{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a643fc58-f147-420f-a9b2-e1a7b7078d63",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "qs 0.25.5\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Warning message:\n",
      "“package ‘IRanges’ was built under R version 4.3.3”\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Warning message:\n",
      "“package ‘AnnotationDbi’ was built under R version 4.3.2”\n",
      "Loading required package: Biobase\n",
      "\n",
      "Warning message:\n",
      "“package ‘Biobase’ was built under R version 4.3.3”\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n",
      "Need specific help about ggbio? try mailing \n",
      " the maintainer or visit https://lawremi.github.io/ggbio/\n",
      "\n",
      "\n",
      "Attaching package: 'ggbio'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ggplot2':\n",
      "\n",
      "    geom_bar, geom_rect, geom_segment, ggsave, stat_bin, stat_identity,\n",
      "    xlim\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'reshape2'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'ggsci' was built under R version 4.3.3\"\n",
      "Warning message:\n",
      "\"package 'viridis' was built under R version 4.3.3\"\n",
      "Loading required package: viridisLite\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Warning message:\n",
      "\"package 'MatrixGenerics' was built under R version 4.3.3\"\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: 'matrixStats'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Biobase':\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'MatrixGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Warning message:\n",
      "\"package 'Biostrings' was built under R version 4.3.3\"\n",
      "Loading required package: XVector\n",
      "\n",
      "Warning message:\n",
      "\"package 'XVector' was built under R version 4.3.3\"\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: BiocIO\n",
      "\n",
      "\n",
      "Attaching package: 'BiocIO'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:rtracklayer':\n",
      "\n",
      "    FileForFormat\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#load package\n",
    ".libPaths(c(\"/cluster/home/liuhengxin/software/miniconda3_2/envs/r-kernel/lib/R/library\",\n",
    "           \"/cluster/home/liuhengxin/software/miniconda3/envs/r4.1.2/lib/R/library\",\n",
    "           \"/cluster/home/liuhengxin/software/miniconda3_2/envs/Rsplot/lib/R/library\"))\n",
    "library(data.table)\n",
    "library(qs)\n",
    "library(GenomicRanges)\n",
    "library(GenomicFeatures)\n",
    "library(rtracklayer)\n",
    "library(dplyr)\n",
    "library(ggbio)\n",
    "library(reshape2)\n",
    "library(ggsci)\n",
    "library(viridis)\n",
    "library(ggpubr)\n",
    "library(ggh4x)\n",
    "library(ggtranscript)\n",
    "library(khroma)\n",
    "library(bambu)\n",
    "setwd(\"/cluster/home/liuhengxin/P1_brain_nanopore/downstream_analysis/\")\n",
    "MySplit = function(str,sep, n){\n",
    "  return(unlist(lapply(strsplit(str,sep),\"[[\",n)))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd414105-6035-40bc-bc67-546c156bd79d",
   "metadata": {},
   "source": [
    "# load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ab502318-511a-4754-a156-fd6aa3e04959",
   "metadata": {},
   "outputs": [],
   "source": [
    "fsraw.t42 = qread(\"data/reads_full_anotation/T42_main_data_read_total_annotation_delete_multiread_total_hubselect_25_12_10.qs\")\n",
    "fsraw.t47 = qread(\"data/reads_full_anotation/T47_main_data_read_total_annotation_delete_multiread_total_hubselect_25_12_10.qs\")\n",
    "fsraw.t44 = qread(\"data/reads_full_anotation/main_data_read_total_annotation_fil_delete_multiread_withsample_25_10_20.qs\")\n",
    "rt_vm116 = qread(\"/cluster/home/liuhengxin/P1_brain_nanopore/processed_data/VM_analysis/Revo/VM11LPT6_Y00983LA_trans_total_anno_withcid_filter_25_12_02.qs\")\n",
    "rt_me01 = qread(\"/cluster/home/liuhengxin/P1_brain_nanopore/processed_data/VM_analysis/Revo/ME012_Y00986G1_trans_total_anno_withcid_filter_25_12_02.qs\")\n",
    "isosigls.t44 = qread(\"data/dtu_analysis/dtu_isoform_ls_fsraw_cor_25_10_10.qs\")\n",
    "isosigls.t42 = qread(\"data/dtu_analysis/T42_dtu_isoform_ls_fsraw_cor_25_12_16.qs\")\n",
    "isosigls.t47 = qread(\"data/dtu_analysis/T47_dtu_isoform_ls_fsraw_cor_25_12_16.qs\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5ac1036-e34e-45a4-afed-c3cdfd0a9ab9",
   "metadata": {},
   "source": [
    "# Table1 isoform count and gtf count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5c6f95f9-7024-42ab-ad7c-25dc48058fbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "datas1 = list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "465b05b5-83f6-417a-bc47-0699cad2e0a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>TXNAME</th><th scope=col>GENEID</th><th scope=col>NDR</th><th scope=col>novelGene</th><th scope=col>novelTranscript</th><th scope=col>txClassDescription</th><th scope=col>readCount</th><th scope=col>relReadCount</th><th scope=col>relSubsetCount</th><th scope=col>txid</th><th scope=col>eqClassById</th><th scope=col>fullLengthCounts</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;list&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>BambuTx1</th><td>BambuTx1</td><td>PGBD2       </td><td>0.5484456</td><td>FALSE</td><td>TRUE</td><td>newLastJunction:newLastExon</td><td>127</td><td>0.413680782</td><td>1.0000000</td><td>1</td><td>        1</td><td>129.00000</td></tr>\n",
       "\t<tr><th scope=row>BambuTx2</th><td>BambuTx2</td><td>ZNF692      </td><td>0.9015748</td><td>FALSE</td><td>TRUE</td><td>newWithin                  </td><td>  7</td><td>0.003280225</td><td>0.5384615</td><td>2</td><td>2, 227827</td><td>  5.00001</td></tr>\n",
       "\t<tr><th scope=row>BambuTx3</th><td>BambuTx3</td><td>ZNF692      </td><td>0.8325641</td><td>FALSE</td><td>TRUE</td><td>newJunction                </td><td> 19</td><td>0.008903468</td><td>0.7916667</td><td>3</td><td>3, 192203</td><td> 19.00000</td></tr>\n",
       "\t<tr><th scope=row>BambuTx4</th><td>BambuTx4</td><td>LOC107130781</td><td>0.2581615</td><td>FALSE</td><td>TRUE</td><td>allNew                     </td><td>629</td><td>0.940209268</td><td>1.0000000</td><td>4</td><td>        4</td><td>629.00000</td></tr>\n",
       "\t<tr><th scope=row>BambuTx5</th><td>BambuTx5</td><td>SH3BP5L     </td><td>0.9214234</td><td>FALSE</td><td>TRUE</td><td>newWithin                  </td><td>  9</td><td>0.005717916</td><td>1.0000000</td><td>5</td><td>        5</td><td>  9.00000</td></tr>\n",
       "\t<tr><th scope=row>BambuTx6</th><td>BambuTx6</td><td>NLRP3       </td><td>0.7253364</td><td>FALSE</td><td>TRUE</td><td>allNew                     </td><td> 33</td><td>0.160194175</td><td>1.0000000</td><td>6</td><td>        6</td><td> 38.00000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 12\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & TXNAME & GENEID & NDR & novelGene & novelTranscript & txClassDescription & readCount & relReadCount & relSubsetCount & txid & eqClassById & fullLengthCounts\\\\\n",
       "  & <chr> & <chr> & <dbl> & <lgl> & <lgl> & <chr> & <int> & <dbl> & <dbl> & <int> & <list> & <dbl>\\\\\n",
       "\\hline\n",
       "\tBambuTx1 & BambuTx1 & PGBD2        & 0.5484456 & FALSE & TRUE & newLastJunction:newLastExon & 127 & 0.413680782 & 1.0000000 & 1 &         1 & 129.00000\\\\\n",
       "\tBambuTx2 & BambuTx2 & ZNF692       & 0.9015748 & FALSE & TRUE & newWithin                   &   7 & 0.003280225 & 0.5384615 & 2 & 2, 227827 &   5.00001\\\\\n",
       "\tBambuTx3 & BambuTx3 & ZNF692       & 0.8325641 & FALSE & TRUE & newJunction                 &  19 & 0.008903468 & 0.7916667 & 3 & 3, 192203 &  19.00000\\\\\n",
       "\tBambuTx4 & BambuTx4 & LOC107130781 & 0.2581615 & FALSE & TRUE & allNew                      & 629 & 0.940209268 & 1.0000000 & 4 &         4 & 629.00000\\\\\n",
       "\tBambuTx5 & BambuTx5 & SH3BP5L      & 0.9214234 & FALSE & TRUE & newWithin                   &   9 & 0.005717916 & 1.0000000 & 5 &         5 &   9.00000\\\\\n",
       "\tBambuTx6 & BambuTx6 & NLRP3        & 0.7253364 & FALSE & TRUE & allNew                      &  33 & 0.160194175 & 1.0000000 & 6 &         6 &  38.00000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 12\n",
       "\n",
       "| <!--/--> | TXNAME &lt;chr&gt; | GENEID &lt;chr&gt; | NDR &lt;dbl&gt; | novelGene &lt;lgl&gt; | novelTranscript &lt;lgl&gt; | txClassDescription &lt;chr&gt; | readCount &lt;int&gt; | relReadCount &lt;dbl&gt; | relSubsetCount &lt;dbl&gt; | txid &lt;int&gt; | eqClassById &lt;list&gt; | fullLengthCounts &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| BambuTx1 | BambuTx1 | PGBD2        | 0.5484456 | FALSE | TRUE | newLastJunction:newLastExon | 127 | 0.413680782 | 1.0000000 | 1 |         1 | 129.00000 |\n",
       "| BambuTx2 | BambuTx2 | ZNF692       | 0.9015748 | FALSE | TRUE | newWithin                   |   7 | 0.003280225 | 0.5384615 | 2 | 2, 227827 |   5.00001 |\n",
       "| BambuTx3 | BambuTx3 | ZNF692       | 0.8325641 | FALSE | TRUE | newJunction                 |  19 | 0.008903468 | 0.7916667 | 3 | 3, 192203 |  19.00000 |\n",
       "| BambuTx4 | BambuTx4 | LOC107130781 | 0.2581615 | FALSE | TRUE | allNew                      | 629 | 0.940209268 | 1.0000000 | 4 |         4 | 629.00000 |\n",
       "| BambuTx5 | BambuTx5 | SH3BP5L      | 0.9214234 | FALSE | TRUE | newWithin                   |   9 | 0.005717916 | 1.0000000 | 5 |         5 |   9.00000 |\n",
       "| BambuTx6 | BambuTx6 | NLRP3        | 0.7253364 | FALSE | TRUE | allNew                      |  33 | 0.160194175 | 1.0000000 | 6 |         6 |  38.00000 |\n",
       "\n"
      ],
      "text/plain": [
       "         TXNAME   GENEID       NDR       novelGene novelTranscript\n",
       "BambuTx1 BambuTx1 PGBD2        0.5484456 FALSE     TRUE           \n",
       "BambuTx2 BambuTx2 ZNF692       0.9015748 FALSE     TRUE           \n",
       "BambuTx3 BambuTx3 ZNF692       0.8325641 FALSE     TRUE           \n",
       "BambuTx4 BambuTx4 LOC107130781 0.2581615 FALSE     TRUE           \n",
       "BambuTx5 BambuTx5 SH3BP5L      0.9214234 FALSE     TRUE           \n",
       "BambuTx6 BambuTx6 NLRP3        0.7253364 FALSE     TRUE           \n",
       "         txClassDescription          readCount relReadCount relSubsetCount txid\n",
       "BambuTx1 newLastJunction:newLastExon 127       0.413680782  1.0000000      1   \n",
       "BambuTx2 newWithin                     7       0.003280225  0.5384615      2   \n",
       "BambuTx3 newJunction                  19       0.008903468  0.7916667      3   \n",
       "BambuTx4 allNew                      629       0.940209268  1.0000000      4   \n",
       "BambuTx5 newWithin                     9       0.005717916  1.0000000      5   \n",
       "BambuTx6 allNew                       33       0.160194175  1.0000000      6   \n",
       "         eqClassById fullLengthCounts\n",
       "BambuTx1         1   129.00000       \n",
       "BambuTx2 2, 227827     5.00001       \n",
       "BambuTx3 3, 192203    19.00000       \n",
       "BambuTx4         4   629.00000       \n",
       "BambuTx5         5     9.00000       \n",
       "BambuTx6         6    38.00000       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtfinfo.t44 = qread(\"ref/bambu_extended_annotations_transinfo.qs\")\n",
    "gtfinfo.t44 = as.data.frame(gtfinfo.t44$transinfo)\n",
    "head(gtfinfo.t44)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bc9750f0-7d5b-4d2e-8f06-5fa1c769fbec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>trantype</th><th scope=col>transnew</th><th scope=col>fulllength</th><th scope=col>transnew2</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>BambuTx1     </td><td>PGBD2       </td><td>novel</td><td>PGBD2-WeiLab-1       </td><td>599</td><td>PGBD2-FS-1       </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>BambuTx10    </td><td>ZNF669      </td><td>novel</td><td>ZNF669-WeiLab-1      </td><td>203</td><td>ZNF669-FS-1      </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>BambuTx1000  </td><td>USP21       </td><td>novel</td><td>USP21-WeiLab-2       </td><td>413</td><td>USP21-FS-2       </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>BambuTx100007</td><td>LOC102142787</td><td>novel</td><td>LOC102142787-WeiLab-1</td><td>105</td><td>LOC102142787-FS-1</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>BambuTx100045</td><td>EXOC3L4     </td><td>novel</td><td>EXOC3L4-WeiLab-1     </td><td> 72</td><td>EXOC3L4-FS-1     </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>BambuTx100047</td><td>EXOC3L4     </td><td>novel</td><td>EXOC3L4-WeiLab-2     </td><td> 37</td><td>EXOC3L4-FS-2     </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & transcript\\_id & gene\\_id & trantype & transnew & fulllength & transnew2\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & BambuTx1      & PGBD2        & novel & PGBD2-WeiLab-1        & 599 & PGBD2-FS-1       \\\\\n",
       "\t2 & BambuTx10     & ZNF669       & novel & ZNF669-WeiLab-1       & 203 & ZNF669-FS-1      \\\\\n",
       "\t3 & BambuTx1000   & USP21        & novel & USP21-WeiLab-2        & 413 & USP21-FS-2       \\\\\n",
       "\t4 & BambuTx100007 & LOC102142787 & novel & LOC102142787-WeiLab-1 & 105 & LOC102142787-FS-1\\\\\n",
       "\t5 & BambuTx100045 & EXOC3L4      & novel & EXOC3L4-WeiLab-1      &  72 & EXOC3L4-FS-1     \\\\\n",
       "\t6 & BambuTx100047 & EXOC3L4      & novel & EXOC3L4-WeiLab-2      &  37 & EXOC3L4-FS-2     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| <!--/--> | transcript_id &lt;chr&gt; | gene_id &lt;chr&gt; | trantype &lt;chr&gt; | transnew &lt;chr&gt; | fulllength &lt;int&gt; | transnew2 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | BambuTx1      | PGBD2        | novel | PGBD2-WeiLab-1        | 599 | PGBD2-FS-1        |\n",
       "| 2 | BambuTx10     | ZNF669       | novel | ZNF669-WeiLab-1       | 203 | ZNF669-FS-1       |\n",
       "| 3 | BambuTx1000   | USP21        | novel | USP21-WeiLab-2        | 413 | USP21-FS-2        |\n",
       "| 4 | BambuTx100007 | LOC102142787 | novel | LOC102142787-WeiLab-1 | 105 | LOC102142787-FS-1 |\n",
       "| 5 | BambuTx100045 | EXOC3L4      | novel | EXOC3L4-WeiLab-1      |  72 | EXOC3L4-FS-1      |\n",
       "| 6 | BambuTx100047 | EXOC3L4      | novel | EXOC3L4-WeiLab-2      |  37 | EXOC3L4-FS-2      |\n",
       "\n"
      ],
      "text/plain": [
       "  transcript_id gene_id      trantype transnew              fulllength\n",
       "1 BambuTx1      PGBD2        novel    PGBD2-WeiLab-1        599       \n",
       "2 BambuTx10     ZNF669       novel    ZNF669-WeiLab-1       203       \n",
       "3 BambuTx1000   USP21        novel    USP21-WeiLab-2        413       \n",
       "4 BambuTx100007 LOC102142787 novel    LOC102142787-WeiLab-1 105       \n",
       "5 BambuTx100045 EXOC3L4      novel    EXOC3L4-WeiLab-1       72       \n",
       "6 BambuTx100047 EXOC3L4      novel    EXOC3L4-WeiLab-2       37       \n",
       "  transnew2        \n",
       "1 PGBD2-FS-1       \n",
       "2 ZNF669-FS-1      \n",
       "3 USP21-FS-2       \n",
       "4 LOC102142787-FS-1\n",
       "5 EXOC3L4-FS-1     \n",
       "6 EXOC3L4-FS-2     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sumfull = read.csv(\"ref/gtf_filter_final_with_newname_25_10_09.csv\")\n",
    "head(sumfull)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cb2025e3-7fc5-43e3-b23c-4be23dbaa960",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "38130"
      ],
      "text/latex": [
       "38130"
      ],
      "text/markdown": [
       "38130"
      ],
      "text/plain": [
       "[1] 38130"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtfinfo.res = merge(gtfinfo.t44[,c(\"TXNAME\",\"GENEID\",\"NDR\",\"novelGene\",\"novelTranscript\",\"txClassDescription\")],\n",
    "                   sumfull[,c(\"transcript_id\",\"transnew2\",\"fulllength\")],by.x = \"TXNAME\",by.y = \"transcript_id\")\n",
    "nrow(gtfinfo.res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fa7adc17-077b-4ba1-86cf-614ad095d545",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>TXNAME</th><th scope=col>GENEID</th><th scope=col>NDR</th><th scope=col>novelGene</th><th scope=col>novelTranscript</th><th scope=col>txClassDescription</th><th scope=col>transnew2</th><th scope=col>fulllength</th><th scope=col>transtype</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>BambuTx1     </td><td>PGBD2       </td><td>0.5484456</td><td>FALSE</td><td>TRUE</td><td>newLastJunction:newLastExon            </td><td>PGBD2-FS-1       </td><td>599</td><td>NNC</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>BambuTx10    </td><td>ZNF669      </td><td>0.7928645</td><td>FALSE</td><td>TRUE</td><td>newLastJunction:newJunction:newLastExon</td><td>ZNF669-FS-1      </td><td>203</td><td>NNC</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>BambuTx1000  </td><td>USP21       </td><td>0.7558931</td><td>FALSE</td><td>TRUE</td><td>newJunction                            </td><td>USP21-FS-2       </td><td>413</td><td>NNC</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>BambuTx100007</td><td>LOC102142787</td><td>0.7591990</td><td>FALSE</td><td>TRUE</td><td>allNew                                 </td><td>LOC102142787-FS-1</td><td>105</td><td>NNC</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>BambuTx100045</td><td>EXOC3L4     </td><td>0.7991295</td><td>FALSE</td><td>TRUE</td><td>newFirstJunction:newFirstExon          </td><td>EXOC3L4-FS-1     </td><td> 72</td><td>NNC</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>BambuTx100047</td><td>EXOC3L4     </td><td>0.7364341</td><td>FALSE</td><td>TRUE</td><td>newLastJunction:newJunction            </td><td>EXOC3L4-FS-2     </td><td> 37</td><td>NNC</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & TXNAME & GENEID & NDR & novelGene & novelTranscript & txClassDescription & transnew2 & fulllength & transtype\\\\\n",
       "  & <chr> & <chr> & <dbl> & <lgl> & <lgl> & <chr> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & BambuTx1      & PGBD2        & 0.5484456 & FALSE & TRUE & newLastJunction:newLastExon             & PGBD2-FS-1        & 599 & NNC\\\\\n",
       "\t2 & BambuTx10     & ZNF669       & 0.7928645 & FALSE & TRUE & newLastJunction:newJunction:newLastExon & ZNF669-FS-1       & 203 & NNC\\\\\n",
       "\t3 & BambuTx1000   & USP21        & 0.7558931 & FALSE & TRUE & newJunction                             & USP21-FS-2        & 413 & NNC\\\\\n",
       "\t4 & BambuTx100007 & LOC102142787 & 0.7591990 & FALSE & TRUE & allNew                                  & LOC102142787-FS-1 & 105 & NNC\\\\\n",
       "\t5 & BambuTx100045 & EXOC3L4      & 0.7991295 & FALSE & TRUE & newFirstJunction:newFirstExon           & EXOC3L4-FS-1      &  72 & NNC\\\\\n",
       "\t6 & BambuTx100047 & EXOC3L4      & 0.7364341 & FALSE & TRUE & newLastJunction:newJunction             & EXOC3L4-FS-2      &  37 & NNC\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | TXNAME &lt;chr&gt; | GENEID &lt;chr&gt; | NDR &lt;dbl&gt; | novelGene &lt;lgl&gt; | novelTranscript &lt;lgl&gt; | txClassDescription &lt;chr&gt; | transnew2 &lt;chr&gt; | fulllength &lt;int&gt; | transtype &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | BambuTx1      | PGBD2        | 0.5484456 | FALSE | TRUE | newLastJunction:newLastExon             | PGBD2-FS-1        | 599 | NNC |\n",
       "| 2 | BambuTx10     | ZNF669       | 0.7928645 | FALSE | TRUE | newLastJunction:newJunction:newLastExon | ZNF669-FS-1       | 203 | NNC |\n",
       "| 3 | BambuTx1000   | USP21        | 0.7558931 | FALSE | TRUE | newJunction                             | USP21-FS-2        | 413 | NNC |\n",
       "| 4 | BambuTx100007 | LOC102142787 | 0.7591990 | FALSE | TRUE | allNew                                  | LOC102142787-FS-1 | 105 | NNC |\n",
       "| 5 | BambuTx100045 | EXOC3L4      | 0.7991295 | FALSE | TRUE | newFirstJunction:newFirstExon           | EXOC3L4-FS-1      |  72 | NNC |\n",
       "| 6 | BambuTx100047 | EXOC3L4      | 0.7364341 | FALSE | TRUE | newLastJunction:newJunction             | EXOC3L4-FS-2      |  37 | NNC |\n",
       "\n"
      ],
      "text/plain": [
       "  TXNAME        GENEID       NDR       novelGene novelTranscript\n",
       "1 BambuTx1      PGBD2        0.5484456 FALSE     TRUE           \n",
       "2 BambuTx10     ZNF669       0.7928645 FALSE     TRUE           \n",
       "3 BambuTx1000   USP21        0.7558931 FALSE     TRUE           \n",
       "4 BambuTx100007 LOC102142787 0.7591990 FALSE     TRUE           \n",
       "5 BambuTx100045 EXOC3L4      0.7991295 FALSE     TRUE           \n",
       "6 BambuTx100047 EXOC3L4      0.7364341 FALSE     TRUE           \n",
       "  txClassDescription                      transnew2         fulllength\n",
       "1 newLastJunction:newLastExon             PGBD2-FS-1        599       \n",
       "2 newLastJunction:newJunction:newLastExon ZNF669-FS-1       203       \n",
       "3 newJunction                             USP21-FS-2        413       \n",
       "4 allNew                                  LOC102142787-FS-1 105       \n",
       "5 newFirstJunction:newFirstExon           EXOC3L4-FS-1       72       \n",
       "6 newLastJunction:newJunction             EXOC3L4-FS-2       37       \n",
       "  transtype\n",
       "1 NNC      \n",
       "2 NNC      \n",
       "3 NNC      \n",
       "4 NNC      \n",
       "5 NNC      \n",
       "6 NNC      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtfinfo.res$transtype = \"NNC\"\n",
    "gtfinfo.res[gtfinfo.res$txClassDescription %in% c(\"newWithin\",\"newLastExon\",\"newFirstExon\",\"newFirstExon:newLastExon\"),\"transtype\"] = \"NIC\"\n",
    "gtfinfo.res[gtfinfo.res$txClassDescription == \"annotation\",\"transtype\"] = \"Known\"\n",
    "head(gtfinfo.res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "93af38bb-a956-432c-bbff-312b1a8bec80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "      1       7      60    5683     433 3852656 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(gtfinfo.res)[7] = \"transcript_id\"\n",
    "summary(gtfinfo.res$fulllength)\n",
    "datas1$gtfinfo = gtfinfo.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fb04331b-8c1c-4fcf-b580-0234cca2057e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fsraw.vm = rt_vm116 %>% group_by(readid) %>% filter(length(unique(transcript_id)) == 1)\n",
    "fsraw.me = rt_me01 %>% group_by(readid) %>% filter(length(unique(transcript_id)) == 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b8e90ef3-236e-4092-b9ec-888c84fe24bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene_id'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'group', 'gene_id'. You can override using\n",
      "the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'group', 'gene_id'. You can override using\n",
      "the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene_id'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene_id'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>gene_id</th><th scope=col>transcript_id</th><th scope=col>count</th><th scope=col>platform</th><th scope=col>sample</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AAAS</td><td>AAAS-FS-1     </td><td>12</td><td>Fullscope-Pacbio</td><td>Macaque1-2</td></tr>\n",
       "\t<tr><td>AAAS</td><td>AAAS-FS-2     </td><td> 6</td><td>Fullscope-Pacbio</td><td>Macaque1-2</td></tr>\n",
       "\t<tr><td>AAAS</td><td>XM_005571000.2</td><td> 6</td><td>Fullscope-Pacbio</td><td>Macaque1-2</td></tr>\n",
       "\t<tr><td>AAAS</td><td>XM_015430971.1</td><td> 2</td><td>Fullscope-Pacbio</td><td>Macaque1-2</td></tr>\n",
       "\t<tr><td>AAAS</td><td>XM_015430972.1</td><td> 4</td><td>Fullscope-Pacbio</td><td>Macaque1-2</td></tr>\n",
       "\t<tr><td>AACS</td><td>NM_001284010.1</td><td>70</td><td>Fullscope-Pacbio</td><td>Macaque1-2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 6 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " gene\\_id & transcript\\_id & count & platform & sample\\\\\n",
       " <chr> & <chr> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t AAAS & AAAS-FS-1      & 12 & Fullscope-Pacbio & Macaque1-2\\\\\n",
       "\t AAAS & AAAS-FS-2      &  6 & Fullscope-Pacbio & Macaque1-2\\\\\n",
       "\t AAAS & XM\\_005571000.2 &  6 & Fullscope-Pacbio & Macaque1-2\\\\\n",
       "\t AAAS & XM\\_015430971.1 &  2 & Fullscope-Pacbio & Macaque1-2\\\\\n",
       "\t AAAS & XM\\_015430972.1 &  4 & Fullscope-Pacbio & Macaque1-2\\\\\n",
       "\t AACS & NM\\_001284010.1 & 70 & Fullscope-Pacbio & Macaque1-2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 6 × 5\n",
       "\n",
       "| gene_id &lt;chr&gt; | transcript_id &lt;chr&gt; | count &lt;int&gt; | platform &lt;chr&gt; | sample &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| AAAS | AAAS-FS-1      | 12 | Fullscope-Pacbio | Macaque1-2 |\n",
       "| AAAS | AAAS-FS-2      |  6 | Fullscope-Pacbio | Macaque1-2 |\n",
       "| AAAS | XM_005571000.2 |  6 | Fullscope-Pacbio | Macaque1-2 |\n",
       "| AAAS | XM_015430971.1 |  2 | Fullscope-Pacbio | Macaque1-2 |\n",
       "| AAAS | XM_015430972.1 |  4 | Fullscope-Pacbio | Macaque1-2 |\n",
       "| AACS | NM_001284010.1 | 70 | Fullscope-Pacbio | Macaque1-2 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene_id transcript_id  count platform         sample    \n",
       "1 AAAS    AAAS-FS-1      12    Fullscope-Pacbio Macaque1-2\n",
       "2 AAAS    AAAS-FS-2       6    Fullscope-Pacbio Macaque1-2\n",
       "3 AAAS    XM_005571000.2  6    Fullscope-Pacbio Macaque1-2\n",
       "4 AAAS    XM_015430971.1  2    Fullscope-Pacbio Macaque1-2\n",
       "5 AAAS    XM_015430972.1  4    Fullscope-Pacbio Macaque1-2\n",
       "6 AACS    NM_001284010.1 70    Fullscope-Pacbio Macaque1-2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##T42 stat raw count for every sample\n",
    "transt.t42 = fsraw.t42 %>% group_by(gene_id,transcript_id) %>% summarise(count = n())\n",
    "transt.t42$platform = \"Fullscope-Pacbio\"\n",
    "transt.t42$sample = \"Macaque1-2\"\n",
    "\n",
    "#T44\n",
    "transt.t44 = fsraw.t44 %>% group_by(group,gene_id,transcript_id) %>% summarise(count = n())\n",
    "renamedf = data.frame(\"group\" = c('ontf','cycf','pacbio','ontd'),\n",
    "                      \"platform\" = c(\"Fullscope-ONT\",\"Fullscope-Cyclone\",\n",
    "                                     \"Fullscope-Pacbio\",\"Direct-ONT\"))\n",
    "transt.t44 = merge(transt.t44,renamedf,by = \"group\")\n",
    "transt.t44$sample = \"Macaque1-1\"\n",
    "transt.t44 = transt.t44[,c(\"gene_id\",\"transcript_id\",\"platform\",\"sample\",\"count\")]\n",
    "\n",
    "#T47\n",
    "transt.t47 = fsraw.t47 %>% group_by(group,gene_id,transcript_id) %>% summarise(count = n())\n",
    "renamedf = data.frame(\"group\" = c('Pacbio','ONT','Cyclone'),\n",
    "                      \"platform\" = c(\"Fullscope-Pacbio\",\"Fullscope-ONT\",\"Fullscope-Cyclone\"))\n",
    "transt.t47 = merge(transt.t47,renamedf,by = \"group\")\n",
    "transt.t47$sample = \"Macaque1-3\"\n",
    "transt.t47 = transt.t47[,c(\"gene_id\",\"transcript_id\",\"platform\",\"sample\",\"count\")]\n",
    "\n",
    "#VM\n",
    "transt.vm = fsraw.vm %>% group_by(gene_id,transcript_id) %>% summarise(count = n())\n",
    "transt.vm$platform = \"Fullscope-Pacbio\"\n",
    "transt.vm$sample = \"Macaque2-1\"\n",
    "\n",
    "#ME\n",
    "transt.me = fsraw.me %>% group_by(gene_id,transcript_id) %>% summarise(count = n())\n",
    "transt.me$platform = \"Fullscope-Pacbio\"\n",
    "transt.me$sample = \"Macaque3-1\"\n",
    "\n",
    "transtl = rbind(transt.t42,transt.t44,transt.t47,transt.vm,transt.me)\n",
    "head(transtl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7591b569-cbb4-41f8-a84f-b702a1353485",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene_id</th><th scope=col>transcript_id</th><th scope=col>Macaque1-1_Direct-ONT</th><th scope=col>Macaque1-1_Fullscope-Cyclone</th><th scope=col>Macaque1-1_Fullscope-ONT</th><th scope=col>Macaque1-1_Fullscope-Pacbio</th><th scope=col>Macaque1-2_Fullscope-Pacbio</th><th scope=col>Macaque1-3_Fullscope-Cyclone</th><th scope=col>Macaque1-3_Fullscope-ONT</th><th scope=col>Macaque1-3_Fullscope-Pacbio</th><th scope=col>Macaque2-1_Fullscope-Pacbio</th><th scope=col>Macaque3-1_Fullscope-Pacbio</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A4GALT</td><td>XM_015457101.1</td><td>0</td><td> 28</td><td> 49</td><td>59</td><td> 0</td><td>2</td><td>0</td><td> 0</td><td>4</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A4GALT</td><td>XM_015457102.1</td><td>2</td><td>  2</td><td>  0</td><td> 0</td><td> 0</td><td>0</td><td>0</td><td> 0</td><td>0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>AAAS  </td><td>AAAS-FS-1     </td><td>3</td><td> 49</td><td> 92</td><td>35</td><td>12</td><td>3</td><td>0</td><td>16</td><td>0</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>AAAS  </td><td>AAAS-FS-2     </td><td>9</td><td>  3</td><td> 18</td><td> 7</td><td> 6</td><td>3</td><td>0</td><td>12</td><td>3</td><td> 0</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>AAAS  </td><td>XM_005571000.2</td><td>3</td><td>375</td><td>589</td><td>11</td><td> 6</td><td>0</td><td>0</td><td>18</td><td>3</td><td>19</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>AAAS  </td><td>XM_015430971.1</td><td>0</td><td>  3</td><td>  4</td><td> 3</td><td> 2</td><td>0</td><td>0</td><td> 2</td><td>0</td><td> 0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 12\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & gene\\_id & transcript\\_id & Macaque1-1\\_Direct-ONT & Macaque1-1\\_Fullscope-Cyclone & Macaque1-1\\_Fullscope-ONT & Macaque1-1\\_Fullscope-Pacbio & Macaque1-2\\_Fullscope-Pacbio & Macaque1-3\\_Fullscope-Cyclone & Macaque1-3\\_Fullscope-ONT & Macaque1-3\\_Fullscope-Pacbio & Macaque2-1\\_Fullscope-Pacbio & Macaque3-1\\_Fullscope-Pacbio\\\\\n",
       "  & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & A4GALT & XM\\_015457101.1 & 0 &  28 &  49 & 59 &  0 & 2 & 0 &  0 & 4 &  0\\\\\n",
       "\t2 & A4GALT & XM\\_015457102.1 & 2 &   2 &   0 &  0 &  0 & 0 & 0 &  0 & 0 &  0\\\\\n",
       "\t3 & AAAS   & AAAS-FS-1      & 3 &  49 &  92 & 35 & 12 & 3 & 0 & 16 & 0 &  0\\\\\n",
       "\t4 & AAAS   & AAAS-FS-2      & 9 &   3 &  18 &  7 &  6 & 3 & 0 & 12 & 3 &  0\\\\\n",
       "\t5 & AAAS   & XM\\_005571000.2 & 3 & 375 & 589 & 11 &  6 & 0 & 0 & 18 & 3 & 19\\\\\n",
       "\t6 & AAAS   & XM\\_015430971.1 & 0 &   3 &   4 &  3 &  2 & 0 & 0 &  2 & 0 &  0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 12\n",
       "\n",
       "| <!--/--> | gene_id &lt;chr&gt; | transcript_id &lt;chr&gt; | Macaque1-1_Direct-ONT &lt;dbl&gt; | Macaque1-1_Fullscope-Cyclone &lt;dbl&gt; | Macaque1-1_Fullscope-ONT &lt;dbl&gt; | Macaque1-1_Fullscope-Pacbio &lt;dbl&gt; | Macaque1-2_Fullscope-Pacbio &lt;dbl&gt; | Macaque1-3_Fullscope-Cyclone &lt;dbl&gt; | Macaque1-3_Fullscope-ONT &lt;dbl&gt; | Macaque1-3_Fullscope-Pacbio &lt;dbl&gt; | Macaque2-1_Fullscope-Pacbio &lt;dbl&gt; | Macaque3-1_Fullscope-Pacbio &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A4GALT | XM_015457101.1 | 0 |  28 |  49 | 59 |  0 | 2 | 0 |  0 | 4 |  0 |\n",
       "| 2 | A4GALT | XM_015457102.1 | 2 |   2 |   0 |  0 |  0 | 0 | 0 |  0 | 0 |  0 |\n",
       "| 3 | AAAS   | AAAS-FS-1      | 3 |  49 |  92 | 35 | 12 | 3 | 0 | 16 | 0 |  0 |\n",
       "| 4 | AAAS   | AAAS-FS-2      | 9 |   3 |  18 |  7 |  6 | 3 | 0 | 12 | 3 |  0 |\n",
       "| 5 | AAAS   | XM_005571000.2 | 3 | 375 | 589 | 11 |  6 | 0 | 0 | 18 | 3 | 19 |\n",
       "| 6 | AAAS   | XM_015430971.1 | 0 |   3 |   4 |  3 |  2 | 0 | 0 |  2 | 0 |  0 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene_id transcript_id  Macaque1-1_Direct-ONT Macaque1-1_Fullscope-Cyclone\n",
       "1 A4GALT  XM_015457101.1 0                      28                         \n",
       "2 A4GALT  XM_015457102.1 2                       2                         \n",
       "3 AAAS    AAAS-FS-1      3                      49                         \n",
       "4 AAAS    AAAS-FS-2      9                       3                         \n",
       "5 AAAS    XM_005571000.2 3                     375                         \n",
       "6 AAAS    XM_015430971.1 0                       3                         \n",
       "  Macaque1-1_Fullscope-ONT Macaque1-1_Fullscope-Pacbio\n",
       "1  49                      59                         \n",
       "2   0                       0                         \n",
       "3  92                      35                         \n",
       "4  18                       7                         \n",
       "5 589                      11                         \n",
       "6   4                       3                         \n",
       "  Macaque1-2_Fullscope-Pacbio Macaque1-3_Fullscope-Cyclone\n",
       "1  0                          2                           \n",
       "2  0                          0                           \n",
       "3 12                          3                           \n",
       "4  6                          3                           \n",
       "5  6                          0                           \n",
       "6  2                          0                           \n",
       "  Macaque1-3_Fullscope-ONT Macaque1-3_Fullscope-Pacbio\n",
       "1 0                         0                         \n",
       "2 0                         0                         \n",
       "3 0                        16                         \n",
       "4 0                        12                         \n",
       "5 0                        18                         \n",
       "6 0                         2                         \n",
       "  Macaque2-1_Fullscope-Pacbio Macaque3-1_Fullscope-Pacbio\n",
       "1 4                            0                         \n",
       "2 0                            0                         \n",
       "3 0                            0                         \n",
       "4 3                            0                         \n",
       "5 3                           19                         \n",
       "6 0                            0                         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>36851</li><li>12</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 36851\n",
       "\\item 12\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 36851\n",
       "2. 12\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 36851    12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "transtmx = dcast(transtl,gene_id+transcript_id~sample+platform,value.var = \"count\",fill = 0,fun.aggregate = sum)\n",
    "head(transtmx)\n",
    "dim(transtmx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "427ca074-8710-4ceb-b470-9f0ced61de37",
   "metadata": {},
   "outputs": [],
   "source": [
    "datas1$transtmx = transtmx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ae67a8bd-542b-413b-806c-593c3e6ed239",
   "metadata": {},
   "outputs": [],
   "source": [
    "#library(xlsx)\n",
    "# install.packages(\"writexl\")\n",
    "writexl::write_xlsx(datas1, \"/cluster/home/liuhengxin/P1_brain_nanopore/downstream_analysis/result/figure_revision/data_summary/tables1_26_01_07.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82a8946f-ca68-4519-8ee4-500e88092eb7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Table2 isoform DTU calculation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "0014d7c3-9c5b-43ae-84b5-f854e3cfd6fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'layer'</li><li>'region'</li><li>'lobe'</li><li>'subclass'</li><li>'class'</li><li>'subcortical'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'layer'\n",
       "\\item 'region'\n",
       "\\item 'lobe'\n",
       "\\item 'subclass'\n",
       "\\item 'class'\n",
       "\\item 'subcortical'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'layer'\n",
       "2. 'region'\n",
       "3. 'lobe'\n",
       "4. 'subclass'\n",
       "5. 'class'\n",
       "6. 'subcortical'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"layer\"       \"region\"      \"lobe\"        \"subclass\"    \"class\"      \n",
       "[6] \"subcortical\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'layer'</li><li>'region'</li><li>'lobe'</li><li>'subclass'</li><li>'class'</li><li>'subcortical'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'layer'\n",
       "\\item 'region'\n",
       "\\item 'lobe'\n",
       "\\item 'subclass'\n",
       "\\item 'class'\n",
       "\\item 'subcortical'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'layer'\n",
       "2. 'region'\n",
       "3. 'lobe'\n",
       "4. 'subclass'\n",
       "5. 'class'\n",
       "6. 'subcortical'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"layer\"       \"region\"      \"lobe\"        \"subclass\"    \"class\"      \n",
       "[6] \"subcortical\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'layer'</li><li>'region'</li><li>'lobe'</li><li>'subclass'</li><li>'class'</li><li>'subcortical'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'layer'\n",
       "\\item 'region'\n",
       "\\item 'lobe'\n",
       "\\item 'subclass'\n",
       "\\item 'class'\n",
       "\\item 'subcortical'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'layer'\n",
       "2. 'region'\n",
       "3. 'lobe'\n",
       "4. 'subclass'\n",
       "5. 'class'\n",
       "6. 'subcortical'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"layer\"       \"region\"      \"lobe\"        \"subclass\"    \"class\"      \n",
       "[6] \"subcortical\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene_id</th><th scope=col>transcript_id</th><th scope=col>areatr</th><th scope=col>count</th><th scope=col>genecount</th><th scope=col>relative</th><th scope=col>meanrelative</th><th scope=col>otherrelative</th><th scope=col>dr</th><th scope=col>sig</th><th scope=col>pvalue</th><th scope=col>padj</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>AAAS</td><td>AAAS-FS-1</td><td>l1</td><td> 4</td><td> 24</td><td>0.16666667</td><td>0.1495626</td><td>0.1461418</td><td> 0.02052482</td><td>N </td><td>0.0001939086</td><td>0.002006998</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>AAAS</td><td>AAAS-FS-1</td><td>l2</td><td> 3</td><td> 41</td><td>0.07317073</td><td>0.1495626</td><td>0.1648410</td><td>-0.09167030</td><td>N </td><td>0.0001939086</td><td>0.002006998</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>AAAS</td><td>AAAS-FS-1</td><td>l3</td><td>45</td><td>346</td><td>0.13005780</td><td>0.1495626</td><td>0.1534636</td><td>-0.02340581</td><td>N </td><td>0.0001939086</td><td>0.002006998</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>AAAS</td><td>AAAS-FS-1</td><td>l4</td><td>14</td><td> 66</td><td>0.21212121</td><td>0.1495626</td><td>0.1370509</td><td> 0.07507028</td><td>N </td><td>0.0001939086</td><td>0.002006998</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>AAAS</td><td>AAAS-FS-1</td><td>l5</td><td>10</td><td>153</td><td>0.06535948</td><td>0.1495626</td><td>0.1664033</td><td>-0.10104381</td><td>N </td><td>0.0001939086</td><td>0.002006998</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>AAAS</td><td>AAAS-FS-1</td><td>l6</td><td>28</td><td>112</td><td>0.25000000</td><td>0.1495626</td><td>0.1294752</td><td> 0.12052482</td><td>Up</td><td>0.0001939086</td><td>0.002006998</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 12\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & gene\\_id & transcript\\_id & areatr & count & genecount & relative & meanrelative & otherrelative & dr & sig & pvalue & padj\\\\\n",
       "  & <chr> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & AAAS & AAAS-FS-1 & l1 &  4 &  24 & 0.16666667 & 0.1495626 & 0.1461418 &  0.02052482 & N  & 0.0001939086 & 0.002006998\\\\\n",
       "\t2 & AAAS & AAAS-FS-1 & l2 &  3 &  41 & 0.07317073 & 0.1495626 & 0.1648410 & -0.09167030 & N  & 0.0001939086 & 0.002006998\\\\\n",
       "\t3 & AAAS & AAAS-FS-1 & l3 & 45 & 346 & 0.13005780 & 0.1495626 & 0.1534636 & -0.02340581 & N  & 0.0001939086 & 0.002006998\\\\\n",
       "\t4 & AAAS & AAAS-FS-1 & l4 & 14 &  66 & 0.21212121 & 0.1495626 & 0.1370509 &  0.07507028 & N  & 0.0001939086 & 0.002006998\\\\\n",
       "\t5 & AAAS & AAAS-FS-1 & l5 & 10 & 153 & 0.06535948 & 0.1495626 & 0.1664033 & -0.10104381 & N  & 0.0001939086 & 0.002006998\\\\\n",
       "\t6 & AAAS & AAAS-FS-1 & l6 & 28 & 112 & 0.25000000 & 0.1495626 & 0.1294752 &  0.12052482 & Up & 0.0001939086 & 0.002006998\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 12\n",
       "\n",
       "| <!--/--> | gene_id &lt;chr&gt; | transcript_id &lt;chr&gt; | areatr &lt;chr&gt; | count &lt;int&gt; | genecount &lt;int&gt; | relative &lt;dbl&gt; | meanrelative &lt;dbl&gt; | otherrelative &lt;dbl&gt; | dr &lt;dbl&gt; | sig &lt;chr&gt; | pvalue &lt;dbl&gt; | padj &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | AAAS | AAAS-FS-1 | l1 |  4 |  24 | 0.16666667 | 0.1495626 | 0.1461418 |  0.02052482 | N  | 0.0001939086 | 0.002006998 |\n",
       "| 2 | AAAS | AAAS-FS-1 | l2 |  3 |  41 | 0.07317073 | 0.1495626 | 0.1648410 | -0.09167030 | N  | 0.0001939086 | 0.002006998 |\n",
       "| 3 | AAAS | AAAS-FS-1 | l3 | 45 | 346 | 0.13005780 | 0.1495626 | 0.1534636 | -0.02340581 | N  | 0.0001939086 | 0.002006998 |\n",
       "| 4 | AAAS | AAAS-FS-1 | l4 | 14 |  66 | 0.21212121 | 0.1495626 | 0.1370509 |  0.07507028 | N  | 0.0001939086 | 0.002006998 |\n",
       "| 5 | AAAS | AAAS-FS-1 | l5 | 10 | 153 | 0.06535948 | 0.1495626 | 0.1664033 | -0.10104381 | N  | 0.0001939086 | 0.002006998 |\n",
       "| 6 | AAAS | AAAS-FS-1 | l6 | 28 | 112 | 0.25000000 | 0.1495626 | 0.1294752 |  0.12052482 | Up | 0.0001939086 | 0.002006998 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene_id transcript_id areatr count genecount relative   meanrelative\n",
       "1 AAAS    AAAS-FS-1     l1      4     24       0.16666667 0.1495626   \n",
       "2 AAAS    AAAS-FS-1     l2      3     41       0.07317073 0.1495626   \n",
       "3 AAAS    AAAS-FS-1     l3     45    346       0.13005780 0.1495626   \n",
       "4 AAAS    AAAS-FS-1     l4     14     66       0.21212121 0.1495626   \n",
       "5 AAAS    AAAS-FS-1     l5     10    153       0.06535948 0.1495626   \n",
       "6 AAAS    AAAS-FS-1     l6     28    112       0.25000000 0.1495626   \n",
       "  otherrelative dr          sig pvalue       padj       \n",
       "1 0.1461418      0.02052482 N   0.0001939086 0.002006998\n",
       "2 0.1648410     -0.09167030 N   0.0001939086 0.002006998\n",
       "3 0.1534636     -0.02340581 N   0.0001939086 0.002006998\n",
       "4 0.1370509      0.07507028 N   0.0001939086 0.002006998\n",
       "5 0.1664033     -0.10104381 N   0.0001939086 0.002006998\n",
       "6 0.1294752      0.12052482 Up  0.0001939086 0.002006998"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "names(isosigls.t44)\n",
    "names(isosigls.t47)\n",
    "names(isosigls.t42)\n",
    "i = 1\n",
    "isosigi.t44 = isosigls.t44[[i]]\n",
    "head(isosigi.t44)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b1d59c6f-b610-4bf5-8298-69f901808753",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 21</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene_id</th><th scope=col>transcript_id</th><th scope=col>areatr</th><th scope=col>sig</th><th scope=col>pvalue</th><th scope=col>padj</th><th scope=col>count.M1-1</th><th scope=col>genecount.M1-1</th><th scope=col>relative.M1-1</th><th scope=col>otherrelative.M1-1</th><th scope=col>⋯</th><th scope=col>count.M1-2</th><th scope=col>genecount.M1-2</th><th scope=col>relative.M1-2</th><th scope=col>otherrelative.M1-2</th><th scope=col>dr.M1-2</th><th scope=col>count.M1-3</th><th scope=col>genecount.M1-3</th><th scope=col>relative.M1-3</th><th scope=col>otherrelative.M1-3</th><th scope=col>dr.M1-3</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ACP1</td><td>XM_005576671.2</td><td>l2</td><td>Up</td><td>2.200806e-06</td><td>4.188992e-05</td><td>159</td><td> 439</td><td>0.3621868</td><td>0.2617331</td><td>⋯</td><td> 26</td><td> 54</td><td>0.4814815</td><td>0.45376915</td><td> 0.02771233</td><td>109</td><td>173</td><td>0.6300578</td><td>0.46530199</td><td> 0.1647558176</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ACP1</td><td>XM_005576671.2</td><td>l3</td><td>N </td><td>2.200806e-06</td><td>4.188992e-05</td><td>565</td><td>1952</td><td>0.2894467</td><td>0.2762811</td><td>⋯</td><td>130</td><td>300</td><td>0.4333333</td><td>0.46580618</td><td>-0.03247285</td><td>240</td><td>530</td><td>0.4528302</td><td>0.50074751</td><td>-0.0479173201</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ACP1</td><td>XM_005576671.2</td><td>l4</td><td>N </td><td>2.200806e-06</td><td>4.188992e-05</td><td> 72</td><td> 301</td><td>0.2392027</td><td>0.2863299</td><td>⋯</td><td> 10</td><td> 22</td><td>0.4545455</td><td>0.46050315</td><td>-0.00595770</td><td> 28</td><td> 74</td><td>0.3783784</td><td>0.51563787</td><td>-0.1372594925</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ACP1</td><td>XM_005576671.2</td><td>l5</td><td>N </td><td>2.200806e-06</td><td>4.188992e-05</td><td>182</td><td> 781</td><td>0.2330346</td><td>0.2875635</td><td>⋯</td><td> 50</td><td>104</td><td>0.4807692</td><td>0.45394721</td><td> 0.02682202</td><td>144</td><td>276</td><td>0.5217391</td><td>0.48696572</td><td> 0.0347734100</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ACP1</td><td>XM_005576671.2</td><td>l6</td><td>N </td><td>2.200806e-06</td><td>4.188992e-05</td><td>194</td><td> 758</td><td>0.2559367</td><td>0.2829831</td><td>⋯</td><td> 50</td><td>112</td><td>0.4464286</td><td>0.46253238</td><td>-0.01610380</td><td>115</td><td>233</td><td>0.4935622</td><td>0.49260110</td><td> 0.0009611316</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>AIDA</td><td>NM_001283182.1</td><td>l3</td><td>N </td><td>2.210511e-04</td><td>2.211496e-03</td><td>577</td><td> 649</td><td>0.8890601</td><td>0.8432234</td><td>⋯</td><td>  2</td><td> 32</td><td>0.0625000</td><td>0.09090909</td><td>-0.02840909</td><td> 11</td><td> 88</td><td>0.1250000</td><td>0.08663835</td><td> 0.0383616514</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 21\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & gene\\_id & transcript\\_id & areatr & sig & pvalue & padj & count.M1-1 & genecount.M1-1 & relative.M1-1 & otherrelative.M1-1 & ⋯ & count.M1-2 & genecount.M1-2 & relative.M1-2 & otherrelative.M1-2 & dr.M1-2 & count.M1-3 & genecount.M1-3 & relative.M1-3 & otherrelative.M1-3 & dr.M1-3\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & ⋯ & <int> & <int> & <dbl> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & ACP1 & XM\\_005576671.2 & l2 & Up & 2.200806e-06 & 4.188992e-05 & 159 &  439 & 0.3621868 & 0.2617331 & ⋯ &  26 &  54 & 0.4814815 & 0.45376915 &  0.02771233 & 109 & 173 & 0.6300578 & 0.46530199 &  0.1647558176\\\\\n",
       "\t2 & ACP1 & XM\\_005576671.2 & l3 & N  & 2.200806e-06 & 4.188992e-05 & 565 & 1952 & 0.2894467 & 0.2762811 & ⋯ & 130 & 300 & 0.4333333 & 0.46580618 & -0.03247285 & 240 & 530 & 0.4528302 & 0.50074751 & -0.0479173201\\\\\n",
       "\t3 & ACP1 & XM\\_005576671.2 & l4 & N  & 2.200806e-06 & 4.188992e-05 &  72 &  301 & 0.2392027 & 0.2863299 & ⋯ &  10 &  22 & 0.4545455 & 0.46050315 & -0.00595770 &  28 &  74 & 0.3783784 & 0.51563787 & -0.1372594925\\\\\n",
       "\t4 & ACP1 & XM\\_005576671.2 & l5 & N  & 2.200806e-06 & 4.188992e-05 & 182 &  781 & 0.2330346 & 0.2875635 & ⋯ &  50 & 104 & 0.4807692 & 0.45394721 &  0.02682202 & 144 & 276 & 0.5217391 & 0.48696572 &  0.0347734100\\\\\n",
       "\t5 & ACP1 & XM\\_005576671.2 & l6 & N  & 2.200806e-06 & 4.188992e-05 & 194 &  758 & 0.2559367 & 0.2829831 & ⋯ &  50 & 112 & 0.4464286 & 0.46253238 & -0.01610380 & 115 & 233 & 0.4935622 & 0.49260110 &  0.0009611316\\\\\n",
       "\t6 & AIDA & NM\\_001283182.1 & l3 & N  & 2.210511e-04 & 2.211496e-03 & 577 &  649 & 0.8890601 & 0.8432234 & ⋯ &   2 &  32 & 0.0625000 & 0.09090909 & -0.02840909 &  11 &  88 & 0.1250000 & 0.08663835 &  0.0383616514\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 21\n",
       "\n",
       "| <!--/--> | gene_id &lt;chr&gt; | transcript_id &lt;chr&gt; | areatr &lt;chr&gt; | sig &lt;chr&gt; | pvalue &lt;dbl&gt; | padj &lt;dbl&gt; | count.M1-1 &lt;int&gt; | genecount.M1-1 &lt;int&gt; | relative.M1-1 &lt;dbl&gt; | otherrelative.M1-1 &lt;dbl&gt; | ⋯ ⋯ | count.M1-2 &lt;int&gt; | genecount.M1-2 &lt;int&gt; | relative.M1-2 &lt;dbl&gt; | otherrelative.M1-2 &lt;dbl&gt; | dr.M1-2 &lt;dbl&gt; | count.M1-3 &lt;int&gt; | genecount.M1-3 &lt;int&gt; | relative.M1-3 &lt;dbl&gt; | otherrelative.M1-3 &lt;dbl&gt; | dr.M1-3 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | ACP1 | XM_005576671.2 | l2 | Up | 2.200806e-06 | 4.188992e-05 | 159 |  439 | 0.3621868 | 0.2617331 | ⋯ |  26 |  54 | 0.4814815 | 0.45376915 |  0.02771233 | 109 | 173 | 0.6300578 | 0.46530199 |  0.1647558176 |\n",
       "| 2 | ACP1 | XM_005576671.2 | l3 | N  | 2.200806e-06 | 4.188992e-05 | 565 | 1952 | 0.2894467 | 0.2762811 | ⋯ | 130 | 300 | 0.4333333 | 0.46580618 | -0.03247285 | 240 | 530 | 0.4528302 | 0.50074751 | -0.0479173201 |\n",
       "| 3 | ACP1 | XM_005576671.2 | l4 | N  | 2.200806e-06 | 4.188992e-05 |  72 |  301 | 0.2392027 | 0.2863299 | ⋯ |  10 |  22 | 0.4545455 | 0.46050315 | -0.00595770 |  28 |  74 | 0.3783784 | 0.51563787 | -0.1372594925 |\n",
       "| 4 | ACP1 | XM_005576671.2 | l5 | N  | 2.200806e-06 | 4.188992e-05 | 182 |  781 | 0.2330346 | 0.2875635 | ⋯ |  50 | 104 | 0.4807692 | 0.45394721 |  0.02682202 | 144 | 276 | 0.5217391 | 0.48696572 |  0.0347734100 |\n",
       "| 5 | ACP1 | XM_005576671.2 | l6 | N  | 2.200806e-06 | 4.188992e-05 | 194 |  758 | 0.2559367 | 0.2829831 | ⋯ |  50 | 112 | 0.4464286 | 0.46253238 | -0.01610380 | 115 | 233 | 0.4935622 | 0.49260110 |  0.0009611316 |\n",
       "| 6 | AIDA | NM_001283182.1 | l3 | N  | 2.210511e-04 | 2.211496e-03 | 577 |  649 | 0.8890601 | 0.8432234 | ⋯ |   2 |  32 | 0.0625000 | 0.09090909 | -0.02840909 |  11 |  88 | 0.1250000 | 0.08663835 |  0.0383616514 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene_id transcript_id  areatr sig pvalue       padj         count.M1-1\n",
       "1 ACP1    XM_005576671.2 l2     Up  2.200806e-06 4.188992e-05 159       \n",
       "2 ACP1    XM_005576671.2 l3     N   2.200806e-06 4.188992e-05 565       \n",
       "3 ACP1    XM_005576671.2 l4     N   2.200806e-06 4.188992e-05  72       \n",
       "4 ACP1    XM_005576671.2 l5     N   2.200806e-06 4.188992e-05 182       \n",
       "5 ACP1    XM_005576671.2 l6     N   2.200806e-06 4.188992e-05 194       \n",
       "6 AIDA    NM_001283182.1 l3     N   2.210511e-04 2.211496e-03 577       \n",
       "  genecount.M1-1 relative.M1-1 otherrelative.M1-1 ⋯ count.M1-2 genecount.M1-2\n",
       "1  439           0.3621868     0.2617331          ⋯  26         54           \n",
       "2 1952           0.2894467     0.2762811          ⋯ 130        300           \n",
       "3  301           0.2392027     0.2863299          ⋯  10         22           \n",
       "4  781           0.2330346     0.2875635          ⋯  50        104           \n",
       "5  758           0.2559367     0.2829831          ⋯  50        112           \n",
       "6  649           0.8890601     0.8432234          ⋯   2         32           \n",
       "  relative.M1-2 otherrelative.M1-2 dr.M1-2     count.M1-3 genecount.M1-3\n",
       "1 0.4814815     0.45376915          0.02771233 109        173           \n",
       "2 0.4333333     0.46580618         -0.03247285 240        530           \n",
       "3 0.4545455     0.46050315         -0.00595770  28         74           \n",
       "4 0.4807692     0.45394721          0.02682202 144        276           \n",
       "5 0.4464286     0.46253238         -0.01610380 115        233           \n",
       "6 0.0625000     0.09090909         -0.02840909  11         88           \n",
       "  relative.M1-3 otherrelative.M1-3 dr.M1-3      \n",
       "1 0.6300578     0.46530199          0.1647558176\n",
       "2 0.4528302     0.50074751         -0.0479173201\n",
       "3 0.3783784     0.51563787         -0.1372594925\n",
       "4 0.5217391     0.48696572          0.0347734100\n",
       "5 0.4935622     0.49260110          0.0009611316\n",
       "6 0.1250000     0.08663835          0.0383616514"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data2ls = list()\n",
    "\n",
    "for(i in 1:6){\n",
    "    isosigi.t44 = isosigls.t44[[i]]\n",
    "    isosigi.t47 = isosigls.t47[[i]]\n",
    "    isosigi.t42 = isosigls.t42[[i]]\n",
    "    \n",
    "    hubtrans = isosigi.t44[isosigi.t44$sig != \"N\",]$transcript_id\n",
    "    isosigi.t44 = isosigi.t44[isosigi.t44$transcript_id %in% hubtrans,]\n",
    "    isosigi.t47 = isosigi.t47[isosigi.t47$transcript_id %in% hubtrans,]\n",
    "    isosigi.t42 = isosigi.t42[isosigi.t42$transcript_id %in% hubtrans,]\n",
    "    commonid1 = c(\"gene_id\",\"transcript_id\",\"areatr\",\"sig\",\"pvalue\",\"padj\",\n",
    "                  \"count\",\"genecount\",\"relative\",\"otherrelative\",\"dr\")\n",
    "    commonid2 = c(\"gene_id\",\"transcript_id\",\"areatr\",\"count\",\"genecount\",\"relative\",\"otherrelative\",\"dr\")\n",
    "    \n",
    "    isosigi = merge(isosigi.t44[,commonid1],isosigi.t42[,commonid2],by = c(\"gene_id\",\"transcript_id\",\"areatr\"),\n",
    "                    suffixes = c(\".M1-1\",\".M1-2\"))\n",
    "    isosigi.t47 = isosigi.t47[,commonid2]\n",
    "    colnames(isosigi.t47)[4:8] = paste0(colnames(isosigi.t47)[4:8],\".M1-3\")\n",
    "    isosigi = merge(isosigi,isosigi.t47,by = c(\"gene_id\",\"transcript_id\",\"areatr\"))\n",
    "    data2ls[[i]] = isosigi\n",
    "}\n",
    "names(data2ls) = names(isosigls.t44)\n",
    "head(data2ls$layer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "bacf71e1-d857-47eb-8f23-18f91a588172",
   "metadata": {},
   "outputs": [],
   "source": [
    "writexl::write_xlsx(data2ls, \"/cluster/home/liuhengxin/P1_brain_nanopore/downstream_analysis/result/figure_revision/data_summary/tables2_26_01_07.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67fb3c55-ee66-46e1-9cb3-1a37c56aeb1b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Table3 Cell type analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "adef7088-9d20-4ac0-b2f0-d61af378a729",
   "metadata": {},
   "outputs": [],
   "source": [
    "datas3 = list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "129c3d00-a704-43ae-b9c4-814007fc75f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'layer', 'region', 'class'. You can\n",
      "override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'layer', 'region', 'class'. You can\n",
      "override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'layer', 'region', 'class'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "cellst.t42 = fsraw.t42 %>% group_by(layer,region,class,subclass) %>% summarise(celln = length(unique(cell_id)))\n",
    "cellst.t44 = fsraw.t44 %>% group_by(layer,region,class,subclass) %>% summarise(celln = length(unique(cell_id)))\n",
    "cellst.t47 = fsraw.t47 %>% group_by(layer,region,class,subclass) %>% summarise(celln = length(unique(cell_label)))\n",
    "cellst.t42$sample = \"Macaque1-2\"\n",
    "cellst.t44$sample = \"Macaque1-1\"\n",
    "cellst.t47$sample = \"Macaque1-3\"\n",
    "datas3$cellst = rbind(cellst.t42,cellst.t44,cellst.t47)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "16ca8746-4e35-4e2f-b137-8b3d8ac7b9f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"L2\"       \"L2_3\"     \"L2_3_4\"   \"L3_4\"     \"L3_4_5\"   \"L4\"      \n",
      " [7] \"L4_5\"     \"L4_5_6\"   \"L5_6\"     \"L6\"       \"LAMP5\"    \"RELN\"    \n",
      "[13] \"VIP_RELN\" \"VIP\"      \"PVALB\"    \"SST\"      \"ASC\"      \"OPC\"     \n",
      "[19] \"OLG\"      \"MG\"      \n",
      "[1] 13\n",
      "[1] 5\n",
      "[1] 11\n",
      "[1] 2\n",
      "[1] 8\n",
      "[1] 4\n",
      "[1] 3\n",
      "[1] 3\n",
      "[1] 9\n",
      "[1] 3\n",
      "[1] 2\n",
      "[1] 13\n",
      "[1] 9\n",
      "[1] 3\n",
      "[1] 9\n",
      "[1] 7\n",
      "[1] 16\n",
      "[1] 10\n",
      "[1] 11\n",
      "[1] 9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 6 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>source</th><th scope=col>term_name</th><th scope=col>term_id</th><th scope=col>highlighted</th><th scope=col>adjusted_p_value</th><th scope=col>negative_log10_of_adjusted_p_value</th><th scope=col>term_size</th><th scope=col>query_size</th><th scope=col>intersection_size</th><th scope=col>effective_domain_size</th><th scope=col>intersections</th><th scope=col>gene_ratio</th><th scope=col>areatr</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>GO:MF</td><td>DNA exonuclease activity, producing 5'-phosphomonoesters</td><td>GO:0016895</td><td>false</td><td>0.013367903</td><td>1.873937</td><td>  17</td><td>173</td><td> 4</td><td>15155</td><td>APTX,EXD2,POLG,TATDN1                                                                                                                                                                                 </td><td>0.23529412</td><td>L2</td></tr>\n",
       "\t<tr><td>GO:MF</td><td>DNA exonuclease activity                                </td><td>GO:0004529</td><td>false</td><td>0.017034582</td><td>1.768669</td><td>  18</td><td>173</td><td> 4</td><td>15155</td><td>APTX,EXD2,POLG,TATDN1                                                                                                                                                                                 </td><td>0.22222222</td><td>L2</td></tr>\n",
       "\t<tr><td>GO:MF</td><td>exonuclease activity                                    </td><td>GO:0004527</td><td>true </td><td>0.005631442</td><td>2.249380</td><td>  70</td><td>173</td><td> 7</td><td>15155</td><td>APTX,CPSF3,EXD2,POLG,RAD1,TATDN1,XRN2                                                                                                                                                                 </td><td>0.10000000</td><td>L2</td></tr>\n",
       "\t<tr><td>HP   </td><td>Neonatal hypotonia                                      </td><td>HP:0001319</td><td>false</td><td>0.026389640</td><td>1.578567</td><td> 180</td><td> 63</td><td>11</td><td> 4759</td><td>AP4B1,BIN1,EGR2,EIF5A,NFIX,NTNG1,PDHX,SLC25A22,SPATA5,SYT1,TPM2                                                                                                                                       </td><td>0.06111111</td><td>L2</td></tr>\n",
       "\t<tr><td>GO:BP</td><td>intracellular transport                                 </td><td>GO:0046907</td><td>false</td><td>0.017172569</td><td>1.765165</td><td> 993</td><td>180</td><td>25</td><td>18644</td><td>AGK,AP3S1,AP4B1,ARFIP2,ATF2,BIN1,BLOC1S6,CCDC91,CMTM6,DYNC1I1,EGR2,GOSR1,IFT20,IFT22,M6PR,OS9,PEX5L,PICK1,PRICKLE1,RABGEF1,RAMP3,REPS2,SARNP,SRSF10,TRMT10B                                           </td><td>0.02517623</td><td>L2</td></tr>\n",
       "\t<tr><td>GO:BP</td><td>establishment of localization in cell                   </td><td>GO:0051649</td><td>true </td><td>0.008733240</td><td>2.058825</td><td>1419</td><td>180</td><td>32</td><td>18644</td><td>AGK,AP3S1,AP4B1,ARFIP2,ATF2,BIN1,BLOC1S6,CACNB4,CCDC91,CEP19,CMTM6,CPLX2,DYNC1I1,EGR2,GOSR1,IFT20,IFT22,M6PR,OS9,PDZD7,PEX5L,PICK1,PRICKLE1,RABGEF1,RACGAP1,RAMP3,REPS2,SARNP,SKA2,SRSF10,SYT1,TRMT10B</td><td>0.02255109</td><td>L2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 6 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " source & term\\_name & term\\_id & highlighted & adjusted\\_p\\_value & negative\\_log10\\_of\\_adjusted\\_p\\_value & term\\_size & query\\_size & intersection\\_size & effective\\_domain\\_size & intersections & gene\\_ratio & areatr\\\\\n",
       " <chr> & <fct> & <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <int> & <int> & <chr> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t GO:MF & DNA exonuclease activity, producing 5'-phosphomonoesters & GO:0016895 & false & 0.013367903 & 1.873937 &   17 & 173 &  4 & 15155 & APTX,EXD2,POLG,TATDN1                                                                                                                                                                                  & 0.23529412 & L2\\\\\n",
       "\t GO:MF & DNA exonuclease activity                                 & GO:0004529 & false & 0.017034582 & 1.768669 &   18 & 173 &  4 & 15155 & APTX,EXD2,POLG,TATDN1                                                                                                                                                                                  & 0.22222222 & L2\\\\\n",
       "\t GO:MF & exonuclease activity                                     & GO:0004527 & true  & 0.005631442 & 2.249380 &   70 & 173 &  7 & 15155 & APTX,CPSF3,EXD2,POLG,RAD1,TATDN1,XRN2                                                                                                                                                                  & 0.10000000 & L2\\\\\n",
       "\t HP    & Neonatal hypotonia                                       & HP:0001319 & false & 0.026389640 & 1.578567 &  180 &  63 & 11 &  4759 & AP4B1,BIN1,EGR2,EIF5A,NFIX,NTNG1,PDHX,SLC25A22,SPATA5,SYT1,TPM2                                                                                                                                        & 0.06111111 & L2\\\\\n",
       "\t GO:BP & intracellular transport                                  & GO:0046907 & false & 0.017172569 & 1.765165 &  993 & 180 & 25 & 18644 & AGK,AP3S1,AP4B1,ARFIP2,ATF2,BIN1,BLOC1S6,CCDC91,CMTM6,DYNC1I1,EGR2,GOSR1,IFT20,IFT22,M6PR,OS9,PEX5L,PICK1,PRICKLE1,RABGEF1,RAMP3,REPS2,SARNP,SRSF10,TRMT10B                                            & 0.02517623 & L2\\\\\n",
       "\t GO:BP & establishment of localization in cell                    & GO:0051649 & true  & 0.008733240 & 2.058825 & 1419 & 180 & 32 & 18644 & AGK,AP3S1,AP4B1,ARFIP2,ATF2,BIN1,BLOC1S6,CACNB4,CCDC91,CEP19,CMTM6,CPLX2,DYNC1I1,EGR2,GOSR1,IFT20,IFT22,M6PR,OS9,PDZD7,PEX5L,PICK1,PRICKLE1,RABGEF1,RACGAP1,RAMP3,REPS2,SARNP,SKA2,SRSF10,SYT1,TRMT10B & 0.02255109 & L2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 6 × 13\n",
       "\n",
       "| source &lt;chr&gt; | term_name &lt;fct&gt; | term_id &lt;chr&gt; | highlighted &lt;chr&gt; | adjusted_p_value &lt;dbl&gt; | negative_log10_of_adjusted_p_value &lt;dbl&gt; | term_size &lt;int&gt; | query_size &lt;int&gt; | intersection_size &lt;int&gt; | effective_domain_size &lt;int&gt; | intersections &lt;chr&gt; | gene_ratio &lt;dbl&gt; | areatr &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| GO:MF | DNA exonuclease activity, producing 5'-phosphomonoesters | GO:0016895 | false | 0.013367903 | 1.873937 |   17 | 173 |  4 | 15155 | APTX,EXD2,POLG,TATDN1                                                                                                                                                                                  | 0.23529412 | L2 |\n",
       "| GO:MF | DNA exonuclease activity                                 | GO:0004529 | false | 0.017034582 | 1.768669 |   18 | 173 |  4 | 15155 | APTX,EXD2,POLG,TATDN1                                                                                                                                                                                  | 0.22222222 | L2 |\n",
       "| GO:MF | exonuclease activity                                     | GO:0004527 | true  | 0.005631442 | 2.249380 |   70 | 173 |  7 | 15155 | APTX,CPSF3,EXD2,POLG,RAD1,TATDN1,XRN2                                                                                                                                                                  | 0.10000000 | L2 |\n",
       "| HP    | Neonatal hypotonia                                       | HP:0001319 | false | 0.026389640 | 1.578567 |  180 |  63 | 11 |  4759 | AP4B1,BIN1,EGR2,EIF5A,NFIX,NTNG1,PDHX,SLC25A22,SPATA5,SYT1,TPM2                                                                                                                                        | 0.06111111 | L2 |\n",
       "| GO:BP | intracellular transport                                  | GO:0046907 | false | 0.017172569 | 1.765165 |  993 | 180 | 25 | 18644 | AGK,AP3S1,AP4B1,ARFIP2,ATF2,BIN1,BLOC1S6,CCDC91,CMTM6,DYNC1I1,EGR2,GOSR1,IFT20,IFT22,M6PR,OS9,PEX5L,PICK1,PRICKLE1,RABGEF1,RAMP3,REPS2,SARNP,SRSF10,TRMT10B                                            | 0.02517623 | L2 |\n",
       "| GO:BP | establishment of localization in cell                    | GO:0051649 | true  | 0.008733240 | 2.058825 | 1419 | 180 | 32 | 18644 | AGK,AP3S1,AP4B1,ARFIP2,ATF2,BIN1,BLOC1S6,CACNB4,CCDC91,CEP19,CMTM6,CPLX2,DYNC1I1,EGR2,GOSR1,IFT20,IFT22,M6PR,OS9,PDZD7,PEX5L,PICK1,PRICKLE1,RABGEF1,RACGAP1,RAMP3,REPS2,SARNP,SKA2,SRSF10,SYT1,TRMT10B | 0.02255109 | L2 |\n",
       "\n"
      ],
      "text/plain": [
       "  source term_name                                                term_id   \n",
       "1 GO:MF  DNA exonuclease activity, producing 5'-phosphomonoesters GO:0016895\n",
       "2 GO:MF  DNA exonuclease activity                                 GO:0004529\n",
       "3 GO:MF  exonuclease activity                                     GO:0004527\n",
       "4 HP     Neonatal hypotonia                                       HP:0001319\n",
       "5 GO:BP  intracellular transport                                  GO:0046907\n",
       "6 GO:BP  establishment of localization in cell                    GO:0051649\n",
       "  highlighted adjusted_p_value negative_log10_of_adjusted_p_value term_size\n",
       "1 false       0.013367903      1.873937                             17     \n",
       "2 false       0.017034582      1.768669                             18     \n",
       "3 true        0.005631442      2.249380                             70     \n",
       "4 false       0.026389640      1.578567                            180     \n",
       "5 false       0.017172569      1.765165                            993     \n",
       "6 true        0.008733240      2.058825                           1419     \n",
       "  query_size intersection_size effective_domain_size\n",
       "1 173         4                15155                \n",
       "2 173         4                15155                \n",
       "3 173         7                15155                \n",
       "4  63        11                 4759                \n",
       "5 180        25                18644                \n",
       "6 180        32                18644                \n",
       "  intersections                                                                                                                                                                                         \n",
       "1 APTX,EXD2,POLG,TATDN1                                                                                                                                                                                 \n",
       "2 APTX,EXD2,POLG,TATDN1                                                                                                                                                                                 \n",
       "3 APTX,CPSF3,EXD2,POLG,RAD1,TATDN1,XRN2                                                                                                                                                                 \n",
       "4 AP4B1,BIN1,EGR2,EIF5A,NFIX,NTNG1,PDHX,SLC25A22,SPATA5,SYT1,TPM2                                                                                                                                       \n",
       "5 AGK,AP3S1,AP4B1,ARFIP2,ATF2,BIN1,BLOC1S6,CCDC91,CMTM6,DYNC1I1,EGR2,GOSR1,IFT20,IFT22,M6PR,OS9,PEX5L,PICK1,PRICKLE1,RABGEF1,RAMP3,REPS2,SARNP,SRSF10,TRMT10B                                           \n",
       "6 AGK,AP3S1,AP4B1,ARFIP2,ATF2,BIN1,BLOC1S6,CACNB4,CCDC91,CEP19,CMTM6,CPLX2,DYNC1I1,EGR2,GOSR1,IFT20,IFT22,M6PR,OS9,PDZD7,PEX5L,PICK1,PRICKLE1,RABGEF1,RACGAP1,RAMP3,REPS2,SARNP,SKA2,SRSF10,SYT1,TRMT10B\n",
       "  gene_ratio areatr\n",
       "1 0.23529412 L2    \n",
       "2 0.22222222 L2    \n",
       "3 0.10000000 L2    \n",
       "4 0.06111111 L2    \n",
       "5 0.02517623 L2    \n",
       "6 0.02255109 L2    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(readxl)\n",
    "file_path <- \"result/figure_revision/Hubgene_by_subclass_25_11_28.xlsx\"\n",
    "sheet_names <- excel_sheets(file_path)\n",
    "print(sheet_names)\n",
    "gofls = list.files(\"result/figure_revision/GO_analysis/\",full.names = T)\n",
    "gofls = gofls[order(gofls)]\n",
    "\n",
    "MyGOFilter = function(godfi,ni){\n",
    "    godfi$gene_ratio = godfi$intersection_size/godfi$term_size\n",
    "    # godfi = godfi[godfi$intersection_size > 2,]\n",
    "    # godfi = godfi[grepl(godfi$term_name,pattern = \"synap|splic|dendrite\") | godfi$source %in% c(\"HP\",\"KEGG\"),]\n",
    "    godfi = godfi %>% group_by(source) %>% arrange(intersection_size) %>% \n",
    "    top_n(wt = intersection_size,n = ni)\n",
    "    godfi$term_name = factor(godfi$term_name,levels = unique(godfi$term_name))\n",
    "    print(nrow(godfi))\n",
    "    godfi = godfi[!godfi$term_name %in% c(\"cell junction\",\"cytoplasm\",\"cytosol\",\"cytoskeleton\"),]\n",
    "    return(godfi)\n",
    "}\n",
    "\n",
    "godf = NULL\n",
    "for(i in 1:length(gofls)){\n",
    "    godfi = read.csv(gofls[i])\n",
    "    godfi = tryCatch(MyGOFilter(godfi,5),error = function(e) 0)\n",
    "    if(!is.null(nrow(godfi))){\n",
    "        if(nrow(godfi) > 0){\n",
    "            godfi$areatr = sheet_names[i]\n",
    "            godf = rbind(godf,godfi)\n",
    "        }\n",
    "    }\n",
    "}\n",
    "head(godf)\n",
    "datas3$godf = godf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "79022354-11eb-496d-863a-f881b9663327",
   "metadata": {},
   "outputs": [],
   "source": [
    "isosig.b62 = qread(file = \"data/fullscope_qc/Mouse_ST_B62_fullscope_DTU_sc26_01_03.qs\")\n",
    "isosigref.ls = qread(\"data/fullscope_qc/Mouse_SC_tilgner_fullscope_witnan_26_01_03.qs\")\n",
    "isosigref.cl = isosigref.ls$class\n",
    "isosigi1 = isosig.b62\n",
    "isosigi2 = isosigref.cl\n",
    "renamedf = data.frame(\"areatr\" = c(\"InhibNeuron\",\"ExciteNeuron\",\"NN\"),new = c(\"GABA\",\"GLU\",\"NN\"))\n",
    "isosigi2 = merge(isosigi2,renamedf,by = \"areatr\")\n",
    "isosigi2$areatr = isosigi2$new\n",
    "# isosigi1 = isosigi1[isosigi1$count > 50,]\n",
    "# isosigi2 = isosigi2[isosigi2$count > 50,]\n",
    "hubtrans = unique(c(isosigi1[isosigi1$sig != \"N\",]$transcript_id,\n",
    "                   isosigi2[isosigi2$sig != \"N\",]$transcript_id))\n",
    "isosigi1 = isosigi1[isosigi1$transcript_id %in% hubtrans & isosigi1$count > 50,]\n",
    "isosigi2 = isosigi2[isosigi2$transcript_id %in% hubtrans & isosigi2$count > 50,]\n",
    "isosig.cmp2 = merge(isosigi1[,c(\"gene_id\",\"transcript_id\",\"areatr\",\"count\",\"dr\")],\n",
    "                   isosigi2[,c(\"gene_id\",\"transcript_id\",\"areatr\",\"count\",\"dr\")],\n",
    "                   by = c(\"gene_id\",\"transcript_id\",\"areatr\"),suffixes = c(\".Fullscope\",\".Tilgner_sc\"))\n",
    "datas3$mouse_cmp = isosig.cmp2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "1f1bbede-bcfb-435d-838f-3c5a73c1510a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NULL"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(datas3$cellanst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "4b011463-844e-48e4-ac28-7ce4df130391",
   "metadata": {},
   "outputs": [],
   "source": [
    "writexl::write_xlsx(datas3, \"/cluster/home/liuhengxin/P1_brain_nanopore/downstream_analysis/result/figure_revision/data_summary/tables3_26_01_07.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f3de9c0-9c08-4554-a7da-12bb89698da7",
   "metadata": {},
   "source": [
    "# Table4 GLM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "03c8b4a5-e896-49b7-95c1-3bc9a5279832",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Estimate</th><th scope=col>Std. Error</th><th scope=col>z value</th><th scope=col>pvalue</th><th scope=col>gene_id</th><th scope=col>areatr</th><th scope=col>group</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>(Intercept)</th><td>-59.727097</td><td>2366015078</td><td>-2.524375e-08</td><td>1</td><td>ABI2</td><td>(Intercept)</td><td>(Intercept)</td></tr>\n",
       "\t<tr><th scope=row>cellm</th><td> 53.450647</td><td>1992326184</td><td> 2.682826e-08</td><td>1</td><td>ABI2</td><td>cellm      </td><td>cell       </td></tr>\n",
       "\t<tr><th scope=row>regionm</th><td> 53.214197</td><td>1100514613</td><td> 4.835392e-08</td><td>1</td><td>ABI2</td><td>regionm    </td><td>region     </td></tr>\n",
       "\t<tr><th scope=row>(Intercept)1</th><td> -5.806286</td><td> 771268832</td><td>-7.528226e-09</td><td>1</td><td>ADD1</td><td>(Intercept)</td><td>(Intercept)</td></tr>\n",
       "\t<tr><th scope=row>cellm1</th><td> 30.355730</td><td> 567226468</td><td> 5.351607e-08</td><td>1</td><td>ADD1</td><td>cellm      </td><td>cell       </td></tr>\n",
       "\t<tr><th scope=row>regionm1</th><td> 21.291255</td><td> 548798317</td><td> 3.879614e-08</td><td>1</td><td>ADD1</td><td>regionm    </td><td>region     </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & Estimate & Std. Error & z value & pvalue & gene\\_id & areatr & group\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t(Intercept) & -59.727097 & 2366015078 & -2.524375e-08 & 1 & ABI2 & (Intercept) & (Intercept)\\\\\n",
       "\tcellm &  53.450647 & 1992326184 &  2.682826e-08 & 1 & ABI2 & cellm       & cell       \\\\\n",
       "\tregionm &  53.214197 & 1100514613 &  4.835392e-08 & 1 & ABI2 & regionm     & region     \\\\\n",
       "\t(Intercept)1 &  -5.806286 &  771268832 & -7.528226e-09 & 1 & ADD1 & (Intercept) & (Intercept)\\\\\n",
       "\tcellm1 &  30.355730 &  567226468 &  5.351607e-08 & 1 & ADD1 & cellm       & cell       \\\\\n",
       "\tregionm1 &  21.291255 &  548798317 &  3.879614e-08 & 1 & ADD1 & regionm     & region     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | Estimate &lt;dbl&gt; | Std. Error &lt;dbl&gt; | z value &lt;dbl&gt; | pvalue &lt;dbl&gt; | gene_id &lt;chr&gt; | areatr &lt;chr&gt; | group &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| (Intercept) | -59.727097 | 2366015078 | -2.524375e-08 | 1 | ABI2 | (Intercept) | (Intercept) |\n",
       "| cellm |  53.450647 | 1992326184 |  2.682826e-08 | 1 | ABI2 | cellm       | cell        |\n",
       "| regionm |  53.214197 | 1100514613 |  4.835392e-08 | 1 | ABI2 | regionm     | region      |\n",
       "| (Intercept)1 |  -5.806286 |  771268832 | -7.528226e-09 | 1 | ADD1 | (Intercept) | (Intercept) |\n",
       "| cellm1 |  30.355730 |  567226468 |  5.351607e-08 | 1 | ADD1 | cellm       | cell        |\n",
       "| regionm1 |  21.291255 |  548798317 |  3.879614e-08 | 1 | ADD1 | regionm     | region      |\n",
       "\n"
      ],
      "text/plain": [
       "             Estimate   Std. Error z value       pvalue gene_id areatr     \n",
       "(Intercept)  -59.727097 2366015078 -2.524375e-08 1      ABI2    (Intercept)\n",
       "cellm         53.450647 1992326184  2.682826e-08 1      ABI2    cellm      \n",
       "regionm       53.214197 1100514613  4.835392e-08 1      ABI2    regionm    \n",
       "(Intercept)1  -5.806286  771268832 -7.528226e-09 1      ADD1    (Intercept)\n",
       "cellm1        30.355730  567226468  5.351607e-08 1      ADD1    cellm      \n",
       "regionm1      21.291255  548798317  3.879614e-08 1      ADD1    regionm    \n",
       "             group      \n",
       "(Intercept)  (Intercept)\n",
       "cellm        cell       \n",
       "regionm      region     \n",
       "(Intercept)1 (Intercept)\n",
       "cellm1       cell       \n",
       "regionm1     region     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "354"
      ],
      "text/latex": [
       "354"
      ],
      "text/markdown": [
       "354"
      ],
      "text/plain": [
       "[1] 354"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "datas4 = list()\n",
    "modelcoef2 = qread(\"data/dtu_analysis/region_subclass_isoform_model_tri_onlyregion_25_12_03.qs\")\n",
    "\n",
    "modelcoef2$group = modelcoef2$areatr\n",
    "modelcoef2[modelcoef2$areatr == \"cellm\",]$group = \"cell\"\n",
    "\n",
    "modelcoef2[modelcoef2$areatr == \"regionm\",]$group = \"region\"\n",
    "head(modelcoef2)\n",
    "\n",
    "colnames(modelcoef2)[4] = \"pvalue\"\n",
    "modelcoef.hub = modelcoef2[modelcoef2$group %in% c(\"cell\",\"region\"),]\n",
    "length(unique(modelcoef.hub$gene_id))\n",
    "modelcoef.hub$pvalue = -log10(modelcoef.hub$pvalue)\n",
    "modelmx1 = dcast(modelcoef.hub,gene_id~group,value.var = \"pvalue\",fun.aggregate = sum)\n",
    "modelmx1.hub = modelmx1\n",
    "modelmx1.hub$sig <- sapply(1:nrow(modelmx1.hub), function(i) {\n",
    "  significant_cols <- names(modelmx1.hub)[-1][modelmx1.hub[i, -1] > -log10(0.05)]\n",
    "  paste(significant_cols, collapse = \"-\")\n",
    "})\n",
    "datas4$region_model = modelmx1.hub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "301d20d3-57fa-4ea0-a252-523ee9db58a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "writexl::write_xlsx(datas4, \"/cluster/home/liuhengxin/P1_brain_nanopore/downstream_analysis/result/figure_revision/data_summary/tables4_26_01_07.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1e640a1-df6d-41d7-b2b8-dc8669e4d0c6",
   "metadata": {},
   "source": [
    "# Table5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "cfe36e6a-8721-45cc-9805-2726b3ed2c41",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: 'tidyr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:reshape2':\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'purrr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:XVector':\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following object is masked from 'package:data.table':\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene_id</th><th scope=col>transcript_id</th><th scope=col>areatr</th><th scope=col>count</th><th scope=col>genecount</th><th scope=col>relative</th><th scope=col>meanrelative</th><th scope=col>otherrelative</th><th scope=col>dr</th><th scope=col>sig</th><th scope=col>pvalue</th><th scope=col>padj</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>9</th><td>AACS</td><td>NM_001284010.1</td><td>Cerebellum </td><td> 146</td><td> 158</td><td>0.9240506</td><td>0.855632</td><td>0.8419483</td><td> 0.08210238</td><td>N   </td><td>0.001852607</td><td>0.005294328</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>AACS</td><td>NM_001284010.1</td><td>Cortex     </td><td>2879</td><td>3272</td><td>0.8798900</td><td>0.855632</td><td>0.8507804</td><td> 0.02910959</td><td>N   </td><td>0.001852607</td><td>0.005294328</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>AACS</td><td>NM_001284010.1</td><td>Hippocampus</td><td>  92</td><td> 117</td><td>0.7863248</td><td>0.855632</td><td>0.8694934</td><td>-0.08316864</td><td>N   </td><td>0.001852607</td><td>0.005294328</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>AACS</td><td>NM_001284010.1</td><td>Midbrain   </td><td>  37</td><td>  48</td><td>0.7708333</td><td>0.855632</td><td>0.8725917</td><td>-0.10175838</td><td>Down</td><td>0.001852607</td><td>0.005294328</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>AACS</td><td>NM_001284010.1</td><td>Striatum   </td><td>  29</td><td>  31</td><td>0.9354839</td><td>0.855632</td><td>0.8396616</td><td> 0.09582226</td><td>N   </td><td>0.001852607</td><td>0.005294328</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>AACS</td><td>NM_001284010.1</td><td>Ventricle  </td><td>  36</td><td>  43</td><td>0.8372093</td><td>0.855632</td><td>0.8593165</td><td>-0.02210722</td><td>N   </td><td>0.001852607</td><td>0.005294328</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 12\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & gene\\_id & transcript\\_id & areatr & count & genecount & relative & meanrelative & otherrelative & dr & sig & pvalue & padj\\\\\n",
       "  & <chr> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t9 & AACS & NM\\_001284010.1 & Cerebellum  &  146 &  158 & 0.9240506 & 0.855632 & 0.8419483 &  0.08210238 & N    & 0.001852607 & 0.005294328\\\\\n",
       "\t10 & AACS & NM\\_001284010.1 & Cortex      & 2879 & 3272 & 0.8798900 & 0.855632 & 0.8507804 &  0.02910959 & N    & 0.001852607 & 0.005294328\\\\\n",
       "\t11 & AACS & NM\\_001284010.1 & Hippocampus &   92 &  117 & 0.7863248 & 0.855632 & 0.8694934 & -0.08316864 & N    & 0.001852607 & 0.005294328\\\\\n",
       "\t12 & AACS & NM\\_001284010.1 & Midbrain    &   37 &   48 & 0.7708333 & 0.855632 & 0.8725917 & -0.10175838 & Down & 0.001852607 & 0.005294328\\\\\n",
       "\t13 & AACS & NM\\_001284010.1 & Striatum    &   29 &   31 & 0.9354839 & 0.855632 & 0.8396616 &  0.09582226 & N    & 0.001852607 & 0.005294328\\\\\n",
       "\t14 & AACS & NM\\_001284010.1 & Ventricle   &   36 &   43 & 0.8372093 & 0.855632 & 0.8593165 & -0.02210722 & N    & 0.001852607 & 0.005294328\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 12\n",
       "\n",
       "| <!--/--> | gene_id &lt;chr&gt; | transcript_id &lt;chr&gt; | areatr &lt;chr&gt; | count &lt;int&gt; | genecount &lt;int&gt; | relative &lt;dbl&gt; | meanrelative &lt;dbl&gt; | otherrelative &lt;dbl&gt; | dr &lt;dbl&gt; | sig &lt;chr&gt; | pvalue &lt;dbl&gt; | padj &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 9 | AACS | NM_001284010.1 | Cerebellum  |  146 |  158 | 0.9240506 | 0.855632 | 0.8419483 |  0.08210238 | N    | 0.001852607 | 0.005294328 |\n",
       "| 10 | AACS | NM_001284010.1 | Cortex      | 2879 | 3272 | 0.8798900 | 0.855632 | 0.8507804 |  0.02910959 | N    | 0.001852607 | 0.005294328 |\n",
       "| 11 | AACS | NM_001284010.1 | Hippocampus |   92 |  117 | 0.7863248 | 0.855632 | 0.8694934 | -0.08316864 | N    | 0.001852607 | 0.005294328 |\n",
       "| 12 | AACS | NM_001284010.1 | Midbrain    |   37 |   48 | 0.7708333 | 0.855632 | 0.8725917 | -0.10175838 | Down | 0.001852607 | 0.005294328 |\n",
       "| 13 | AACS | NM_001284010.1 | Striatum    |   29 |   31 | 0.9354839 | 0.855632 | 0.8396616 |  0.09582226 | N    | 0.001852607 | 0.005294328 |\n",
       "| 14 | AACS | NM_001284010.1 | Ventricle   |   36 |   43 | 0.8372093 | 0.855632 | 0.8593165 | -0.02210722 | N    | 0.001852607 | 0.005294328 |\n",
       "\n"
      ],
      "text/plain": [
       "   gene_id transcript_id  areatr      count genecount relative  meanrelative\n",
       "9  AACS    NM_001284010.1 Cerebellum   146   158      0.9240506 0.855632    \n",
       "10 AACS    NM_001284010.1 Cortex      2879  3272      0.8798900 0.855632    \n",
       "11 AACS    NM_001284010.1 Hippocampus   92   117      0.7863248 0.855632    \n",
       "12 AACS    NM_001284010.1 Midbrain      37    48      0.7708333 0.855632    \n",
       "13 AACS    NM_001284010.1 Striatum      29    31      0.9354839 0.855632    \n",
       "14 AACS    NM_001284010.1 Ventricle     36    43      0.8372093 0.855632    \n",
       "   otherrelative dr          sig  pvalue      padj       \n",
       "9  0.8419483      0.08210238 N    0.001852607 0.005294328\n",
       "10 0.8507804      0.02910959 N    0.001852607 0.005294328\n",
       "11 0.8694934     -0.08316864 N    0.001852607 0.005294328\n",
       "12 0.8725917     -0.10175838 Down 0.001852607 0.005294328\n",
       "13 0.8396616      0.09582226 N    0.001852607 0.005294328\n",
       "14 0.8593165     -0.02210722 N    0.001852607 0.005294328"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "375"
      ],
      "text/latex": [
       "375"
      ],
      "text/markdown": [
       "375"
      ],
      "text/plain": [
       "[1] 375"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "375"
      ],
      "text/latex": [
       "375"
      ],
      "text/markdown": [
       "375"
      ],
      "text/plain": [
       "[1] 375"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(purrr)\n",
    "# major isoform shift stat\n",
    "isosigi = isosigls.t44$subcortical\n",
    "hubgene = unique(isosigi[isosigi$sig != \"N\",]$gene_id)\n",
    "isosigi = isosigi[isosigi$gene_id %in% hubgene,]\n",
    "head(isosigi)\n",
    "#construct phase dataset\n",
    "isosigi$ismajor <- FALSE\n",
    "isosigi <- isosigi %>%\n",
    "  group_by(areatr, gene_id) %>%\n",
    "  mutate(ismajor = ifelse(relative == max(relative), TRUE, FALSE)) %>%\n",
    "  ungroup()\n",
    "isosigi = isosigi %>% group_by(gene_id) %>% filter(meanrelative == max(meanrelative))\n",
    "#isosigi[isosigi$relative > 0.5,]$ismajor = TRUE\n",
    "isosigi = isosigi[isosigi$ismajor == TRUE | (isosigi$ismajor == FALSE & abs(isosigi$relative - isosigi$otherrelative) > 0.1),]\n",
    "isosigi = isosigi %>% group_by(transcript_id) %>% filter(length(unique(ismajor)) > 1)\n",
    "length(unique(isosigi$gene_id))\n",
    "length(unique(isosigi$transcript_id))\n",
    "library(tidyr)\n",
    "df_wide <- pivot_wider(isosigi, \n",
    "                       id_cols = transcript_id, \n",
    "                       names_from = areatr, \n",
    "                       values_from = ismajor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "ee3299be-5eb4-4c16-8be0-8e40e73a318f",
   "metadata": {},
   "outputs": [],
   "source": [
    "datas5 = list()\n",
    "datas5$shift_isoform = isosigi\n",
    "datas5$shift_isoform_st = df_wide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "28d4978b-aeb3-40a1-b722-4cdf81cc397c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4515"
      ],
      "text/latex": [
       "4515"
      ],
      "text/markdown": [
       "4515"
      ],
      "text/plain": [
       "[1] 4515"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "transt = qread(file = \"data/cross_region/transt_subcortical_cell_level_25_10_28.qs\")\n",
    "transt = transt[transt$transcript_id %in% isosigi$transcript_id,]\n",
    "nrow(transt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "6fecb21a-2fe1-486d-898b-9bbec26d1e6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "datas5$celldtu = transt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "afc197a2-8347-4bc9-b8f4-29d4c1639a4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "writexl::write_xlsx(datas5, \"/cluster/home/liuhengxin/P1_brain_nanopore/downstream_analysis/result/figure_revision/data_summary/tables5_26_01_07.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b7eb56f-8a2a-4801-8d2e-5859aab48a52",
   "metadata": {},
   "source": [
    "# Table 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "42fd755f-0029-45f9-98eb-740b8fb01bea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene_id</th><th scope=col>source</th><th scope=col>group</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>MED13  </td><td>ASD.Talkowski</td><td>ASD</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>SYNGAP1</td><td>ASD.Talkowski</td><td>ASD</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>TRIP12 </td><td>ASD.Talkowski</td><td>ASD</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>GRIN2B </td><td>ASD.Talkowski</td><td>ASD</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ASH1L  </td><td>ASD.Talkowski</td><td>ASD</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>MED13L </td><td>ASD.Talkowski</td><td>ASD</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & gene\\_id & source & group\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & MED13   & ASD.Talkowski & ASD\\\\\n",
       "\t2 & SYNGAP1 & ASD.Talkowski & ASD\\\\\n",
       "\t3 & TRIP12  & ASD.Talkowski & ASD\\\\\n",
       "\t4 & GRIN2B  & ASD.Talkowski & ASD\\\\\n",
       "\t5 & ASH1L   & ASD.Talkowski & ASD\\\\\n",
       "\t6 & MED13L  & ASD.Talkowski & ASD\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | gene_id &lt;chr&gt; | source &lt;chr&gt; | group &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | MED13   | ASD.Talkowski | ASD |\n",
       "| 2 | SYNGAP1 | ASD.Talkowski | ASD |\n",
       "| 3 | TRIP12  | ASD.Talkowski | ASD |\n",
       "| 4 | GRIN2B  | ASD.Talkowski | ASD |\n",
       "| 5 | ASH1L   | ASD.Talkowski | ASD |\n",
       "| 6 | MED13L  | ASD.Talkowski | ASD |\n",
       "\n"
      ],
      "text/plain": [
       "  gene_id source        group\n",
       "1 MED13   ASD.Talkowski ASD  \n",
       "2 SYNGAP1 ASD.Talkowski ASD  \n",
       "3 TRIP12  ASD.Talkowski ASD  \n",
       "4 GRIN2B  ASD.Talkowski ASD  \n",
       "5 ASH1L   ASD.Talkowski ASD  \n",
       "6 MED13L  ASD.Talkowski ASD  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "refgdf = read.csv(\"ref/disease_gene/Total_gene_list_summary_25_03_19.csv\")\n",
    "head(refgdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "78a73535-ba8d-4d09-b706-648350a7d044",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "11982"
      ],
      "text/latex": [
       "11982"
      ],
      "text/markdown": [
       "11982"
      ],
      "text/plain": [
       "[1] 11982"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "isosigls = isosigls.t44\n",
    "# gtf.hp = rtracklayer::import(\"ref/cross_species/gencode.v19.annotation.gtf\")\n",
    "# gtf = rtracklayer::import(\"../ref/bambu_extended_annotations_novelname.gtf\")\n",
    "cmgeneid = intersect(gtf.hp$gene_name,gtf$gene_id)\n",
    "cmgeneid = cmgeneid[cmgeneid %in% fsraw.t44$gene_id]\n",
    "cmgenen = length(cmgeneid)\n",
    "cmgenen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "30e5f474-6e24-4adf-85d0-98cbfe22e0c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "hubsource = c('ASD.DTU','BD.DTU','SCZ.DTU','ASD.GWAS','BD.GWAS','SCZ.GWAS','NDD.GWAS','DDD.GWAS',\n",
    "              'BIP.GWAS','EPI.GWAS','ASD.TWAS','BD.TWAS',\"SCZ.TWAS\")\n",
    "FishEnrich = function(geneset,funcset,total_genes){\n",
    "    drugovl = intersect(unique(funcset),geneset)\n",
    "    geneset_size <- length(geneset)    # 感兴趣的基因集大小\n",
    "    functional_category_size <- length(unique(funcset))  # 特定功能类别中的基因总数\n",
    "    both <- length(drugovl)  # 两个集合都包含的基因数量\n",
    "    #total_genes <- 20000  # 整个基因组中的基因总数\n",
    "\n",
    "    contingency_table <- matrix(c(both, geneset_size - both, \n",
    "                              functional_category_size - both, \n",
    "                              total_genes - (geneset_size + functional_category_size - both)), \n",
    "                            nrow = 2)\n",
    "\n",
    "    # 进行Fisher精确检验\n",
    "    fisher_test_result <- fisher.test(contingency_table)\n",
    "    # 输出p值\n",
    "    p_value <- fisher_test_result$p.value\n",
    "\n",
    "    # 计算富集分数（Enrichment Ratio），这通常是感兴趣的基因集中具有特定功能的基因比例与整个基因组中该功能基因比例的比值\n",
    "    enrichment_ratio <- (both / geneset_size) / (functional_category_size / total_genes)\n",
    "    res = data.frame(\"p.value\" = p_value,\"enrichscore\" = enrichment_ratio,\"overlap_size\" = both,\n",
    "                     \"term_size\" = functional_category_size,\"query_size\" = geneset_size,\n",
    "                    \"overlap_ratio\" = both/functional_category_size)\n",
    "    return(res)\n",
    "}\n",
    "\n",
    "DiseaseEnrich = function(isosigt,refgdf){\n",
    "    isosigt = isosigt[substr(isosigt$gene_id,1,9) != \"BambuGene\",]\n",
    "    areatrs = unique(isosigt$areatr)\n",
    "    goana = NULL\n",
    "    refgdfi = refgdf\n",
    "    sourceid = unique(refgdfi$source)\n",
    "    for(i in 1:length(areatrs)){\n",
    "        isosigii = isosigt[isosigt$areatr == areatrs[i],]\n",
    "        hubgene = unique(isosigii$gene_id)\n",
    "        for(j in 1:length(sourceid)){\n",
    "            refgene = unique(refgdfi[refgdfi$source == sourceid[j],]$gene_id)\n",
    "            goanai = FishEnrich(hubgene,refgene,cmgenen)\n",
    "            goanai$group = sourceid[j]\n",
    "            goanai$areatr = areatrs[i]\n",
    "            goana = rbind(goana,goanai)\n",
    "        }\n",
    "    }\n",
    "    goana$fdr = p.adjust(goana$p.value)\n",
    "    goanaf = goana[goana$group %in% hubsource,]\n",
    "    #goanaf\n",
    "    #goanaf$group = MySplit(goanaf$group,\"\\\\.\",1)\n",
    "    goanaf$group = factor(goanaf$group,levels = hubsource)\n",
    "    goanaf$stars <- cut(goanaf$p.value, breaks=c(-Inf, 0.001, 0.01, 0.05, Inf), label=c(\"***\", \"**\", \"*\", \"\"))  # Create column of significance labels\n",
    "    return(goanaf)\n",
    "}\n",
    "#subclass\n",
    "isosigt = isosigls$subclass\n",
    "isosigt = isosigt[isosigt$sig != \"N\",]\n",
    "isosigt = isosigt %>% group_by(transcript_id,sig) %>% filter(length(unique(areatr)) == 1)\n",
    "isosigt = isosigt[isosigt$gene_id %in% cmgeneid,]\n",
    "goanaf.subuni = DiseaseEnrich(isosigt,refgdf)\n",
    "goanaf.subuni$sample = \"subclass\"\n",
    "#class\n",
    "isosigt = isosigls$class\n",
    "isosigt = isosigt[isosigt$sig != \"N\",]\n",
    "isosigt = isosigt %>% group_by(transcript_id,sig) %>% filter(length(unique(areatr)) == 1)\n",
    "goanaf.class = DiseaseEnrich(isosigt,refgdf)\n",
    "goanaf.class$sample = \"class\"\n",
    "\n",
    "#layer\n",
    "isosigt = isosigls$layer\n",
    "isosigt = isosigt[isosigt$sig != \"N\",]\n",
    "isosigt = isosigt %>% group_by(transcript_id,sig) %>% filter(length(unique(areatr)) == 1)\n",
    "goanaf.layer = DiseaseEnrich(isosigt,refgdf)\n",
    "goanaf.layer$sample = \"layer\"\n",
    "\n",
    "#lobe\n",
    "isosigt = isosigls$lobe\n",
    "isosigt = isosigt[isosigt$sig != \"N\",]\n",
    "isosigt = isosigt %>% group_by(transcript_id,sig) %>% filter(length(unique(areatr)) == 1)\n",
    "isosigt = isosigt[isosigt$gene_id %in% cmgeneid,]\n",
    "goanaf.lobe = DiseaseEnrich(isosigt,refgdf)\n",
    "goanaf.lobe$sample = \"lobe\"\n",
    "\n",
    "#subcortical\n",
    "isosigt = isosigls$subcortical\n",
    "isosigt = isosigt[isosigt$sig != \"N\",]\n",
    "isosigt = isosigt %>% group_by(transcript_id,sig) %>% filter(length(unique(areatr)) == 1)\n",
    "isosigt = isosigt[isosigt$gene_id %in% cmgeneid,]\n",
    "goanaf.subc = DiseaseEnrich(isosigt,refgdf)\n",
    "goanaf.subc$sample = \"subcortical\"\n",
    "\n",
    "goanaft = rbind(goanaf.subuni,goanaf.class,goanaf.layer,goanaf.lobe,goanaf.subc)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "7ba4297d-232f-457a-9672-44acf613a4c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "20837"
      ],
      "text/latex": [
       "20837"
      ],
      "text/markdown": [
       "20837"
      ],
      "text/plain": [
       "[1] 20837"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.221590063168641"
      ],
      "text/latex": [
       "0.221590063168641"
      ],
      "text/markdown": [
       "0.221590063168641"
      ],
      "text/plain": [
       "[1] 0.2215901"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtf.fil = rtracklayer::import(\"ref/cross_species/bambu_extended_annotations_novelname_fil_04_30.gtf\")\n",
    "hgvmac.tmap = read.table(\"ref/cross_species/bambu_macaca_vs_hg19lifted_25_03_16.gtf.bambu_extended_annotations_novelname_fil_04_30.gtf.tmap\",\n",
    "                        header = T)\n",
    "hgvmac.tmapf = hgvmac.tmap[hgvmac.tmap$class_code == \"=\",]\n",
    "hgvmac.tmapf$ref_id = substr(hgvmac.tmapf$ref_id,1,nchar(hgvmac.tmapf$ref_id)-2)\n",
    "trnumt = length(unique(gtf.fil$transcript_id))\n",
    "trnumhit = length(unique(hgvmac.tmapf$qry_id))\n",
    "trnumhit\n",
    "trnumhit/trnumt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "e543a686-42f7-40bb-a9f6-a6b888461bd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data6s = list()\n",
    "data6s$ref = refgdf\n",
    "data6s$enrich = goanaft\n",
    "data6s$cross_species = hgvmac.tmapf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "c07709dc-46e1-48db-bae4-94fe4fb54e2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "discor = read_xlsx(\"result/figure6_snp_analysis//Disease_analysis_of_top_DTUs_by_disgener_25_12_24.xlsx\")\n",
    "data6s$Diseas_score = discor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "6baa1f69-0b04-4d11-83e6-f4ea564701a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "writexl::write_xlsx(data6s, \"/cluster/home/liuhengxin/P1_brain_nanopore/downstream_analysis/result/figure_revision/data_summary/tables6_26_01_07.xlsx\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
