{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8c986c7-2015-483c-9ea3-ae58e7fc3cf7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "qs 0.25.5\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Warning message:\n",
      "“package ‘IRanges’ was built under R version 4.3.3”\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Warning message:\n",
      "“package ‘AnnotationDbi’ was built under R version 4.3.2”\n",
      "Loading required package: Biobase\n",
      "\n",
      "Warning message:\n",
      "“package ‘Biobase’ was built under R version 4.3.3”\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n",
      "Need specific help about ggbio? try mailing \n",
      " the maintainer or visit https://lawremi.github.io/ggbio/\n",
      "\n",
      "\n",
      "Attaching package: 'ggbio'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ggplot2':\n",
      "\n",
      "    geom_bar, geom_rect, geom_segment, ggsave, stat_bin, stat_identity,\n",
      "    xlim\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'reshape2'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'ggsci' was built under R version 4.3.3\"\n",
      "Warning message:\n",
      "\"package 'viridis' was built under R version 4.3.3\"\n",
      "Loading required package: viridisLite\n",
      "\n",
      "Warning message:\n",
      "\"package 'circlize' was built under R version 4.3.3\"\n",
      "========================================\n",
      "circlize version 0.4.16\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Warning message:\n",
      "\"package 'MatrixGenerics' was built under R version 4.3.3\"\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: 'matrixStats'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Biobase':\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'MatrixGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Warning message:\n",
      "\"package 'Biostrings' was built under R version 4.3.3\"\n",
      "Loading required package: XVector\n",
      "\n",
      "Warning message:\n",
      "\"package 'XVector' was built under R version 4.3.3\"\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: BiocIO\n",
      "\n",
      "\n",
      "Attaching package: 'BiocIO'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:rtracklayer':\n",
      "\n",
      "    FileForFormat\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'Seurat' was built under R version 4.3.3\"\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Warning message:\n",
      "\"package 'SeuratObject' was built under R version 4.3.3\"\n",
      "Loading required package: sp\n",
      "\n",
      "Warning message:\n",
      "\"package 'sp' was built under R version 4.3.3\"\n",
      "\n",
      "Attaching package: 'sp'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'SeuratObject'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biostrings':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from 'package:SummarizedExperiment':\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from 'package:BiocGenerics':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Seurat'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:SummarizedExperiment':\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "\u001b[1m\u001b[22mscCustomize v2.1.2\n",
      "If you find the scCustomize useful please cite.\n",
      "See 'samuel-marsh.github.io/scCustomize/articles/FAQ.html' for citation info.\n",
      "\n",
      "Warning message:\n",
      "\"package 'ComplexUpset' was built under R version 4.3.3\"\n",
      "Warning message:\n",
      "\"package 'ComplexHeatmap' was built under R version 4.3.2\"\n",
      "Loading required package: grid\n",
      "\n",
      "\n",
      "Attaching package: 'grid'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biostrings':\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.18.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite either one:\n",
      "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
      "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "    genomic data. Bioinformatics 2016.\n",
      "\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#load package\n",
    ".libPaths(c(\"/cluster/home/liuhengxin/software/miniconda3_2/envs/r-kernel/lib/R/library\",\n",
    "           \"/cluster/home/liuhengxin/software/miniconda3/envs/r4.1.2/lib/R/library\",\n",
    "           \"/cluster/home/liuhengxin/software/miniconda3_2/envs/Rsplot/lib/R/library\"))\n",
    "library(data.table)\n",
    "library(qs)\n",
    "library(GenomicRanges)\n",
    "library(GenomicFeatures)\n",
    "library(rtracklayer)\n",
    "library(dplyr)\n",
    "library(ggbio)\n",
    "library(reshape2)\n",
    "library(ggsci)\n",
    "library(viridis)\n",
    "library(ggpubr)\n",
    "library(circlize)\n",
    "library(ggh4x)\n",
    "library(ggtranscript)\n",
    "library(khroma)\n",
    "library(randomcoloR)\n",
    "library(bambu)\n",
    "library(ggdensity)\n",
    "library(Seurat)\n",
    "library(scCustomize)\n",
    "library(ComplexUpset)\n",
    "library(ComplexHeatmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "36616b57-d162-439a-a79a-491b43c4a55f",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'unique': object 'areaan' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'unique': object 'areaan' not found\nTraceback:\n",
      "1. .handleSimpleError(function (cond) \n . .Internal(C_tryCatchHelper(addr, 1L, cond)), \"object 'areaan' not found\", \n .     base::quote(eval(expr, envir)))",
      "2. h(simpleError(msg, call))",
      "3. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = NULL)\n . }, \"error in evaluating the argument 'x' in selecting a method for function 'unique': object 'areaan' not found\", \n .     base::quote(h(simpleError(msg, call))))"
     ]
    }
   ],
   "source": [
    "#load data\n",
    "setwd(\"/cluster/home/liuhengxin/P1_brain_nanopore/downstream_analysis/\")\n",
    "MySplit = function(str,sep, n){\n",
    "  return(unlist(lapply(strsplit(str,sep),\"[[\",n)))\n",
    "}\n",
    "# cellan = read.csv(\"../ref/cell_type_type20230220-macaque1-contour2_v1.csv\")\n",
    "# cellan = unique(cellan[,c(\"subclass\",\"class\")])\n",
    "# areaan = read.csv(\"../ref/area_annotation.csv\")\n",
    "gtf = rtracklayer::import(\"ref/bambu_extended_annotations_novelname.gtf\")\n",
    "areaan = unique(areaan[c(\"L0\",\"L3\")])\n",
    "colnames(areaan) = c(\"lobe\",\"area\")\n",
    "areaan[areaan$area == \"1/2\",]$area = \"1|2\"\n",
    "gtfh = rtracklayer::import(\"ref/bambu_extended_annotations_novelname.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "63b038c6-336b-4e33-bc49-90e4d0ca3581",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>NA</li><li>'XM_005568473.2'</li><li>'DDRGK1-WeiLab-1'</li><li>'BambuTx20710'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item NA\n",
       "\\item 'XM\\_005568473.2'\n",
       "\\item 'DDRGK1-WeiLab-1'\n",
       "\\item 'BambuTx20710'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. NA\n",
       "2. 'XM_005568473.2'\n",
       "3. 'DDRGK1-WeiLab-1'\n",
       "4. 'BambuTx20710'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] NA                \"XM_005568473.2\"  \"DDRGK1-WeiLab-1\" \"BambuTx20710\"   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fsraw = qread(\"data/reads_full_anotation/read_total_annotation_nobackground_24_08_22.qs\")\n",
    "unique(fsraw[fsraw$gene_id == \"DDRGK1\",]$transcript_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "b6506ce8-f806-4bbb-b667-f3b569b996c0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>readid</th><th scope=col>origin_name</th><th scope=col>cell_id</th><th scope=col>cell_type_id</th><th scope=col>subclass</th><th scope=col>class</th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>x</th><th scope=col>y</th><th scope=col>layer</th><th scope=col>region</th><th scope=col>lobe</th><th scope=col>subcortical</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>375</th><td>14492470-8f38-4166-a200-16d025097700-8;st_ed:A:3091-3566;AD:A:adapterD.rc-end;AP:A:Boundary_21-45:-                 </td><td>L-1|2-l1</td><td>NA</td><td>NA</td><td></td><td></td><td>NA            </td><td>NA       </td><td>10000</td><td>26429</td><td>l1</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "\t<tr><th scope=row>376</th><td>9eda4dea-9dc0-434f-871e-9448b11daf53-12;st_ed:A:6349-7977;AD:A:adapterC.rc-adapterM;AP:A:RC-Wrong_21-45:-           </td><td>L-1|2-l1</td><td>NA</td><td>NA</td><td></td><td></td><td>XM_005543577.2</td><td>RPS8     </td><td>10000</td><td>26457</td><td>l1</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "\t<tr><th scope=row>377</th><td>41fcd1d6-b6f1-49a8-98c4-1c64118f7381-10;st_ed:A:4570-5935;AD:A:adapterI-adapterK;AP:A:Jump|AD:A:no-split_1322-1346:+</td><td>L-1|2-l1</td><td>NA</td><td>NA</td><td></td><td></td><td>XM_005561040.2</td><td>ARHGAP5  </td><td>10000</td><td>26465</td><td>l1</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "\t<tr><th scope=row>378</th><td>8e9fe2ef-d7b3-459a-94a2-f2638c24f2e0-3;st_ed:A:709-1447;AD:A:adapterE-adapterF;AP:A:Correct_695-719:+               </td><td>L-1|2-l1</td><td>NA</td><td>NA</td><td></td><td></td><td>XM_005594282.1</td><td>CXHXorf57</td><td>10000</td><td>26495</td><td>l1</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "\t<tr><th scope=row>379</th><td>d3ff0bb6-eb5b-4e44-a69a-58a35ddf805b-6;st_ed:A:7011-7543;AD:A:adapterL-adapterM;AP:A:Correct_486-510:+              </td><td>L-1|2-l1</td><td>NA</td><td>NA</td><td></td><td></td><td>NA            </td><td>NA       </td><td>10000</td><td>26501</td><td>l1</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "\t<tr><th scope=row>380</th><td>03f0a4dd-0f1f-4d24-b429-21d043ccc63d-7;st_ed:A:3452-4287;AD:A:adapterI-adapterJ;AP:A:Correct|ts-te:115-823          </td><td>L-1|2-l1</td><td>NA</td><td>NA</td><td></td><td></td><td>NA            </td><td>NA       </td><td>10000</td><td>26505</td><td>l1</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 14\n",
       "\\begin{tabular}{r|llllllllllllll}\n",
       "  & readid & origin\\_name & cell\\_id & cell\\_type\\_id & subclass & class & transcript\\_id & gene\\_id & x & y & layer & region & lobe & subcortical\\\\\n",
       "  & <chr> & <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t375 & 14492470-8f38-4166-a200-16d025097700-8;st\\_ed:A:3091-3566;AD:A:adapterD.rc-end;AP:A:Boundary\\_21-45:-                  & L-1\\textbar{}2-l1 & NA & NA &  &  & NA             & NA        & 10000 & 26429 & l1 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\t376 & 9eda4dea-9dc0-434f-871e-9448b11daf53-12;st\\_ed:A:6349-7977;AD:A:adapterC.rc-adapterM;AP:A:RC-Wrong\\_21-45:-            & L-1\\textbar{}2-l1 & NA & NA &  &  & XM\\_005543577.2 & RPS8      & 10000 & 26457 & l1 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\t377 & 41fcd1d6-b6f1-49a8-98c4-1c64118f7381-10;st\\_ed:A:4570-5935;AD:A:adapterI-adapterK;AP:A:Jump\\textbar{}AD:A:no-split\\_1322-1346:+ & L-1\\textbar{}2-l1 & NA & NA &  &  & XM\\_005561040.2 & ARHGAP5   & 10000 & 26465 & l1 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\t378 & 8e9fe2ef-d7b3-459a-94a2-f2638c24f2e0-3;st\\_ed:A:709-1447;AD:A:adapterE-adapterF;AP:A:Correct\\_695-719:+                & L-1\\textbar{}2-l1 & NA & NA &  &  & XM\\_005594282.1 & CXHXorf57 & 10000 & 26495 & l1 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\t379 & d3ff0bb6-eb5b-4e44-a69a-58a35ddf805b-6;st\\_ed:A:7011-7543;AD:A:adapterL-adapterM;AP:A:Correct\\_486-510:+               & L-1\\textbar{}2-l1 & NA & NA &  &  & NA             & NA        & 10000 & 26501 & l1 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\t380 & 03f0a4dd-0f1f-4d24-b429-21d043ccc63d-7;st\\_ed:A:3452-4287;AD:A:adapterI-adapterJ;AP:A:Correct\\textbar{}ts-te:115-823           & L-1\\textbar{}2-l1 & NA & NA &  &  & NA             & NA        & 10000 & 26505 & l1 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 14\n",
       "\n",
       "| <!--/--> | readid &lt;chr&gt; | origin_name &lt;chr&gt; | cell_id &lt;dbl&gt; | cell_type_id &lt;dbl&gt; | subclass &lt;chr&gt; | class &lt;chr&gt; | transcript_id &lt;chr&gt; | gene_id &lt;chr&gt; | x &lt;dbl&gt; | y &lt;dbl&gt; | layer &lt;chr&gt; | region &lt;chr&gt; | lobe &lt;chr&gt; | subcortical &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 375 | 14492470-8f38-4166-a200-16d025097700-8;st_ed:A:3091-3566;AD:A:adapterD.rc-end;AP:A:Boundary_21-45:-                  | L-1|2-l1 | NA | NA | <!----> | <!----> | NA             | NA        | 10000 | 26429 | l1 | 1|2 | somatosensory | Cortex |\n",
       "| 376 | 9eda4dea-9dc0-434f-871e-9448b11daf53-12;st_ed:A:6349-7977;AD:A:adapterC.rc-adapterM;AP:A:RC-Wrong_21-45:-            | L-1|2-l1 | NA | NA | <!----> | <!----> | XM_005543577.2 | RPS8      | 10000 | 26457 | l1 | 1|2 | somatosensory | Cortex |\n",
       "| 377 | 41fcd1d6-b6f1-49a8-98c4-1c64118f7381-10;st_ed:A:4570-5935;AD:A:adapterI-adapterK;AP:A:Jump|AD:A:no-split_1322-1346:+ | L-1|2-l1 | NA | NA | <!----> | <!----> | XM_005561040.2 | ARHGAP5   | 10000 | 26465 | l1 | 1|2 | somatosensory | Cortex |\n",
       "| 378 | 8e9fe2ef-d7b3-459a-94a2-f2638c24f2e0-3;st_ed:A:709-1447;AD:A:adapterE-adapterF;AP:A:Correct_695-719:+                | L-1|2-l1 | NA | NA | <!----> | <!----> | XM_005594282.1 | CXHXorf57 | 10000 | 26495 | l1 | 1|2 | somatosensory | Cortex |\n",
       "| 379 | d3ff0bb6-eb5b-4e44-a69a-58a35ddf805b-6;st_ed:A:7011-7543;AD:A:adapterL-adapterM;AP:A:Correct_486-510:+               | L-1|2-l1 | NA | NA | <!----> | <!----> | NA             | NA        | 10000 | 26501 | l1 | 1|2 | somatosensory | Cortex |\n",
       "| 380 | 03f0a4dd-0f1f-4d24-b429-21d043ccc63d-7;st_ed:A:3452-4287;AD:A:adapterI-adapterJ;AP:A:Correct|ts-te:115-823           | L-1|2-l1 | NA | NA | <!----> | <!----> | NA             | NA        | 10000 | 26505 | l1 | 1|2 | somatosensory | Cortex |\n",
       "\n"
      ],
      "text/plain": [
       "    readid                                                                                                              \n",
       "375 14492470-8f38-4166-a200-16d025097700-8;st_ed:A:3091-3566;AD:A:adapterD.rc-end;AP:A:Boundary_21-45:-                 \n",
       "376 9eda4dea-9dc0-434f-871e-9448b11daf53-12;st_ed:A:6349-7977;AD:A:adapterC.rc-adapterM;AP:A:RC-Wrong_21-45:-           \n",
       "377 41fcd1d6-b6f1-49a8-98c4-1c64118f7381-10;st_ed:A:4570-5935;AD:A:adapterI-adapterK;AP:A:Jump|AD:A:no-split_1322-1346:+\n",
       "378 8e9fe2ef-d7b3-459a-94a2-f2638c24f2e0-3;st_ed:A:709-1447;AD:A:adapterE-adapterF;AP:A:Correct_695-719:+               \n",
       "379 d3ff0bb6-eb5b-4e44-a69a-58a35ddf805b-6;st_ed:A:7011-7543;AD:A:adapterL-adapterM;AP:A:Correct_486-510:+              \n",
       "380 03f0a4dd-0f1f-4d24-b429-21d043ccc63d-7;st_ed:A:3452-4287;AD:A:adapterI-adapterJ;AP:A:Correct|ts-te:115-823          \n",
       "    origin_name cell_id cell_type_id subclass class transcript_id  gene_id  \n",
       "375 L-1|2-l1    NA      NA                          NA             NA       \n",
       "376 L-1|2-l1    NA      NA                          XM_005543577.2 RPS8     \n",
       "377 L-1|2-l1    NA      NA                          XM_005561040.2 ARHGAP5  \n",
       "378 L-1|2-l1    NA      NA                          XM_005594282.1 CXHXorf57\n",
       "379 L-1|2-l1    NA      NA                          NA             NA       \n",
       "380 L-1|2-l1    NA      NA                          NA             NA       \n",
       "    x     y     layer region lobe          subcortical\n",
       "375 10000 26429 l1    1|2    somatosensory Cortex     \n",
       "376 10000 26457 l1    1|2    somatosensory Cortex     \n",
       "377 10000 26465 l1    1|2    somatosensory Cortex     \n",
       "378 10000 26495 l1    1|2    somatosensory Cortex     \n",
       "379 10000 26501 l1    1|2    somatosensory Cortex     \n",
       "380 10000 26505 l1    1|2    somatosensory Cortex     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(fsraw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "336ec158-89ad-41bf-9efa-88a1ddc2cf77",
   "metadata": {},
   "outputs": [],
   "source": [
    "#bambu annotation\n",
    "library(bambu)\n",
    "bambuAnnotations <- prepareAnnotations(\"/cluster/home/liuhengxin/software/refdata/GRCm39/genes/genes.gtf\")\n",
    "se <- bambu(reads = c(\"/cluster/home/liuhengxin/P_fullscope_crosspecies/data//preprocess/mouse_near_T44_C12cyclone/Alignment/C12T4_cyclone.bam\",\n",
    "                     \"/cluster/home/liuhengxin/P_fullscope_crosspecies/data/preprocess/mouse_near_T44_C12ONT/Alignment/C12T4.bam\",\n",
    "                     \"/cluster/home/liuhengxin/P_fullscope_crosspecies/data/preprocess/mouse_near_T44_C12pacbio/alignment/C12L_pacbio_250604.sorted.bam\"), \n",
    "            annotations = bambuAnnotations, genome = \"/cluster/home/liuhengxin/software/refdata/GRCm39/fasta/genome.fa\",NDR = 1,\n",
    "           trackReads = TRUE)\n",
    "se\n",
    "#bambu gtf annotation\n",
    "#se = qread(\"/cluster/home/liuhengxin/P_fullscope_mousebrain/data/preprocess/mouse_near_T44/bambu/bambu_transcript_annotation.qs\")\n",
    "traninfo = mcols(se)\n",
    "flct = assays(se)$fullLengthCounts\n",
    "traninfo$fullLengthCounts = rowSums(flct)\n",
    "traninfofil = traninfo[(!is.na(traninfo$NDR) & traninfo$NDR<0.80) | is.na(traninfo$NDR) | traninfo$txClassDescription == \"annotation\",]\n",
    "print(nrow(traninfofil))\n",
    "readtrans = rowRanges(se)\n",
    "readtrans\n",
    "readtranshub = readtrans[c(traninfofil$TXNAME)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1db8bd2b-e339-4c9c-a2e5-3c1d2ee3bec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "T47 = qread(\"/cluster/home/liuhengxin/P1_brain_nanopore/processed_data/T47_analysis/T47_bambu_annotations_25_06_26.qs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7f062988-5639-4b67-8f15-d11b717c7842",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NULL"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bb.raw = qread(\"/cluster/home/hongyanhong/projects/macaca_new_data_bam/bambu2/bambu.qs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4714f2f4-a36c-4c74-92ff-65c7215ea00b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 8 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ONT_BV827-001N0001_clean.part_001.splitted.fasta.gz.sorted.rename2</th><th scope=col>ONT_BV827-001N0001_clean.part_002.splitted.fasta.gz.sorted.rename2</th><th scope=col>ONT_BV827-001N0001_clean.part_003.splitted.fasta.gz.sorted.rename2</th><th scope=col>ONT_BV827-001N0001_clean.part_004.splitted.fasta.gz.sorted.rename2</th><th scope=col>ONT_BV827-001N0001_clean.part_005.splitted.fasta.gz.sorted.rename2</th><th scope=col>ONT_BV827-001N0001_clean.part_006.splitted.fasta.gz.sorted.rename2</th><th scope=col>ONT_BV827-001N0001_clean.part_008.splitted.fasta.gz.sorted.rename2</th><th scope=col>ONT_BV827-001N0001_clean.part_009.splitted.fasta.gz.sorted.rename2</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>BambuTx1</th><td>14</td><td>17</td><td>16</td><td>15</td><td>25</td><td>15</td><td>17</td><td>8</td></tr>\n",
       "\t<tr><th scope=row>BambuTx10</th><td> 9</td><td> 4</td><td> 5</td><td> 2</td><td> 5</td><td> 2</td><td> 7</td><td>3</td></tr>\n",
       "\t<tr><th scope=row>BambuTx1000</th><td> 1</td><td> 5</td><td> 8</td><td> 7</td><td> 5</td><td> 6</td><td> 4</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>BambuTx10000</th><td> 0</td><td> 0</td><td> 1</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>BambuTx100000</th><td> 0</td><td> 3</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td> 0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>BambuTx100007</th><td> 1</td><td> 1</td><td> 0</td><td> 2</td><td> 0</td><td> 0</td><td> 3</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 8 of type dbl\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & ONT\\_BV827-001N0001\\_clean.part\\_001.splitted.fasta.gz.sorted.rename2 & ONT\\_BV827-001N0001\\_clean.part\\_002.splitted.fasta.gz.sorted.rename2 & ONT\\_BV827-001N0001\\_clean.part\\_003.splitted.fasta.gz.sorted.rename2 & ONT\\_BV827-001N0001\\_clean.part\\_004.splitted.fasta.gz.sorted.rename2 & ONT\\_BV827-001N0001\\_clean.part\\_005.splitted.fasta.gz.sorted.rename2 & ONT\\_BV827-001N0001\\_clean.part\\_006.splitted.fasta.gz.sorted.rename2 & ONT\\_BV827-001N0001\\_clean.part\\_008.splitted.fasta.gz.sorted.rename2 & ONT\\_BV827-001N0001\\_clean.part\\_009.splitted.fasta.gz.sorted.rename2\\\\\n",
       "\\hline\n",
       "\tBambuTx1 & 14 & 17 & 16 & 15 & 25 & 15 & 17 & 8\\\\\n",
       "\tBambuTx10 &  9 &  4 &  5 &  2 &  5 &  2 &  7 & 3\\\\\n",
       "\tBambuTx1000 &  1 &  5 &  8 &  7 &  5 &  6 &  4 & 1\\\\\n",
       "\tBambuTx10000 &  0 &  0 &  1 &  0 &  0 &  0 &  0 & 0\\\\\n",
       "\tBambuTx100000 &  0 &  3 &  0 &  0 &  0 &  0 &  0 & 0\\\\\n",
       "\tBambuTx100007 &  1 &  1 &  0 &  2 &  0 &  0 &  3 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 8 of type dbl\n",
       "\n",
       "| <!--/--> | ONT_BV827-001N0001_clean.part_001.splitted.fasta.gz.sorted.rename2 | ONT_BV827-001N0001_clean.part_002.splitted.fasta.gz.sorted.rename2 | ONT_BV827-001N0001_clean.part_003.splitted.fasta.gz.sorted.rename2 | ONT_BV827-001N0001_clean.part_004.splitted.fasta.gz.sorted.rename2 | ONT_BV827-001N0001_clean.part_005.splitted.fasta.gz.sorted.rename2 | ONT_BV827-001N0001_clean.part_006.splitted.fasta.gz.sorted.rename2 | ONT_BV827-001N0001_clean.part_008.splitted.fasta.gz.sorted.rename2 | ONT_BV827-001N0001_clean.part_009.splitted.fasta.gz.sorted.rename2 |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| BambuTx1 | 14 | 17 | 16 | 15 | 25 | 15 | 17 | 8 |\n",
       "| BambuTx10 |  9 |  4 |  5 |  2 |  5 |  2 |  7 | 3 |\n",
       "| BambuTx1000 |  1 |  5 |  8 |  7 |  5 |  6 |  4 | 1 |\n",
       "| BambuTx10000 |  0 |  0 |  1 |  0 |  0 |  0 |  0 | 0 |\n",
       "| BambuTx100000 |  0 |  3 |  0 |  0 |  0 |  0 |  0 | 0 |\n",
       "| BambuTx100007 |  1 |  1 |  0 |  2 |  0 |  0 |  3 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "              ONT_BV827-001N0001_clean.part_001.splitted.fasta.gz.sorted.rename2\n",
       "BambuTx1      14                                                                \n",
       "BambuTx10      9                                                                \n",
       "BambuTx1000    1                                                                \n",
       "BambuTx10000   0                                                                \n",
       "BambuTx100000  0                                                                \n",
       "BambuTx100007  1                                                                \n",
       "              ONT_BV827-001N0001_clean.part_002.splitted.fasta.gz.sorted.rename2\n",
       "BambuTx1      17                                                                \n",
       "BambuTx10      4                                                                \n",
       "BambuTx1000    5                                                                \n",
       "BambuTx10000   0                                                                \n",
       "BambuTx100000  3                                                                \n",
       "BambuTx100007  1                                                                \n",
       "              ONT_BV827-001N0001_clean.part_003.splitted.fasta.gz.sorted.rename2\n",
       "BambuTx1      16                                                                \n",
       "BambuTx10      5                                                                \n",
       "BambuTx1000    8                                                                \n",
       "BambuTx10000   1                                                                \n",
       "BambuTx100000  0                                                                \n",
       "BambuTx100007  0                                                                \n",
       "              ONT_BV827-001N0001_clean.part_004.splitted.fasta.gz.sorted.rename2\n",
       "BambuTx1      15                                                                \n",
       "BambuTx10      2                                                                \n",
       "BambuTx1000    7                                                                \n",
       "BambuTx10000   0                                                                \n",
       "BambuTx100000  0                                                                \n",
       "BambuTx100007  2                                                                \n",
       "              ONT_BV827-001N0001_clean.part_005.splitted.fasta.gz.sorted.rename2\n",
       "BambuTx1      25                                                                \n",
       "BambuTx10      5                                                                \n",
       "BambuTx1000    5                                                                \n",
       "BambuTx10000   0                                                                \n",
       "BambuTx100000  0                                                                \n",
       "BambuTx100007  0                                                                \n",
       "              ONT_BV827-001N0001_clean.part_006.splitted.fasta.gz.sorted.rename2\n",
       "BambuTx1      15                                                                \n",
       "BambuTx10      2                                                                \n",
       "BambuTx1000    6                                                                \n",
       "BambuTx10000   0                                                                \n",
       "BambuTx100000  0                                                                \n",
       "BambuTx100007  0                                                                \n",
       "              ONT_BV827-001N0001_clean.part_008.splitted.fasta.gz.sorted.rename2\n",
       "BambuTx1      17                                                                \n",
       "BambuTx10      7                                                                \n",
       "BambuTx1000    4                                                                \n",
       "BambuTx10000   0                                                                \n",
       "BambuTx100000  0                                                                \n",
       "BambuTx100007  3                                                                \n",
       "              ONT_BV827-001N0001_clean.part_009.splitted.fasta.gz.sorted.rename2\n",
       "BambuTx1      8                                                                 \n",
       "BambuTx10     3                                                                 \n",
       "BambuTx1000   1                                                                 \n",
       "BambuTx10000  0                                                                 \n",
       "BambuTx100000 0                                                                 \n",
       "BambuTx100007 0                                                                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fgtfi = assays(bb.pcb)$fullLengthCounts\n",
    "head(fgtfi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "114d0c5d-f06b-4c88-9f38-fd9f370a046d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>transid</th><th scope=col>fulllength</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>BambuTx1</th><td>BambuTx1     </td><td>254</td></tr>\n",
       "\t<tr><th scope=row>BambuTx10</th><td>BambuTx10    </td><td> 74</td></tr>\n",
       "\t<tr><th scope=row>BambuTx1000</th><td>BambuTx1000  </td><td> 74</td></tr>\n",
       "\t<tr><th scope=row>BambuTx10000</th><td>BambuTx10000 </td><td>  2</td></tr>\n",
       "\t<tr><th scope=row>BambuTx100000</th><td>BambuTx100000</td><td>  6</td></tr>\n",
       "\t<tr><th scope=row>BambuTx100007</th><td>BambuTx100007</td><td> 14</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & transid & fulllength\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tBambuTx1 & BambuTx1      & 254\\\\\n",
       "\tBambuTx10 & BambuTx10     &  74\\\\\n",
       "\tBambuTx1000 & BambuTx1000   &  74\\\\\n",
       "\tBambuTx10000 & BambuTx10000  &   2\\\\\n",
       "\tBambuTx100000 & BambuTx100000 &   6\\\\\n",
       "\tBambuTx100007 & BambuTx100007 &  14\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | transid &lt;chr&gt; | fulllength &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| BambuTx1 | BambuTx1      | 254 |\n",
       "| BambuTx10 | BambuTx10     |  74 |\n",
       "| BambuTx1000 | BambuTx1000   |  74 |\n",
       "| BambuTx10000 | BambuTx10000  |   2 |\n",
       "| BambuTx100000 | BambuTx100000 |   6 |\n",
       "| BambuTx100007 | BambuTx100007 |  14 |\n",
       "\n"
      ],
      "text/plain": [
       "              transid       fulllength\n",
       "BambuTx1      BambuTx1      254       \n",
       "BambuTx10     BambuTx10      74       \n",
       "BambuTx1000   BambuTx1000    74       \n",
       "BambuTx10000  BambuTx10000    2       \n",
       "BambuTx100000 BambuTx100000   6       \n",
       "BambuTx100007 BambuTx100007  14       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fgtfi = as.data.frame(fgtfi)\n",
    "fgtfi$fulllength = rowSums(fgtfi)\n",
    "fgtfi$transid = rownames(fgtfi)\n",
    "fgtfi = fgtfi[,c(\"transid\",\"fulllength\")]\n",
    "head(fgtfi)\n",
    "write.table(fgtfi,file = \"data/fullscope_qc/fullscope-pacbio.tsv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7234e922-ae9a-450b-a9f7-b47fef183b06",
   "metadata": {},
   "source": [
    "## Filter gtf make annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8717bc67-4307-4493-8051-bbbd4336c1b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 8 metadata columns:\n",
       "         seqnames    ranges strand |   source       type     score     phase\n",
       "            <Rle> <IRanges>  <Rle> | <factor>   <factor> <numeric> <integer>\n",
       "  [1] NC_012670.1  620-2452      + |    Bambu transcript        NA      <NA>\n",
       "  [2] NC_012670.1  620-1566      + |    Bambu exon              NA      <NA>\n",
       "  [3] NC_012670.1 1633-2452      + |    Bambu exon              NA      <NA>\n",
       "  [4] NC_012670.1 1637-2483      - |    Bambu transcript        NA      <NA>\n",
       "  [5] NC_012670.1 1637-2300      - |    Bambu exon              NA      <NA>\n",
       "  [6] NC_012670.1 2354-2483      - |    Bambu exon              NA      <NA>\n",
       "              gene_id        transcript_id        transcript_name exon_number\n",
       "          <character>          <character>            <character> <character>\n",
       "  [1] BambuGene146379 BambuGene146379-FS-1 BambuGene146379-WeiL..        <NA>\n",
       "  [2] BambuGene146379 BambuGene146379-FS-1 BambuGene146379-WeiL..           1\n",
       "  [3] BambuGene146379 BambuGene146379-FS-1 BambuGene146379-WeiL..           2\n",
       "  [4] BambuGene146394 BambuGene146394-FS-1 BambuGene146394-WeiL..        <NA>\n",
       "  [5] BambuGene146394 BambuGene146394-FS-1 BambuGene146394-WeiL..           2\n",
       "  [6] BambuGene146394 BambuGene146394-FS-1 BambuGene146394-WeiL..           1\n",
       "  -------\n",
       "  seqinfo: 997 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 8 metadata columns:\n",
       "         seqnames    ranges strand |   source       type     score     phase\n",
       "            <Rle> <IRanges>  <Rle> | <factor>   <factor> <numeric> <integer>\n",
       "  [1] NC_012670.1  620-2452      + |    Bambu transcript        NA      <NA>\n",
       "  [2] NC_012670.1  620-1566      + |    Bambu exon              NA      <NA>\n",
       "  [3] NC_012670.1 1633-2452      + |    Bambu exon              NA      <NA>\n",
       "  [4] NC_012670.1 1637-2483      - |    Bambu transcript        NA      <NA>\n",
       "  [5] NC_012670.1 1637-2300      - |    Bambu exon              NA      <NA>\n",
       "  [6] NC_012670.1 2354-2483      - |    Bambu exon              NA      <NA>\n",
       "              gene_id        transcript_id        transcript_name exon_number\n",
       "          <character>          <character>            <character> <character>\n",
       "  [1] BambuGene146379 BambuGene146379-FS-1 BambuGene146379-WeiL..        <NA>\n",
       "  [2] BambuGene146379 BambuGene146379-FS-1 BambuGene146379-WeiL..           1\n",
       "  [3] BambuGene146379 BambuGene146379-FS-1 BambuGene146379-WeiL..           2\n",
       "  [4] BambuGene146394 BambuGene146394-FS-1 BambuGene146394-WeiL..        <NA>\n",
       "  [5] BambuGene146394 BambuGene146394-FS-1 BambuGene146394-WeiL..           2\n",
       "  [6] BambuGene146394 BambuGene146394-FS-1 BambuGene146394-WeiL..           1\n",
       "  -------\n",
       "  seqinfo: 997 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#rename gtf\n",
    "head(gtf)\n",
    "gtf$transcript_id = gsub(gtf$transcript_id,pattern = \"WeiLab\",replacement = \"FS\")\n",
    "head(gtf)\n",
    "rtracklayer::export(gtf,\"ref/bambu_extended_annotations_rename_25_12_09.gtf\",format = \"gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37e42b6c-7664-4ce9-8cf8-71071b39a4b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(bambu)\n",
    "library(qs)\n",
    "bb.ont = qread(\"/cluster/home/hongyanhong/projects/MAS-seq/240318_newdata/bambu/bambu2.qs\")\n",
    "bb.cyl = qread(\"/cluster/home/hongyanhong/projects/MAS-seq/BGI_240325_newdata/bambu/bambu2.qs\")\n",
    "bb.raw = qread(\"/cluster/home/hongyanhong/projects/macaca_new_data_bam/bambu2/bambu.qs\")\n",
    "bb.pcb = qread(\"/cluster/home/hongyanhong/projects/MAS-seq/240318_newdata/bambu/bambu2.qs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "22fc28c7-f419-4afd-9ccf-8e4546e4f830",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'/cluster/home/hongyanhong/projects/MAS-seq/combined_fulllenmatrix//directONT.tsv'</li><li>'/cluster/home/hongyanhong/projects/MAS-seq/combined_fulllenmatrix//fullscope_BGI_df.tsv'</li><li>'/cluster/home/hongyanhong/projects/MAS-seq/combined_fulllenmatrix//fullscope_ONT_df.tsv'</li><li>'data/fullscope_qc/fullscope-pacbio.tsv'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '/cluster/home/hongyanhong/projects/MAS-seq/combined\\_fulllenmatrix//directONT.tsv'\n",
       "\\item '/cluster/home/hongyanhong/projects/MAS-seq/combined\\_fulllenmatrix//fullscope\\_BGI\\_df.tsv'\n",
       "\\item '/cluster/home/hongyanhong/projects/MAS-seq/combined\\_fulllenmatrix//fullscope\\_ONT\\_df.tsv'\n",
       "\\item 'data/fullscope\\_qc/fullscope-pacbio.tsv'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '/cluster/home/hongyanhong/projects/MAS-seq/combined_fulllenmatrix//directONT.tsv'\n",
       "2. '/cluster/home/hongyanhong/projects/MAS-seq/combined_fulllenmatrix//fullscope_BGI_df.tsv'\n",
       "3. '/cluster/home/hongyanhong/projects/MAS-seq/combined_fulllenmatrix//fullscope_ONT_df.tsv'\n",
       "4. 'data/fullscope_qc/fullscope-pacbio.tsv'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"/cluster/home/hongyanhong/projects/MAS-seq/combined_fulllenmatrix//directONT.tsv\"       \n",
       "[2] \"/cluster/home/hongyanhong/projects/MAS-seq/combined_fulllenmatrix//fullscope_BGI_df.tsv\"\n",
       "[3] \"/cluster/home/hongyanhong/projects/MAS-seq/combined_fulllenmatrix//fullscope_ONT_df.tsv\"\n",
       "[4] \"data/fullscope_qc/fullscope-pacbio.tsv\"                                                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "files = list.files(\"/cluster/home/hongyanhong/projects/MAS-seq/combined_fulllenmatrix/\",full.names = T)\n",
    "files = c(files[-4],\"data/fullscope_qc/fullscope-pacbio.tsv\")\n",
    "files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "21155566-ec51-47b3-948c-1c1c1172b2c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>transid</th><th scope=col>fulllength</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>BambuTx1</th><td>BambuTx1</td><td>129.00000</td></tr>\n",
       "\t<tr><th scope=row>BambuTx2</th><td>BambuTx2</td><td>  5.00001</td></tr>\n",
       "\t<tr><th scope=row>BambuTx3</th><td>BambuTx3</td><td> 19.00000</td></tr>\n",
       "\t<tr><th scope=row>BambuTx4</th><td>BambuTx4</td><td>629.00000</td></tr>\n",
       "\t<tr><th scope=row>BambuTx5</th><td>BambuTx5</td><td>  9.00000</td></tr>\n",
       "\t<tr><th scope=row>BambuTx6</th><td>BambuTx6</td><td> 38.00000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & transid & fulllength\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tBambuTx1 & BambuTx1 & 129.00000\\\\\n",
       "\tBambuTx2 & BambuTx2 &   5.00001\\\\\n",
       "\tBambuTx3 & BambuTx3 &  19.00000\\\\\n",
       "\tBambuTx4 & BambuTx4 & 629.00000\\\\\n",
       "\tBambuTx5 & BambuTx5 &   9.00000\\\\\n",
       "\tBambuTx6 & BambuTx6 &  38.00000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | transid &lt;chr&gt; | fulllength &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| BambuTx1 | BambuTx1 | 129.00000 |\n",
       "| BambuTx2 | BambuTx2 |   5.00001 |\n",
       "| BambuTx3 | BambuTx3 |  19.00000 |\n",
       "| BambuTx4 | BambuTx4 | 629.00000 |\n",
       "| BambuTx5 | BambuTx5 |   9.00000 |\n",
       "| BambuTx6 | BambuTx6 |  38.00000 |\n",
       "\n"
      ],
      "text/plain": [
       "         transid  fulllength\n",
       "BambuTx1 BambuTx1 129.00000 \n",
       "BambuTx2 BambuTx2   5.00001 \n",
       "BambuTx3 BambuTx3  19.00000 \n",
       "BambuTx4 BambuTx4 629.00000 \n",
       "BambuTx5 BambuTx5   9.00000 \n",
       "BambuTx6 BambuTx6  38.00000 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fgtf = NULL\n",
    "for(i in files[1:4]){\n",
    "    fi = read.table(i)\n",
    "    colnames(fi) = c(\"transid\",\"fulllength\")\n",
    "    fgtf = rbind(fgtf,fi)\n",
    "}\n",
    "head(fgtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d815a3f7-2879-4996-a214-6ac6626fdd52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene_id</th><th scope=col>transcript_id</th><th scope=col>trantype</th><th scope=col>transnew</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>AAAS </td><td>BambuTx122957</td><td>novel</td><td>AAAS-WeiLab-1 </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>AAAS </td><td>BambuTx22391 </td><td>novel</td><td>AAAS-WeiLab-2 </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>AADAT</td><td>BambuTx11228 </td><td>novel</td><td>AADAT-WeiLab-1</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>AAED1</td><td>BambuTx30778 </td><td>novel</td><td>AAED1-WeiLab-1</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>AAED1</td><td>BambuTx30780 </td><td>novel</td><td>AAED1-WeiLab-2</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>AAMDC</td><td>BambuTx28589 </td><td>novel</td><td>AAMDC-WeiLab-1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & gene\\_id & transcript\\_id & trantype & transnew\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & AAAS  & BambuTx122957 & novel & AAAS-WeiLab-1 \\\\\n",
       "\t2 & AAAS  & BambuTx22391  & novel & AAAS-WeiLab-2 \\\\\n",
       "\t3 & AADAT & BambuTx11228  & novel & AADAT-WeiLab-1\\\\\n",
       "\t4 & AAED1 & BambuTx30778  & novel & AAED1-WeiLab-1\\\\\n",
       "\t5 & AAED1 & BambuTx30780  & novel & AAED1-WeiLab-2\\\\\n",
       "\t6 & AAMDC & BambuTx28589  & novel & AAMDC-WeiLab-1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| <!--/--> | gene_id &lt;chr&gt; | transcript_id &lt;chr&gt; | trantype &lt;chr&gt; | transnew &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 1 | AAAS  | BambuTx122957 | novel | AAAS-WeiLab-1  |\n",
       "| 2 | AAAS  | BambuTx22391  | novel | AAAS-WeiLab-2  |\n",
       "| 3 | AADAT | BambuTx11228  | novel | AADAT-WeiLab-1 |\n",
       "| 4 | AAED1 | BambuTx30778  | novel | AAED1-WeiLab-1 |\n",
       "| 5 | AAED1 | BambuTx30780  | novel | AAED1-WeiLab-2 |\n",
       "| 6 | AAMDC | BambuTx28589  | novel | AAMDC-WeiLab-1 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene_id transcript_id trantype transnew      \n",
       "1 AAAS    BambuTx122957 novel    AAAS-WeiLab-1 \n",
       "2 AAAS    BambuTx22391  novel    AAAS-WeiLab-2 \n",
       "3 AADAT   BambuTx11228  novel    AADAT-WeiLab-1\n",
       "4 AAED1   BambuTx30778  novel    AAED1-WeiLab-1\n",
       "5 AAED1   BambuTx30780  novel    AAED1-WeiLab-2\n",
       "6 AAMDC   BambuTx28589  novel    AAMDC-WeiLab-1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "127050"
      ],
      "text/latex": [
       "127050"
      ],
      "text/markdown": [
       "127050"
      ],
      "text/plain": [
       "[1] 127050"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "135984"
      ],
      "text/latex": [
       "135984"
      ],
      "text/markdown": [
       "135984"
      ],
      "text/plain": [
       "[1] 135984"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#make a full annotation and rename matrix\n",
    "gtfdfan = read.csv(\"ref/gtf_new_transcript_name_map_04_30.csv\")\n",
    "gtfdfan = unique(gtfdfan[,c(\"gene_id\",\"transcript_id\",\"trantype\",\"transnew\")])\n",
    "nrow(gtfdfan)\n",
    "gtfdfant = unique(fsraw[,c(\"gene_id\",\"transcript_id\")])\n",
    "gtfdfant = gtfdfant[!is.na(gtfdfant$transcript_id),]\n",
    "head(gtfdfant)\n",
    "nrow(gtfdfant)\n",
    "head(gtfdfan)\n",
    "nrow(gtfdfant)\n",
    "gtfdfant = merge(gtfdfant,gtfdfan,by = c(\"gene_id\",\"transcript_id\"),all = TRUE)\n",
    "nrow(gtfdfant)\n",
    "qsave(gtfdfant,file = \"ref/bambu_trans_total_oldname_newname_25_10_09.qs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "9e3d0514-9266-4563-b312-0b9243ff6ebd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 6 rows and 12 columns\n",
       "               TXNAME        GENEID       NDR novelGene novelTranscript\n",
       "          <character>   <character> <numeric> <logical>       <logical>\n",
       "BambuTx1     BambuTx1         PGBD2  0.548446     FALSE            TRUE\n",
       "BambuTx4     BambuTx4  LOC107130781  0.258162     FALSE            TRUE\n",
       "BambuTx6     BambuTx6         NLRP3  0.725336     FALSE            TRUE\n",
       "BambuTx8     BambuTx8        ZNF124  0.783240     FALSE            TRUE\n",
       "BambuTx10   BambuTx10        ZNF669  0.792865     FALSE            TRUE\n",
       "BambuTx11   BambuTx11 BambuGene6274  0.507450      TRUE            TRUE\n",
       "              txClassDescription readCount relReadCount relSubsetCount\n",
       "                     <character> <integer>    <numeric>      <numeric>\n",
       "BambuTx1  newLastJunction:newL..       127    0.4136808       1.000000\n",
       "BambuTx4                  allNew       629    0.9402093       1.000000\n",
       "BambuTx6                  allNew        33    0.1601942       1.000000\n",
       "BambuTx8  newLastJunction:newF..         2    0.0140845       1.000000\n",
       "BambuTx10 newLastJunction:newJ..        65    0.0718232       1.000000\n",
       "BambuTx11        newGene-spliced        71    0.8352941       0.835294\n",
       "               txid     eqClassById fullLengthCounts\n",
       "          <integer>   <IntegerList>        <numeric>\n",
       "BambuTx1          1               1         129.0000\n",
       "BambuTx4          4               4         629.0000\n",
       "BambuTx6          6               6          38.0000\n",
       "BambuTx8          8               8           5.0000\n",
       "BambuTx10        10              10          67.0000\n",
       "BambuTx11        11 11,42504,192214          40.1418"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "33308"
      ],
      "text/latex": [
       "33308"
      ],
      "text/markdown": [
       "33308"
      ],
      "text/plain": [
       "[1] 33308"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "30147"
      ],
      "text/latex": [
       "30147"
      ],
      "text/markdown": [
       "30147"
      ],
      "text/plain": [
       "[1] 30147"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "3.05357083623578"
      ],
      "text/latex": [
       "3.05357083623578"
      ],
      "text/markdown": [
       "3.05357083623578"
      ],
      "text/plain": [
       "[1] 3.053571"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#filter novel gene use exon number and width\n",
    "\n",
    "gtfinfo = qread(\"ref/bambu_extended_annotations_transinfo.qs\")\n",
    "gtfinfo = gtfinfo$transinfo\n",
    "gtfinfokn.tr = gtfinfo[gtfinfo$novelTranscript != TRUE,]\n",
    "gtfinfonv.tr = gtfinfo[gtfinfo$novelTranscript == TRUE,]\n",
    "gtfinfonv.trf = gtfinfonv.tr[!is.na(gtfinfonv.tr$NDR) & gtfinfonv.tr$NDR < 0.8,]\n",
    "gtfinfofil = rbind(gtfinfonv.trf,gtfinfokn.tr)\n",
    "head(gtfinfofil)\n",
    "gtfo.st.nv = gtfo[gtfo$gene_id %in% gtfinfonv.trf$GENEID,] %>% \n",
    "        group_by(gene_id) %>% summarise(width = max(width),\n",
    "                                        exon = max(as.numeric(exon_number[!is.na(exon_number)])))\n",
    "#head(gtfo.st)\n",
    "# gtfo.st$genetype = \"known\"\n",
    "# gtfo.st[substr(gtfo.st$gene_id,1,5) == \"Bambu\",]$genetype = \"novel\"\n",
    "# gtfo.st.nv = gtfo.st[gtfo.st$genetype == \"novel\",]\n",
    "nrow(gtfo.st.nv)\n",
    "gtfo.st.nv = gtfo.st.nv[gtfo.st.nv$width > 1000 & gtfo.st.nv$exon > 1,]\n",
    "gtfinfofil = gtfinfofil[gtfinfofil$novelGene == FALSE | gtfinfofil$GENEID %in% gtfo.st.nv$gene_id,]\n",
    "nrow(gtfo.st.nv)\n",
    "mean(gtfo.st.nv$exon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "9b0f5c62-fc03-4ec9-bc03-8333fae70c92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "401839"
      ],
      "text/latex": [
       "401839"
      ],
      "text/markdown": [
       "401839"
      ],
      "text/plain": [
       "[1] 401839"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>transid</th><th scope=col>fulllength</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>BambuTx1     </td><td>599</td></tr>\n",
       "\t<tr><td>BambuTx10    </td><td>203</td></tr>\n",
       "\t<tr><td>BambuTx100   </td><td>  5</td></tr>\n",
       "\t<tr><td>BambuTx1000  </td><td>413</td></tr>\n",
       "\t<tr><td>BambuTx10000 </td><td> 45</td></tr>\n",
       "\t<tr><td>BambuTx100000</td><td> 43</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 2\n",
       "\\begin{tabular}{ll}\n",
       " transid & fulllength\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t BambuTx1      & 599\\\\\n",
       "\t BambuTx10     & 203\\\\\n",
       "\t BambuTx100    &   5\\\\\n",
       "\t BambuTx1000   & 413\\\\\n",
       "\t BambuTx10000  &  45\\\\\n",
       "\t BambuTx100000 &  43\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 2\n",
       "\n",
       "| transid &lt;chr&gt; | fulllength &lt;int&gt; |\n",
       "|---|---|\n",
       "| BambuTx1      | 599 |\n",
       "| BambuTx10     | 203 |\n",
       "| BambuTx100    |   5 |\n",
       "| BambuTx1000   | 413 |\n",
       "| BambuTx10000  |  45 |\n",
       "| BambuTx100000 |  43 |\n",
       "\n"
      ],
      "text/plain": [
       "  transid       fulllength\n",
       "1 BambuTx1      599       \n",
       "2 BambuTx10     203       \n",
       "3 BambuTx100      5       \n",
       "4 BambuTx1000   413       \n",
       "5 BambuTx10000   45       \n",
       "6 BambuTx100000  43       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "71455"
      ],
      "text/latex": [
       "71455"
      ],
      "text/markdown": [
       "71455"
      ],
      "text/plain": [
       "[1] 71455"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "31671"
      ],
      "text/latex": [
       "31671"
      ],
      "text/markdown": [
       "31671"
      ],
      "text/plain": [
       "[1] 31671"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "39784"
      ],
      "text/latex": [
       "39784"
      ],
      "text/markdown": [
       "39784"
      ],
      "text/plain": [
       "[1] 39784"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "49606"
      ],
      "text/latex": [
       "49606"
      ],
      "text/markdown": [
       "49606"
      ],
      "text/plain": [
       "[1] 49606"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fgtf$fulllength = as.integer(fgtf$fulllength)\n",
    "fgtf.st = fgtf %>% group_by(transid) %>% summarise(fulllength = sum(fulllength))\n",
    "nrow(fgtf.st)\n",
    "head(fgtf.st)\n",
    "fgtf.st = fgtf.st[((substr(fgtf.st$transid,1,5) == \"Bambu\") & (fgtf.st$fulllength > 30)) |\n",
    "                  ((substr(fgtf.st$transid,1,5) != \"Bambu\") & (fgtf.st$fulllength > 0)),]\n",
    "\n",
    "nrow(fgtf.st)\n",
    "nrow(fgtf.st[substr(fgtf.st$transid,1,5) != \"Bambu\",])\n",
    "nrow(fgtf.st[substr(fgtf.st$transid,1,5) == \"Bambu\",])\n",
    "gtfdfant.fil = merge(gtfdfant,fgtf.st,by.x = \"transcript_id\",by.y = \"transid\")\n",
    "nrow(gtfdfant.fil)\n",
    "#gtfdfant.fil = gtfdfant.fil[gtfdfant.fil$gene_id %in% gtfo.st.nv$gene_id,]\n",
    "# gtfdfant.fil = gtfdfant.fil[gtfdfant.fil$transcript_id %in% gtfinfofil$TXNAME,]\n",
    "# nrow(gtfdfant.fil)\n",
    "#length(unique(gtfo.st.nv[gtfo.st.nv$gene_id %in% gtfdfant.fil$gene_id,]$gene_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "495a20ce-b230-43e6-a109-1bfebbf53463",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "6459"
      ],
      "text/latex": [
       "6459"
      ],
      "text/markdown": [
       "6459"
      ],
      "text/plain": [
       "[1] 6459"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "31671"
      ],
      "text/latex": [
       "31671"
      ],
      "text/markdown": [
       "31671"
      ],
      "text/plain": [
       "[1] 31671"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "      1       7      60    5683     433 3852656 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtfdfant.kn = gtfdfant.fil[substr(gtfdfant.fil$gene_id,1,5) != \"Bambu\",]\n",
    "length(unique(gtfdfant.kn[substr(gtfdfant.kn$transcript_id,1,5) == \"Bambu\",]$transcript_id))\n",
    "length(unique(gtfdfant.kn[substr(gtfdfant.kn$transcript_id,1,5) != \"Bambu\",]$transcript_id))\n",
    "summary(gtfdfant.kn$fulllength)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "2602a3f0-26e6-404f-9f88-7394b97a4b19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>trantype</th><th scope=col>transnew</th><th scope=col>fulllength</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>7957</th><td>BambuTx20710  </td><td>DDRGK1</td><td>novel</td><td>DDRGK1-WeiLab-1</td><td>35354</td></tr>\n",
       "\t<tr><th scope=row>28097</th><td>XM_005568473.2</td><td>DDRGK1</td><td>known</td><td>XM_005568473.2 </td><td>36851</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & transcript\\_id & gene\\_id & trantype & transnew & fulllength\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t7957 & BambuTx20710   & DDRGK1 & novel & DDRGK1-WeiLab-1 & 35354\\\\\n",
       "\t28097 & XM\\_005568473.2 & DDRGK1 & known & XM\\_005568473.2  & 36851\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 5\n",
       "\n",
       "| <!--/--> | transcript_id &lt;chr&gt; | gene_id &lt;chr&gt; | trantype &lt;chr&gt; | transnew &lt;chr&gt; | fulllength &lt;int&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 7957 | BambuTx20710   | DDRGK1 | novel | DDRGK1-WeiLab-1 | 35354 |\n",
       "| 28097 | XM_005568473.2 | DDRGK1 | known | XM_005568473.2  | 36851 |\n",
       "\n"
      ],
      "text/plain": [
       "      transcript_id  gene_id trantype transnew        fulllength\n",
       "7957  BambuTx20710   DDRGK1  novel    DDRGK1-WeiLab-1 35354     \n",
       "28097 XM_005568473.2 DDRGK1  known    XM_005568473.2  36851     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtfdfant.kn[gtfdfant.kn$gene_id == \"DDRGK1\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "17017bb5-3dcb-4aa7-89ba-7851d951183d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mReturning more (or less) than 1 row per `summarise()` group was deprecated in\n",
      "dplyr 1.1.0.\n",
      "\u001b[36mℹ\u001b[39m Please use `reframe()` instead.\n",
      "\u001b[36mℹ\u001b[39m When switching from `summarise()` to `reframe()`, remember that `reframe()`\n",
      "  always returns an ungrouped data frame and adjust accordingly.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene_id'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "8324"
      ],
      "text/latex": [
       "8324"
      ],
      "text/markdown": [
       "8324"
      ],
      "text/plain": [
       "[1] 8324"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# #rename\n",
    "# gtfo = rtracklayer::import(\"../ref/bambu_extended_annotations.gtf\")\n",
    "# gtfo = as.data.frame(gtfo)\n",
    "# #head(gtfo)\n",
    "# gtfo.nv = gtfo[(substr(gtfo$gene_id,1,5) != \"Bambu\"),]\n",
    "\n",
    "# gtfo.nv = gtfo.nv[(substr(gtfo.nv$transcript_id,1,5) == \"Bambu\") | grepl(x = gtfo.nv$transcript_id,pattern = \"WeiLab\"),]\n",
    "# gtfo.nv = unique(gtfo.nv[,c(\"gene_id\",\"transcript_id\")])\n",
    "# gtfo.nv = gtfo.nv %>% group_by(gene_id) %>% summarise(transcript_id = transcript_id,\n",
    "#                                                       transrname = paste0(gene_id,\"-FS-\",1:n()))\n",
    "# nrow(gtfo.nv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "53e3dcfc-7b57-4531-b5be-8e848fe5c468",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "38130"
      ],
      "text/latex": [
       "38130"
      ],
      "text/markdown": [
       "38130"
      ],
      "text/plain": [
       "[1] 38130"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "Length  Class   Mode \n",
       "     0   NULL   NULL "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "38130"
      ],
      "text/latex": [
       "38130"
      ],
      "text/markdown": [
       "38130"
      ],
      "text/plain": [
       "[1] 38130"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(gtfdfant.kn)\n",
    "summary(gtfdfant.kn$fulllength)\n",
    "nrow(gtfdfant.kn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "2cea6e20-a77b-4a1c-8821-e6b6dbd647a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>trantype</th><th scope=col>transnew</th><th scope=col>fulllength</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>BambuTx1     </td><td>PGBD2       </td><td>novel</td><td>PGBD2-WeiLab-1       </td><td>599</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>BambuTx10    </td><td>ZNF669      </td><td>novel</td><td>ZNF669-WeiLab-1      </td><td>203</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>BambuTx1000  </td><td>USP21       </td><td>novel</td><td>USP21-WeiLab-2       </td><td>413</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>BambuTx100007</td><td>LOC102142787</td><td>novel</td><td>LOC102142787-WeiLab-1</td><td>105</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>BambuTx100045</td><td>EXOC3L4     </td><td>novel</td><td>EXOC3L4-WeiLab-1     </td><td> 72</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>BambuTx100047</td><td>EXOC3L4     </td><td>novel</td><td>EXOC3L4-WeiLab-2     </td><td> 37</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & transcript\\_id & gene\\_id & trantype & transnew & fulllength\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & BambuTx1      & PGBD2        & novel & PGBD2-WeiLab-1        & 599\\\\\n",
       "\t2 & BambuTx10     & ZNF669       & novel & ZNF669-WeiLab-1       & 203\\\\\n",
       "\t3 & BambuTx1000   & USP21        & novel & USP21-WeiLab-2        & 413\\\\\n",
       "\t6 & BambuTx100007 & LOC102142787 & novel & LOC102142787-WeiLab-1 & 105\\\\\n",
       "\t11 & BambuTx100045 & EXOC3L4      & novel & EXOC3L4-WeiLab-1      &  72\\\\\n",
       "\t12 & BambuTx100047 & EXOC3L4      & novel & EXOC3L4-WeiLab-2      &  37\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | transcript_id &lt;chr&gt; | gene_id &lt;chr&gt; | trantype &lt;chr&gt; | transnew &lt;chr&gt; | fulllength &lt;int&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | BambuTx1      | PGBD2        | novel | PGBD2-WeiLab-1        | 599 |\n",
       "| 2 | BambuTx10     | ZNF669       | novel | ZNF669-WeiLab-1       | 203 |\n",
       "| 3 | BambuTx1000   | USP21        | novel | USP21-WeiLab-2        | 413 |\n",
       "| 6 | BambuTx100007 | LOC102142787 | novel | LOC102142787-WeiLab-1 | 105 |\n",
       "| 11 | BambuTx100045 | EXOC3L4      | novel | EXOC3L4-WeiLab-1      |  72 |\n",
       "| 12 | BambuTx100047 | EXOC3L4      | novel | EXOC3L4-WeiLab-2      |  37 |\n",
       "\n"
      ],
      "text/plain": [
       "   transcript_id gene_id      trantype transnew              fulllength\n",
       "1  BambuTx1      PGBD2        novel    PGBD2-WeiLab-1        599       \n",
       "2  BambuTx10     ZNF669       novel    ZNF669-WeiLab-1       203       \n",
       "3  BambuTx1000   USP21        novel    USP21-WeiLab-2        413       \n",
       "6  BambuTx100007 LOC102142787 novel    LOC102142787-WeiLab-1 105       \n",
       "11 BambuTx100045 EXOC3L4      novel    EXOC3L4-WeiLab-1       72       \n",
       "12 BambuTx100047 EXOC3L4      novel    EXOC3L4-WeiLab-2       37       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gtfdfant.kn)\n",
    "gtfdfant.kn$transnew2 = gsub(gtfdfant.kn$transnew,pattern = \"WeiLab\",replacement = \"FS\")\n",
    "write.csv(gtfdfant.kn,file = \"ref/gtf_filter_final_with_newname_25_10_09.csv\",row.names = F,quote = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "a45ffd25-8fb7-44d6-91c1-18a9d7f005c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'XM_005568473.2'</li><li>'DDRGK1-WeiLab-1'</li><li>NA</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'XM\\_005568473.2'\n",
       "\\item 'DDRGK1-WeiLab-1'\n",
       "\\item NA\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'XM_005568473.2'\n",
       "2. 'DDRGK1-WeiLab-1'\n",
       "3. NA\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"XM_005568473.2\"  \"DDRGK1-WeiLab-1\" NA               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fsraw = qread(\"data/reads_full_anotation/read_total_annotation_nobackground_24_09_04.qs\")\n",
    "unique(fsraw[fsraw$gene_id == \"DDRGK1\",]$transcript_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "3a334d52-03b2-456a-8056-4d18dbe51011",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "219542188"
      ],
      "text/latex": [
       "219542188"
      ],
      "text/markdown": [
       "219542188"
      ],
      "text/plain": [
       "[1] 219542188"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "123495999"
      ],
      "text/latex": [
       "123495999"
      ],
      "text/markdown": [
       "123495999"
      ],
      "text/plain": [
       "[1] 123495999"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "159948291"
      ],
      "text/latex": [
       "159948291"
      ],
      "text/markdown": [
       "159948291"
      ],
      "text/plain": [
       "[1] 159948291"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "105872017"
      ],
      "text/latex": [
       "105872017"
      ],
      "text/markdown": [
       "105872017"
      ],
      "text/plain": [
       "[1] 105872017"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'XM_005568473.2'</li><li>'DDRGK1-WeiLab-1'</li><li>NA</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'XM\\_005568473.2'\n",
       "\\item 'DDRGK1-WeiLab-1'\n",
       "\\item NA\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'XM_005568473.2'\n",
       "2. 'DDRGK1-WeiLab-1'\n",
       "3. NA\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"XM_005568473.2\"  \"DDRGK1-WeiLab-1\" NA               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(fsraw)\n",
    "length(unique(fsraw$readid))\n",
    "fsraw.fil = fsraw[fsraw$transcript_id %in% c(gtfdfant.kn$transnew),]\n",
    "nrow(fsraw.fil)\n",
    "length(unique(fsraw.fil$readid))\n",
    "unique(fsraw.fil[fsraw.fil$gene_id == \"DDRGK1\",]$transcript_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "1ab73944-7f7d-4bed-b7f1-dad0b6efa854",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "76543529"
      ],
      "text/latex": [
       "76543529"
      ],
      "text/markdown": [
       "76543529"
      ],
      "text/plain": [
       "[1] 76543529"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "64830100"
      ],
      "text/latex": [
       "64830100"
      ],
      "text/markdown": [
       "64830100"
      ],
      "text/plain": [
       "[1] 64830100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "qsave(fsraw.fil,file = \"data/reads_full_anotation/main_data_read_total_annotation_fil_25_10_09.qs\")\n",
    "fsraw.fil = fsraw.fil[!is.na(fsraw.fil$transcript_id),]\n",
    "fsraw.cor = fsraw.fil %>% group_by(readid) %>% filter(length(unique(transcript_id)) == 1)\n",
    "nrow(fsraw.cor)\n",
    "length(unique(fsraw.cor$readid))\n",
    "qsave(fsraw.cor,\"data/reads_full_anotation/main_data_read_total_annotation_fil_delete_multiread_25_10_09.qs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "95d9b85d-f3ee-4c7e-ab6c-69c4981be9d2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 6 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>readid</th><th scope=col>origin_name</th><th scope=col>cell_id</th><th scope=col>cell_type_id</th><th scope=col>subclass</th><th scope=col>class</th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>x</th><th scope=col>y</th><th scope=col>layer</th><th scope=col>region</th><th scope=col>lobe</th><th scope=col>subcortical</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>9eda4dea-9dc0-434f-871e-9448b11daf53-12;st_ed:A:6349-7977;AD:A:adapterC.rc-adapterM;AP:A:RC-Wrong_21-45:-         </td><td>L-1|2-l1</td><td>NA</td><td>NA</td><td></td><td></td><td>XM_005543577.2</td><td>RPS8     </td><td>10000</td><td>26457</td><td>l1</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "\t<tr><td>8e9fe2ef-d7b3-459a-94a2-f2638c24f2e0-3;st_ed:A:709-1447;AD:A:adapterE-adapterF;AP:A:Correct_695-719:+             </td><td>L-1|2-l1</td><td>NA</td><td>NA</td><td></td><td></td><td>XM_005594282.1</td><td>CXHXorf57</td><td>10000</td><td>26495</td><td>l1</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "\t<tr><td>cddeed9c-26d3-4e0f-abb2-709ef8d310c6-4;st_ed:A:803-1394;AD:A:adapterB-adapterC;AP:A:Correct_548-572:+             </td><td>L-1|2-l2</td><td>NA</td><td>NA</td><td></td><td></td><td>NM_001285041.1</td><td>CALM2    </td><td>10000</td><td>26590</td><td>l2</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "\t<tr><td>03977eea-b45b-42ad-b59d-953ddeb99a39-3;st_ed:A:91-914;AD:A:adapterP.rc-adapterN.rc;AP:A:Jump|AD:A:no-split_21-47:-</td><td>L-1|2-l2</td><td>NA</td><td>NA</td><td></td><td></td><td>XM_015437862.1</td><td>WBP2     </td><td>10000</td><td>26612</td><td>l2</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "\t<tr><td>19839030-fc62-41df-9c6f-170ec6b9fb74-15;st_ed:A:6994-7454;AD:A:adapterO-adapterP;AP:A:Correct_417-441:+           </td><td>L-1|2-l2</td><td>NA</td><td>NA</td><td></td><td></td><td>XM_015437862.1</td><td>WBP2     </td><td>10000</td><td>26612</td><td>l2</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "\t<tr><td>5b2f6398-a7f9-409e-84d3-b492e5bd2168-4;st_ed:A:1607-2156;AD:A:adapterM.rc-adapterL.rc;AP:A:Correct_21-45:-        </td><td>L-1|2-l2</td><td>NA</td><td>NA</td><td></td><td></td><td>XM_005548045.2</td><td>NDUFB4   </td><td>10000</td><td>26652</td><td>l2</td><td>1|2</td><td>somatosensory</td><td>Cortex</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 6 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " readid & origin\\_name & cell\\_id & cell\\_type\\_id & subclass & class & transcript\\_id & gene\\_id & x & y & layer & region & lobe & subcortical\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 9eda4dea-9dc0-434f-871e-9448b11daf53-12;st\\_ed:A:6349-7977;AD:A:adapterC.rc-adapterM;AP:A:RC-Wrong\\_21-45:-          & L-1\\textbar{}2-l1 & NA & NA &  &  & XM\\_005543577.2 & RPS8      & 10000 & 26457 & l1 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\t 8e9fe2ef-d7b3-459a-94a2-f2638c24f2e0-3;st\\_ed:A:709-1447;AD:A:adapterE-adapterF;AP:A:Correct\\_695-719:+              & L-1\\textbar{}2-l1 & NA & NA &  &  & XM\\_005594282.1 & CXHXorf57 & 10000 & 26495 & l1 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\t cddeed9c-26d3-4e0f-abb2-709ef8d310c6-4;st\\_ed:A:803-1394;AD:A:adapterB-adapterC;AP:A:Correct\\_548-572:+              & L-1\\textbar{}2-l2 & NA & NA &  &  & NM\\_001285041.1 & CALM2     & 10000 & 26590 & l2 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\t 03977eea-b45b-42ad-b59d-953ddeb99a39-3;st\\_ed:A:91-914;AD:A:adapterP.rc-adapterN.rc;AP:A:Jump\\textbar{}AD:A:no-split\\_21-47:- & L-1\\textbar{}2-l2 & NA & NA &  &  & XM\\_015437862.1 & WBP2      & 10000 & 26612 & l2 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\t 19839030-fc62-41df-9c6f-170ec6b9fb74-15;st\\_ed:A:6994-7454;AD:A:adapterO-adapterP;AP:A:Correct\\_417-441:+            & L-1\\textbar{}2-l2 & NA & NA &  &  & XM\\_015437862.1 & WBP2      & 10000 & 26612 & l2 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\t 5b2f6398-a7f9-409e-84d3-b492e5bd2168-4;st\\_ed:A:1607-2156;AD:A:adapterM.rc-adapterL.rc;AP:A:Correct\\_21-45:-         & L-1\\textbar{}2-l2 & NA & NA &  &  & XM\\_005548045.2 & NDUFB4    & 10000 & 26652 & l2 & 1\\textbar{}2 & somatosensory & Cortex\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 6 × 14\n",
       "\n",
       "| readid &lt;chr&gt; | origin_name &lt;chr&gt; | cell_id &lt;dbl&gt; | cell_type_id &lt;dbl&gt; | subclass &lt;chr&gt; | class &lt;chr&gt; | transcript_id &lt;chr&gt; | gene_id &lt;chr&gt; | x &lt;dbl&gt; | y &lt;dbl&gt; | layer &lt;chr&gt; | region &lt;chr&gt; | lobe &lt;chr&gt; | subcortical &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 9eda4dea-9dc0-434f-871e-9448b11daf53-12;st_ed:A:6349-7977;AD:A:adapterC.rc-adapterM;AP:A:RC-Wrong_21-45:-          | L-1|2-l1 | NA | NA | <!----> | <!----> | XM_005543577.2 | RPS8      | 10000 | 26457 | l1 | 1|2 | somatosensory | Cortex |\n",
       "| 8e9fe2ef-d7b3-459a-94a2-f2638c24f2e0-3;st_ed:A:709-1447;AD:A:adapterE-adapterF;AP:A:Correct_695-719:+              | L-1|2-l1 | NA | NA | <!----> | <!----> | XM_005594282.1 | CXHXorf57 | 10000 | 26495 | l1 | 1|2 | somatosensory | Cortex |\n",
       "| cddeed9c-26d3-4e0f-abb2-709ef8d310c6-4;st_ed:A:803-1394;AD:A:adapterB-adapterC;AP:A:Correct_548-572:+              | L-1|2-l2 | NA | NA | <!----> | <!----> | NM_001285041.1 | CALM2     | 10000 | 26590 | l2 | 1|2 | somatosensory | Cortex |\n",
       "| 03977eea-b45b-42ad-b59d-953ddeb99a39-3;st_ed:A:91-914;AD:A:adapterP.rc-adapterN.rc;AP:A:Jump|AD:A:no-split_21-47:- | L-1|2-l2 | NA | NA | <!----> | <!----> | XM_015437862.1 | WBP2      | 10000 | 26612 | l2 | 1|2 | somatosensory | Cortex |\n",
       "| 19839030-fc62-41df-9c6f-170ec6b9fb74-15;st_ed:A:6994-7454;AD:A:adapterO-adapterP;AP:A:Correct_417-441:+            | L-1|2-l2 | NA | NA | <!----> | <!----> | XM_015437862.1 | WBP2      | 10000 | 26612 | l2 | 1|2 | somatosensory | Cortex |\n",
       "| 5b2f6398-a7f9-409e-84d3-b492e5bd2168-4;st_ed:A:1607-2156;AD:A:adapterM.rc-adapterL.rc;AP:A:Correct_21-45:-         | L-1|2-l2 | NA | NA | <!----> | <!----> | XM_005548045.2 | NDUFB4    | 10000 | 26652 | l2 | 1|2 | somatosensory | Cortex |\n",
       "\n"
      ],
      "text/plain": [
       "  readid                                                                                                            \n",
       "1 9eda4dea-9dc0-434f-871e-9448b11daf53-12;st_ed:A:6349-7977;AD:A:adapterC.rc-adapterM;AP:A:RC-Wrong_21-45:-         \n",
       "2 8e9fe2ef-d7b3-459a-94a2-f2638c24f2e0-3;st_ed:A:709-1447;AD:A:adapterE-adapterF;AP:A:Correct_695-719:+             \n",
       "3 cddeed9c-26d3-4e0f-abb2-709ef8d310c6-4;st_ed:A:803-1394;AD:A:adapterB-adapterC;AP:A:Correct_548-572:+             \n",
       "4 03977eea-b45b-42ad-b59d-953ddeb99a39-3;st_ed:A:91-914;AD:A:adapterP.rc-adapterN.rc;AP:A:Jump|AD:A:no-split_21-47:-\n",
       "5 19839030-fc62-41df-9c6f-170ec6b9fb74-15;st_ed:A:6994-7454;AD:A:adapterO-adapterP;AP:A:Correct_417-441:+           \n",
       "6 5b2f6398-a7f9-409e-84d3-b492e5bd2168-4;st_ed:A:1607-2156;AD:A:adapterM.rc-adapterL.rc;AP:A:Correct_21-45:-        \n",
       "  origin_name cell_id cell_type_id subclass class transcript_id  gene_id  \n",
       "1 L-1|2-l1    NA      NA                          XM_005543577.2 RPS8     \n",
       "2 L-1|2-l1    NA      NA                          XM_005594282.1 CXHXorf57\n",
       "3 L-1|2-l2    NA      NA                          NM_001285041.1 CALM2    \n",
       "4 L-1|2-l2    NA      NA                          XM_015437862.1 WBP2     \n",
       "5 L-1|2-l2    NA      NA                          XM_015437862.1 WBP2     \n",
       "6 L-1|2-l2    NA      NA                          XM_005548045.2 NDUFB4   \n",
       "  x     y     layer region lobe          subcortical\n",
       "1 10000 26457 l1    1|2    somatosensory Cortex     \n",
       "2 10000 26495 l1    1|2    somatosensory Cortex     \n",
       "3 10000 26590 l2    1|2    somatosensory Cortex     \n",
       "4 10000 26612 l2    1|2    somatosensory Cortex     \n",
       "5 10000 26612 l2    1|2    somatosensory Cortex     \n",
       "6 10000 26652 l2    1|2    somatosensory Cortex     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fsraw.cor$transcript_id = gsub(fsraw.cor$transcript_id,pattern = \"WeiLab\",replacement = \"FS\")\n",
    "head(fsraw.cor)\n",
    "qsave(fsraw.cor,\"data/reads_full_anotation/main_data_read_total_annotation_fil_delete_multiread_25_10_09.qs\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58709723-4a50-4fd9-a628-4b3b28a5c15f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Isoform plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ae6bfb82-7cf4-45b3-a5b0-06ee0b597079",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>trantype</th><th scope=col>transnew</th><th scope=col>fulllength</th><th scope=col>transnew2</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>BambuTx1     </td><td>PGBD2       </td><td>novel</td><td>PGBD2-WeiLab-1       </td><td>599</td><td>PGBD2-FS-1       </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>BambuTx10    </td><td>ZNF669      </td><td>novel</td><td>ZNF669-WeiLab-1      </td><td>203</td><td>ZNF669-FS-1      </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>BambuTx1000  </td><td>USP21       </td><td>novel</td><td>USP21-WeiLab-2       </td><td>413</td><td>USP21-FS-2       </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>BambuTx100007</td><td>LOC102142787</td><td>novel</td><td>LOC102142787-WeiLab-1</td><td>105</td><td>LOC102142787-FS-1</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>BambuTx100045</td><td>EXOC3L4     </td><td>novel</td><td>EXOC3L4-WeiLab-1     </td><td> 72</td><td>EXOC3L4-FS-1     </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>BambuTx100047</td><td>EXOC3L4     </td><td>novel</td><td>EXOC3L4-WeiLab-2     </td><td> 37</td><td>EXOC3L4-FS-2     </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & transcript\\_id & gene\\_id & trantype & transnew & fulllength & transnew2\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & BambuTx1      & PGBD2        & novel & PGBD2-WeiLab-1        & 599 & PGBD2-FS-1       \\\\\n",
       "\t2 & BambuTx10     & ZNF669       & novel & ZNF669-WeiLab-1       & 203 & ZNF669-FS-1      \\\\\n",
       "\t3 & BambuTx1000   & USP21        & novel & USP21-WeiLab-2        & 413 & USP21-FS-2       \\\\\n",
       "\t4 & BambuTx100007 & LOC102142787 & novel & LOC102142787-WeiLab-1 & 105 & LOC102142787-FS-1\\\\\n",
       "\t5 & BambuTx100045 & EXOC3L4      & novel & EXOC3L4-WeiLab-1      &  72 & EXOC3L4-FS-1     \\\\\n",
       "\t6 & BambuTx100047 & EXOC3L4      & novel & EXOC3L4-WeiLab-2      &  37 & EXOC3L4-FS-2     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| <!--/--> | transcript_id &lt;chr&gt; | gene_id &lt;chr&gt; | trantype &lt;chr&gt; | transnew &lt;chr&gt; | fulllength &lt;int&gt; | transnew2 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | BambuTx1      | PGBD2        | novel | PGBD2-WeiLab-1        | 599 | PGBD2-FS-1        |\n",
       "| 2 | BambuTx10     | ZNF669       | novel | ZNF669-WeiLab-1       | 203 | ZNF669-FS-1       |\n",
       "| 3 | BambuTx1000   | USP21        | novel | USP21-WeiLab-2        | 413 | USP21-FS-2        |\n",
       "| 4 | BambuTx100007 | LOC102142787 | novel | LOC102142787-WeiLab-1 | 105 | LOC102142787-FS-1 |\n",
       "| 5 | BambuTx100045 | EXOC3L4      | novel | EXOC3L4-WeiLab-1      |  72 | EXOC3L4-FS-1      |\n",
       "| 6 | BambuTx100047 | EXOC3L4      | novel | EXOC3L4-WeiLab-2      |  37 | EXOC3L4-FS-2      |\n",
       "\n"
      ],
      "text/plain": [
       "  transcript_id gene_id      trantype transnew              fulllength\n",
       "1 BambuTx1      PGBD2        novel    PGBD2-WeiLab-1        599       \n",
       "2 BambuTx10     ZNF669       novel    ZNF669-WeiLab-1       203       \n",
       "3 BambuTx1000   USP21        novel    USP21-WeiLab-2        413       \n",
       "4 BambuTx100007 LOC102142787 novel    LOC102142787-WeiLab-1 105       \n",
       "5 BambuTx100045 EXOC3L4      novel    EXOC3L4-WeiLab-1       72       \n",
       "6 BambuTx100047 EXOC3L4      novel    EXOC3L4-WeiLab-2       37       \n",
       "  transnew2        \n",
       "1 PGBD2-FS-1       \n",
       "2 ZNF669-FS-1      \n",
       "3 USP21-FS-2       \n",
       "4 LOC102142787-FS-1\n",
       "5 EXOC3L4-FS-1     \n",
       "6 EXOC3L4-FS-2     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "DataFrame with 6 rows and 12 columns\n",
       "              TXNAME       GENEID       NDR novelGene novelTranscript\n",
       "         <character>  <character> <numeric> <logical>       <logical>\n",
       "BambuTx1    BambuTx1        PGBD2  0.548446     FALSE            TRUE\n",
       "BambuTx2    BambuTx2       ZNF692  0.901575     FALSE            TRUE\n",
       "BambuTx3    BambuTx3       ZNF692  0.832564     FALSE            TRUE\n",
       "BambuTx4    BambuTx4 LOC107130781  0.258162     FALSE            TRUE\n",
       "BambuTx5    BambuTx5      SH3BP5L  0.921423     FALSE            TRUE\n",
       "BambuTx6    BambuTx6        NLRP3  0.725336     FALSE            TRUE\n",
       "             txClassDescription readCount relReadCount relSubsetCount      txid\n",
       "                    <character> <integer>    <numeric>      <numeric> <integer>\n",
       "BambuTx1 newLastJunction:newL..       127   0.41368078       1.000000         1\n",
       "BambuTx2              newWithin         7   0.00328022       0.538462         2\n",
       "BambuTx3            newJunction        19   0.00890347       0.791667         3\n",
       "BambuTx4                 allNew       629   0.94020927       1.000000         4\n",
       "BambuTx5              newWithin         9   0.00571792       1.000000         5\n",
       "BambuTx6                 allNew        33   0.16019417       1.000000         6\n",
       "           eqClassById fullLengthCounts\n",
       "         <IntegerList>        <numeric>\n",
       "BambuTx1             1        129.00000\n",
       "BambuTx2      2,227827          5.00001\n",
       "BambuTx3      3,192203         19.00000\n",
       "BambuTx4             4        629.00000\n",
       "BambuTx5             5          9.00000\n",
       "BambuTx6             6         38.00000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtfdfant.kn = read.csv(\"ref/gtf_filter_final_with_newname_25_10_09.csv\")\n",
    "head(gtfdfant.kn)\n",
    "gtfinfo = qread(\"ref/bambu_extended_annotations_transinfo.qs\")\n",
    "gtfinfo = gtfinfo$transinfo\n",
    "head(gtfinfo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5144d2f0-6940-4af8-aab4-d431fc240d5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#stat gene pie plot\n",
    "# gtfkn = gtf[substr(gtf$transcript_id,1,5) != \"Bambu\",]\n",
    "# genet.st = data.frame(\"type\" = c(\"Known\",\"Novel\"), \"count\" = c(length(unique(gtfkn$gene_id)),length(unique(gtfinfonv$GENEID))))\n",
    "# genet.st <- genet.st %>% arrange(desc(type)) %>%\n",
    "#   mutate(prop = count / sum(genet.st$count) *100) %>%\n",
    "#   mutate(ypos = cumsum(prop)- 0.5*prop )\n",
    "# genet.st$proplabel = paste0(genet.st$count,\"(\", round(genet.st$prop,1),\"%)\")\n",
    "\n",
    "# p1.1 = ggplot(genet.st, aes(x=\"\", y=prop, fill=type)) +\n",
    "#   geom_bar(stat=\"identity\", width=1, color= NA) +\n",
    "#   coord_polar(\"y\", start=0) +\n",
    "#   theme_void() + \n",
    "#   theme(legend.position=\"none\") +\n",
    "#   geom_text(aes(y = ypos, label = proplabel), color = \"white\", size=6) +\n",
    "#   scale_fill_nejm()\n",
    "# p1.1\n",
    "# ggexport(p1.1,filename = \"result/figure1/novel_gene_stat.pdf\",width = 5,height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "00c50cfd-6949-461e-8d7b-da6eafec5d85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "401839"
      ],
      "text/latex": [
       "401839"
      ],
      "text/markdown": [
       "401839"
      ],
      "text/plain": [
       "[1] 401839"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "38130"
      ],
      "text/latex": [
       "38130"
      ],
      "text/markdown": [
       "38130"
      ],
      "text/plain": [
       "[1] 38130"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "38130"
      ],
      "text/latex": [
       "38130"
      ],
      "text/markdown": [
       "38130"
      ],
      "text/plain": [
       "[1] 38130"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(gtfinfo)\n",
    "nrow(gtfdfant.kn)\n",
    "gtfinfo = as.data.frame(gtfinfo)\n",
    "gtfinfo = gtfinfo[gtfinfo$TXNAME %in% gtfdfant.kn$transcript_id,]\n",
    "nrow(gtfinfo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "39bd7e38-0d7f-4e6d-815a-43877c45c14e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "                                                               allNew \n",
       "                                                                 3023 \n",
       "                                                           annotation \n",
       "                                                                31671 \n",
       "                                                     newFirstJunction \n",
       "                                                                  282 \n",
       "                                        newFirstJunction:newFirstExon \n",
       "                                                                  343 \n",
       "                                         newFirstJunction:newJunction \n",
       "                                                                  357 \n",
       "                            newFirstJunction:newJunction:newFirstExon \n",
       "                                                                  455 \n",
       "                                                          newJunction \n",
       "                                                                  560 \n",
       "                                                      newLastJunction \n",
       "                                                                  168 \n",
       "                         newLastJunction:newFirstJunction:newJunction \n",
       "                                                                   12 \n",
       "            newLastJunction:newFirstJunction:newJunction:newFirstExon \n",
       "                                                                   29 \n",
       "newLastJunction:newFirstJunction:newJunction:newFirstExon:newLastExon \n",
       "                                                                   35 \n",
       "             newLastJunction:newFirstJunction:newJunction:newLastExon \n",
       "                                                                   18 \n",
       "                                          newLastJunction:newJunction \n",
       "                                                                  228 \n",
       "                              newLastJunction:newJunction:newLastExon \n",
       "                                                                  321 \n",
       "                                          newLastJunction:newLastExon \n",
       "                                                                  262 \n",
       "                                                            newWithin \n",
       "                                                                  341 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "file saved to result/figure1/bambu_novel_isoform_stat_bar_25_10_13.pdf\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzdd6CWc//A8c+p0zrtQUtDRUkl2RmdkhYysx9bkc3z2KPsn71HVp6IskJGVMpo\neMyoUNpKQrvTOp3fH6FN0fx6vf7qXPf3uu7ruo3e53uNOyMvLy8AANjy5dvUOwAAwPoh7AAA\nEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewA\nABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEpG5qXcA2OJNOWj39b7Nir0/\nWu/b/Pva9h2xfjf4avO663eDG1TGU7es3w3mnXTZ+t0gYMYOACARwg4AIBHCDgAgEcIOACAR\nwg4AIBHCDgAgEcIOtkxTnmuffUyXbzb1bmxC017omJ2dfcjNQxesuHzhO5dnZ1/y+pyIiNGP\nH5+dffGrM5d/fcmMr3rde2WHYw5t3aL1ocecfN51j/cdNStv4+33uhrd5fjs7KbHPvL1khWX\nT+vRMbvFjR+s9ZhfzR3V55FOZx9/+IEtW7U96sSOVz3Y+6ufF2/QA/ibuhyfnX3xq3OWXzTz\n/c6HN21+8kMjFqxpJfjnEnawRZrZ5+knBg4ZM3tT78cmtGDSpwMHDnz16jNv+SJ3+eVLpn45\ncOCwKYsjIuaMHTpw4BeTF/32Wt6UPpfsVWPn9s9M2mb/E869+NzjW9XPN/T/Dt1hh9a3Dp61\nsQ9gLc0ZM3TgwAHPXdzxoQkrLF8w8dOB7438aa3HRMSMQTe1qFX3qAdGlNz7qLMuOv/kg3ct\n+c3DJ+5Ue7/L3v5hIxzJXzNm6MCBX0xe1p7zPrvxwIM6Dynxr6vPqFtoE+4XbKY8oBi2OJ/c\ndtDFzw0bmht53TtkDylesu0pe/Z9sk9es86vXtOkwPIDf3z+rKMf+Hq7M57psturx7Z/LueQ\ne144evrjD3Tv/+WE6YuzKu24/7/Obb9/1eVXmfVN76eeevWDL8dPzy1SrnrD5sec9q/9qhSI\nzVaZ/VvVeu/ms+47/sMLamX86ejcEXccethtUw974fNuR1T69bfaNu1OPvNfnfbf+9LjO+09\n8s69N9NSqNWyVbx95XlPH9HrhAp/eczEricceOWnjR/+X68O2//6z7T14f8689T7D97j3BMu\najyqe9uSG2Tn16eF3zx4eMurvtnjgYE9T6rp7y9YDTN2sMWp0fayTm1rLYwotU/7Tp06XXJw\nq0blvxr41m33r3i+KqY8/2CXAQPn1NyjUiyY+OnAgW++eO0he1/wYdYeR3a44KxDtx3f7dzm\nO7W6b+Rv01253z55ZP36ba8fkH/3ozuee3KbmlOePqtJ3b0ufXfGxj/EtZVTtcNDF9UYes3Z\nj03+88Fze11/40eZh9/20O9Vt1TRXa5+7ZuJX96xuVZdROQ2+M/9pxZ/5d8Xv7bmfxh/Mib3\nw1uvfn3Oftc++nvVLVVwu3Oe+2Lc6G5bQNUtmfDM8c3PGVrn1rd7dfx9su6nF8/Obn7Th/O/\nffXmjkcf3Oqgo06/+r/DZv52aj1vxhfP3XzOcW1bHtDykOPOvvnZz6fnRcRndxySffSDw3/b\n7vQXz2ua3ew/b879bcEX9x6WfUyXb7579ITsox74atawZ68548gDW7c97qwbXh41byMeMPwV\nwg62OKVr75O949YZEYWr7pKdnb33duVbdDh525jz6uM9f1pu2OQez763JLPJaSfViihQoEDE\nwg/fK3xTv/9eeVq7A1sd1v6ON7qfXnlG/8sv7TE9IiLv69uPbf/ipNqX9Xn3oYtPOOTgo9pf\n//LQp4/O+uTW4y99e+4a9mSTW7Qoo9E1D3Ys8/ZlF6xw7Kuz5IPeb84o1PKoQ0qs8lL+rapt\nU/TPZ/w2nUWLCrX8v7uPXNz93Cv6r6ks/mTM5717T4p9jjqq0qovla5SrUT+9bq/G8LUV9s3\nP/mtite80fs/O2ctW5z341cD+73c6agTu+fudVzHjodV+/rBk/Y9ufsvEREz+5y9+26n/Xd6\nvSM7Xnh2u/oznjpt9z3OfXtW1K6Qb0jPp9749bT13D7dugwY+O7DPd799ZT9iFce6/V54ao1\nF076bODbz154eMe+pZqddHb7VkUH33J4k/P75WzsI4d1IuwgARl7tj+jQSzs83i3Sb8vm9jj\n2UFLsg48/bhl5+VKHdH+qLK//1Sk2TFtt4q5fXq/uygiBnd58NPF+Vtc8O+dl81blT3ywhOr\nxJSnH3t9/kY5jHWXl5cXWdk33HNCwecv/E+fP77icNro0TNj2zp1Cm6kfVuf8vLyouwRd/9f\n658f6Xj9/xb+hTGLR48eH6Xq1Cm/4Xd2Q5gx4N8tj3589E5XPt1prxXDPCMjI+KLuU17PHfV\nvw5p3fa02+4+o+as1197b0nEiHsufGjs3rf2feaKEw9p0/aEy5958/pdRj9w6UPfZbVotW/G\nx+/0nR4Rsaj/m/1zGzfefc7bfQZFRMTkt976ssgBrZtkZmRkxMxhWae98vhFxx7c5rCOD91y\nTMkpr732ycY/fFgHwg6SsP2pHZoWWDLo8a5f/7pg7HPPfZRX8sjTjljur8Edd9pphYmZqlWr\nRiycOPHHiB8++mhCxHZ77llm+QEZu+yxe2bMGz587AY/gL+j5CG3337wvKfOueaDPwrQBQsW\nRBQsuCV23a8qn3pfpz3H33HWbSOXrPOYhQsW5G25R/9h5+O6Fj1g77Kf3HzGPSMWrfp6nVYt\nq/325ypVqsSiKVN+jh/69x8ZDdq0qfz7sG1bttw+Pn/vvVlbtWzZaMn77/SfH5H3/pt9Zu9y\n2OUt6kzp02dYREx/662P8me3PqDI0nV2bdnyt1+GMqtUqRhTp0xZ86cPmwFhB2kof0KHQ4rG\n8CceH5oXETHm2Wc/iorHnda6yHJjSpcuvcI6pUqViohp06ZF/PzzzxEx+ekzslfQ/IZBETFp\n0qTYvJU//t4bsr+/76xbPl/zkzu2qlAhX0yevBYX4222Mmqe/9BldT6/seODay7tNYzJqlCh\nePwyefLmOvf6x7Y+5KmPB7z6+vPnVBx0cduz+vy88utlyiz7hSR//vwRubm5MXXq1Iitt956\nuXFbbbVVxI8/TovqrVrVnt/3nQ/y4tM335q6Q4sWBx5wQPlRffqMjZx33no/d4/WrcqsedPC\njs2au4ogESUO63DMVi883q3re7fs0WT0s89+Fttddtp+K/wnnpu7wnNBIicnJyKKFi0aka9w\n4Yio1OKCTu22jpVlbF1vQ+75elH9rAev6rrTdWfddfy7lVf/C2uRxo13jtf69B66eP89ttj/\n82U2uOyh857Z96pzuh35aL41/GK++jG7NW6c+eiA3n3mHn1I0Y21t+vNdtkHVC8QkX3XK3d9\nufu5Rx+545C3L6zzZ3drFy5c+Nd52t8PeMGCBRFFihSJqNmyZeUbe7/z+dfF3xxboUWLBhl7\nNG9W9P4+fabU+1/f+Ttd1Ho1lyLClsGMHaSi4P4dTqkZU5/v3n/h5926fRUNTzttlxVHrDxd\nNW3atIioUKFCxDY1axaMmFu0evZqNKlbbuMdxl+Vr86/H7p42486n/3Uj4VX/z+26ied1arI\n2IevfnyV+cd5Q+/oePGTn2zGt/8uU2jvzvefXOyNf1/0ek7hdRlTut1Zx5T76dlr7x6Ru9Lo\nxSMfO+/ce97ffJ9kt0xm7XNe6H5qqff+c/A5b/3yZ4Nr1KuXFSOHD1/28OnFw4d/G2V23LFi\nROzVqmXJMf27PvTm8OLND9gzokB28/3yfdjv9tfemV6zdetaG/IoYIMSdrBFyrea6ZqM3dq3\n3znj5xeevv7R/36T2eS0E1f+2+mrfv2mLvfj5AEDRkXU3m234hGFmh90QKGY2Oulj1f4Doaf\nXrnilEvuG/D9BjiC9a/grlc90GGrdy6/5b013OJZ8dSHHzqkdL8LW7d/+stZv55Py5v19cuX\ntjzo0pe/K1Bu83/gR0REFG9x653tFj979X0j1mlM0UPu7Hpq9RGdDj7mwaHTfou7eePevrlt\ni7O6fZlbpuzqt7S5KXfgg72u321yl6OPvHfkH39jRoFW7U+pNrXbDfeMWHqT8MxPb76555za\nZ5yenRERmdktmxX+vEuXjwo2bZGdGRElmu+/2/zXH+o6cevWrRpt+OOADUXYwRapQoUKEdPG\nj1/x0Qs1T+mwf8Ffut/SZWKhNqcdX3GldTKKDL3rnGe+W3qV1ZIfXr/2gSFRdP8LT68XEVHu\nuBsu27nQqPtOv+D175c2z5Lpn9x9Uoebn35pbEbp2DJkNbvp3mMLTJ26msvrIyIio9pJL3z0\n5tX1Prl4t/IV6+zWeK+da1estNuFg2te0u/zl06otjk/8GQF5Y66++YW86ZOXccxWx342OCB\ndzadclOzbcrXarRX413rblNhx5NfLn7aS5+/c9GOm/FzqFdUqOEVLz3erui7Fx187qpX260w\ncN9bX3uk1aRrd6lWe9c9Gm5bvfE9c4956tXOuyzt/iIHtNo3b/78xY1bHLD0VG3V5s1r5eTk\nlGjZuvEW828CrCojL28z/o5EYA2WjLg3e98L3l9UdZcdy+7QoXe3k3+tuNkvHFOpXY85JY99\nZUr3tstunPi6U70dOo9o98Cb271w5qOji1avlH/yl8OmlW528VPP3rT/7xfV/TLoznPPvfn5\nEZm1GlQvOmf8yG/mVW55zt2PXNd6mz/+FXDKQbuv9wOs2PujPxmxYNKng78rUq/JDiueJ57+\n7YdfTF5UpvY+DSpmxpyxQz8eX3j7vXeqtGK2LJk7ZdSoCT8vKFS2YrVqVUoXXtu/yNv2/YNZ\nsr/i1eZ1/2zInDFDP55adue9aq0woZj34/D3RkzLKF9vvx3Krd2Y5cyfNvqbcT/m5C9doUq1\nqltlrfUv+BlP3bK2Q9dO3kmXrd8NAsIOtlS5syZ8NWJiTsHSVbevU6nYb385D7um9k7XLzj/\n/dF377PcHQJfd6q3Q+fhhz63+OWj882b8vXIcdNzi1WuXbdayVVPWi6Z8/3X30z4ZUH+kpW2\nq1O99NrM42yasNsUNkXYbUaEHWz+tth7w+AfL3+JqjvtWXXFZXNevf3hbwvsc//F+6zuv+28\nvLyIjKyKO+yy8lna5eQrVrnuLpXX/DoAmy9hB8nI+ebxkzp0m73nzQ92qLKp9wWATcHNE5CA\nSU+etPfudao26PBh3StfeemS+n5jA/hn8v9/SMDWrS+5rdaMzK2226nO1oVWN6DqyU++mz23\nXN3N/6veAfgbhB0koGCFHRtX+KMBWdV3y66+kXYGgE3GqVgAgER43AkAQCLM2AEAJELYAQAk\nQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEA\nJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgB\nACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkInNT78AWZvHixU888URENG/evEaNGpt6dwAA\nlsnIy8vb1PuwJcnJycnKyoqInj17tmvXblPvDgDAMk7FAgAkQtgBACRC2AEAJELYAQAkQtgB\nACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELY\nAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC\n2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAk\nQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEA\nJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACQic1PvAKuX8dQtm3oXAIB1k3fSZZt2B8zYAQAk\nQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEA\nJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgB\nACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELY\nAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC\n2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAk\nQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEA\nJELYAQAkQtgBACRC2AEAJCJzg7/DW6cXa/343EO6Lep1wgpvNvPDK5q3uvnjQvvf8c5rF+1c\nZIPvB+ti6jfx+ofx9Q/xy7woVDyq14z9s6NRuZWHTf40HnwlJi2MU66NJoVXfjV3ZrzdNz74\nJn7MiSIlYvt6ccj+UaVgRESMiXMejTlrePd9TorT6yz7cd6kePzZ+OSX2Otf0aHu+jlAAEjR\nhg+71Zo5+KoWrW/+uHCLu/v2On8nVbd5+eatuGNgVNozzjw8tikSP42LZ1+Ke7+Mo8+K1hV+\nG7QwBr4S3b+O0gUiFq5mI7k/xd0Px/gKcerpUbd0/DAyujwfN42LK9pHlfwRNeL+m1dZJy9e\nezBenB171Vi2ZPSH8VCfKFBywxwqACRlU5yKnTXkmlatbvyocKv7+r+i6jY3eVPiyYERNeL8\nQ6JGqShYKCrVjo5touDCePW9yF06aHE8d1/0nBZnnhv7llr9dvq9GF9mxlknRsOto2CBqNog\n2u8bORPi5S/X+NZTB8Wrk2LPg2PHgr8u+eSFuGVg7HdKnG6iDgD+3EYPu9kfdWrV8vqhhQ98\n4N1e59RfdvZu2kNNMjLyt+sx66tuFx+y27ZbFS9UMGurWk3PfuKzWcutnffTR49cdPie25Uv\nUahAoeLlauxyYMe73p2cGxHx82Ot8mdkNH3gh2Wj5zxzcMGMjIzql3283CYGXVQ5I6Pyee8t\nef7ojIyMpg9MGNPr2qP33r58ycIFC5eutteJ93z445IN/BlsKLOHxMmXxwPDYtJncc8Dce61\ncdo1ccmj8f7kddjIrKmRuVU02DlKL7ewSMUoF5EzM+Yv/Xlx5N8uOneInddQdfF9vDMuquwe\nOxRctqzqfnHTpXF2wzWsMiO6vh0F6sRxOy5bNqdIXHReHFIjMtbhCADgH2vjnoqd87/rW7fo\nPCTr4If7v9B++b/zIwoXLhyx5Ku72hxdqs2tTw56ukbhnz9+5IwjLj+t6dSs7144pmxExM9v\nnL7HoU/M2e+6R9/qsW/t0jnjP+pxxSn/bv7qkO4f9zi6QqtWu0Wfwe8OmH/2MUt7cfHAvgMX\nFStWbHy/ft/FrjWXvs/Ifv0mR8nT2jTOt+jxghHjuh3dtsB+197Xt0udUnO+ev78dqdf0HxM\n/pEfnFN9o34w60eBAhERkz6IB4vE0UdGhzIxZ1I8+Uw8/mgU/E/skRXfvRHXv7/G1fM1iCeO\njZIN44ZV2mv29/FzRLnKUXTpz4WjXds/2pMfvo1pEc23XXFp4ai0ynV4v/v4rRi5OI45KEos\nt7DJQX/0LgDAijZi2M355MbWLa4ZlHXoo+/2PK12gZVezcjIiIivJ9Tv++4V+xeJiCi+36U3\nnvJw39te6fHW/GOOLxy573U+84kxRY7q+cLVbUtFRJTeofmFT9//abVDnz630xmHP9y8det6\nFw4dOGBI3jHZGRERQ/v2m139hLN3fPSh/v1nXFqzVETEtH79voysI9tkZ0bfjIyIcV9XeG7M\n/7UrExFRcvfTbj37kRcv/bDHK9+fc37l5feub9++s2bNiohFixaVK1du+vTpG/jD+humzIhL\nzoi6BSIiCm8bR+wSw9+Lod/GHg2jZpvo2mZdtpUX8+fG2JHxbO8otG20b7a2630/NSKiTIHo\n+2q8NzJ+mB35smLb7aNti9ihxGrG502NXsOiRKNoVnZddg8AWMHGCru5n9584HnXfDAjitTc\nqcE2K1fd78ocfNT+yy66y9h++1oR4ydNmhpRLYY8//zEKHhEu0OWP/1XtM1hLQo+/WyvXh8+\n3Dy7VattO93+7rtfRXb9iPj6nXcmFm3c4qwdhz5wY/+BC884pGBETv9+g/MKHNDmgN9mjvLv\nf9ThZZZtrsb22+eP/02aNClihbDr0qXLmDFjlv65WrVqc+as6X7OzUDRHX6tuqUqlIuImD5z\n3Tc0Ik7tFksiIjPq7xNnNY2KBf90nV/NnhcR8U63qLhrnH5WlM+MiSPjv6/FraPj3HOiUbGV\nx3/UNyZlxDHNYq3fAQBY1ca6xq7vXdcOr3PRjWfUzf2k80FHdx2Tu/phVatWXf7HzMzMiFi0\naFFEzBg58oeIKrVqrfh3f4EaNapETP366xkRu7VqWTa+HTBgckTED337Ds+3b7PsHZs0qTiv\nf78hERFL3u83YGHGPm1aF/9t9YpVq65QmZmZmb+94Zaq7IrXveXLFxGxeA0f+B+pG0/cHI9c\nHVe1i8LD4+q7o/+UtV118eKIiLzt4rwDomqJKJQVtXaJC1pF/pnx9IDIW2n0tOg9PIrUj2Zl\nVrMpAGCtbayw2+6U3p+/d8cVD7/+6CFb//h6+9Znv/3z6oblz59/TVtYOkuWlZW10vIiRYpE\nxNy5cyPy7du6RbEY+u6AnIjZffsOjYbNmpWO3bObZP3Yv/9XEfFJv34zYpcD25Rfmzdc3i23\n3NKtW7du3bo99thjI0eOXLhwdQ/42EzkW683GhTKiloN4uzTo9bseLpHfL92axUuFBFRt24s\nf01dmfpRK+KXUbFSH479OCbmxe6NTNcBwN+0sU7F1m3WYpvMiKh+YvdXRu3X9IZHjmxb9f1+\nV+y05ovpV1a8ePGImDdv3krLly5Z+mrBZq2aFnz2vQGD846e+86Axdu233/biNgvu3G+5/r3\nn9qpWL9+Y6LuCW2qr/Pu16jx65PVcnJyVt2FLcba3DwREUsWRW7+KLB89JeIhpXi6wkxYlZU\nXt1FcivZqkxERO5KtxdnRYl8ETmx0uf3v+ERBWPnmmt3DADAGm38BxRn7Xnda//9bo+jn72y\nzfFVhzx/QpW1nDQsWa/eNtF3wrffLoiGhZYtXjBq1MSIynXrloiIKNay1d4Zrw0YMOyjmf1m\nlm7VrGFEROns7AZxXf93Jxfr/1lUv6jNDuv/oLYQa3PzxKc9477PIvv0OGnF0pq/ICJWOYu6\nBtWrR8FBMfnHFZfOiV+WRJRa4UEq8XN8+XNk1I46azV1CgD8gU3xgOKMiu2e7H3T3sUnv3Rq\n64v6z1jb1XY75tiaseidHi/9stzCmb2ef3tRVDnqqL2W/ly+deudY9Sgx7q/832Bpvvvu/Tw\najdpUmHh+/3veG/IknJt2uy2Po8lPdvXigIRn34SOcstzJseX/0UUXb197SuqlCd2Kt4TPxf\njF7u2r5pX8TYiGo7xPJ3vi7+PiZHlCsfaz93CwCswaYIu4go1OCyXj1Orblk+D2HH3bXV2t3\nwVq+Pa7qcm6dRS+d3+7617+cPHvejInDXr/hqPNeyNn+rC7XNP5tvmfbVq22jyFdu34dezRr\n+utT12L37CZZvzzb5cXZJVq12WcTHfKWoljDOHb7mPVZ3PVWjJseCxfG1LHxRLcYHdGsbVRZ\ny60UiCOPiPKz4t6nYuT0WDg/xnwW974T+SvGv/ZeYeC0nyJ3lRs+AIC/ZBN9V2xElGv98Ov3\njt/r7H4Xtz6lypCnW63FKiWa3fvhhw1uuKHLec3+b9KMxQXLVK2/70lPDLr8lF2W64L6rVtv\nc9M9k6Jus2a/f69pwf2y98ro0W9O4cPbNFvjo1ZYKl80PSkqDom3Pos7Bse8xVGwWFTbNtq3\ni8YVfx0ypX9c/s4KKz3ZOZ6MiIgSe8W9bSMiiteOa86KXv3jsftjxoLIKhU7NI4O2bHNirdI\nzJsf8dvNFisbF+c9Est/8cjgbjF46Z/Kx80XRMXVrQQA/2AZeXlredkUERE5OTlLb8zt2bNn\nu3btNtwbZTx1y4bbOACwIeSddNmm3QHnJQEAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh\n7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAAS\nIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAA\nEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewA\nABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHs\nAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh\n7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABIh7AAAEiHsAAASIewAABKRual3gNXLO+my\nTb0LAMAWxowdAEAihB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQd\nAEAihB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAihB0AQCKE\nHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAi\nhB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQdAJv/4lgAACAASURB\nVEAihB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAihB0AQCKE\nHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAihB0AQCKEHQBAIoQdAEAi\nhB0AQCKEHQBAIoQdAEAiMjf1DrB6bfuO2NS78I/2avO6m3oXAGCdmbEDAEiEsAMASISwAwBI\nhLADAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMA\nSISwAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBIhLAD\nAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMASISw\nAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBIhLADAEiE\nsAMASISwAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBI\nhLADAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMASISwAwBIhLADAEiEsAMA\nSETmpt6BNVkw8dPB382KgpUbNt6u1Cqv5oz/eOjYnAoN9q1TZp0Hr2DRjInfjf9hek6+kuUr\nVa5UoWShjA1wLFue3Bnj5vwwM7LKZ1WqUGB1/44smTlhzpRfcvOXzKpavdDqPrQlc76f8/20\nxVG0yDY1ihTNv5pN5C1a9NOkedNm5hUpV7TKNqt9FwBgnWy2f51O69Gx6X+GRmx91EsjehxW\ndqVXv+92etOrR5/05pyurdZ58FIzv+xx5w13Pvn6pzNKb7/dNsUWTflu1MR55fY89F9nX3b5\ncQ2Kb9iD24zl/fjeN/fcNPaLnCLbVoipo+ZmbLfNaTfXb1Hj96ndvKnvjrzrlnFfzC1cs2qB\neRNnTS1Ypu3lO53ePOu3eMub9t4399469qPv81WslVVg2qxJuSUPuGinsw8rXuj3N8mZ1f/B\nkU89+9OsssUql8j9ccy8nLLlWp9f7/QDixXc2McLACnZ7E/F/tiz4/m9pq/fwbnfdTt2112P\nuWt0w2v7Tfpp4vBPBg8dNu6n6aNfPLvM4NuO33P3014Yn/f39npLNev9Ly46Z+ysFrs/3T/7\nvh7Zz7y2U/1pE2476fMPZ/w6IOejr/593pgf92j03wHN7u++7xMD9r+myfyXLxz8wODFSwfM\nHTTswrNHDS9b6853W3Z9ft9H+zW7qtmivlcNuum1+b9uIm/uq+d8eHPXufVuaNqjT5P7n2/W\no2/jI7b65eVLBt397uJNcdAAkIzNO+wqtj2+WbEfnul4Ye+1SLu1Hbxw6NWHnPrc2O0uen3A\nI6ftUf73OaLC27a+rFe/B1sU+vqJU057bOLf3vmNaNGkz37+YnjOgojIy50+btbXn08fPX7h\nonXdTO4vT1838acda19xUdlS+SMiClTepmP7rUoWmfPF50uTa26v28b9UKLKmVdU2GrpXG/+\nwntctmPzUvN63zbu+4iIub3vmTA1r9TR12+349Jz4vmL7H1NgzYVFg66bdTnuRERiweN6jpk\ncYmD653XKmvpZ59RusxJ51UuEQv6vzh1wd/+LADgH2yzPRUbERFzy/3r0Vsm1j/nqTMvPGZ4\n11Yl18fgH/575Z3DF291yp037rvqCdf8NU6/49YvHv5p93r5FkUUWA+HsFHMeuXCQb2KbHfL\ndRkvXTt+XKEiJRfnTBizsNgetf5za51dykVE7o/DZ0yZt8b1i1crU2PrjMWDxr8zOaP+adtU\nXO6lMsfs+cIxv/0wYfKAEVGoTflGhZcbUbDc3vvk7/Pq9wPG1jq+0i9fDY+oUK5h1eUG5C+b\nfUChV7pNHvhR/YZ7xdzM4vsfVaVcq7KFlhuSsVXh0hGzFi7J/bsfBQD8k23eYbdw4eIaHR+7\nscdOFz7V/t/HDn+05R9d+7Z2g39+tWf/BbHV4ccdUHh1L0dGvTPuv3997PvGlC+zQMS0Cbfe\nvNWZDzZvWj1fxJKpL/yv/bWjbri9dI9byheMRV89/83r49a4/vYn7d5h68yxH/8yJ7Jq1y8Y\nsWT2pDnf/5Sbv2RWteqFCv52b8SiETPGR1SrWXzFf2vyVa1RNGL2t18tiTK58/MiimcWXXH7\n5SsViZgxemRO7FWk5B41z95jpffP/fa1HyZE/vpNymX97c8CAP7BNu+wy8vLi4ztznusc4+d\nLnms/SXHfPXQ/mtOu7Ub/Nmnn+ZFNGzUaJ1uf+3evfvPP/8cEYsXL65cufLUqVPX9VA2vJy8\nnc+u37T60pPr+cofVm3Pm3/sP/Sn0VG+bhRu1qlxsz9Zf8mkcfMiyhSYMPq6K8YMX1S4XMEF\nE8fMz6tU/uhrdzp+r0IZETN+XJAbUXqFubaIX5fM+unHBVEiq3zxiMlzJ+fGNsvdCTt75qKI\n+Hna/Igiy5ZOnz1s1Px5v8z7pv+4V95dvNuZu110TJEAAP66zfsau6XybX/RE513LTThkTMu\nHTj37w7+5ZfpEVGmTOl12oVevXo99dRTTz311DPPPFOhQoXMzM0xiEs22G253cpfsHSpiOkL\nZ63t6ovnzI6IGc91/qHS5fs988Z+D/Q64MXejfbN/+NTHYd2/zovIubPWxwRBQqu/C/N0iU5\n8xZHRrnWR2RlzJn80gvzl91+8svkZ1+cGxEL5i9ZYbWxU7o9+G23h0e91G9x7cOqt2xZuuzq\nnooCAKy1LSHsIvLvcPETV+2SOfbh0y8fmPP3BhcpUjgicnLmr7riFq5g4TIllv85IyNfxJK8\ntb5qLS8vLyKWVD1p59P2LLy0EAtUrXzeJZVLLJzZ4/GpiyIyC+SLiCW5K98ynJubFxEFCuSL\nyLdjx4bH7Jjvkxs/uOq+iYM/nDrw2RFXHvvVtB3KFIsoXGTFcGu0/W1d936gV/Pn39p5r5/G\n3tBuwM1vpvdPBQA2ps1x5ml18te/7InLX9j1ugdOv+roYXfs/dcH16hRI+Krb7/9NmKntX/7\nJ598Mjc3NyLmz59fqVKlJUuW/Okqm5W1uXmiQLESEZG53Y5Flz9LXXi3stvHpI+HzRgXFSqW\nLBAR8+Yujlgh0ebNWRwRxUsWiIgoWvbUZ7Mb9RzzzuDvX/kqf6lqpZrd2qTJz8PbvhdVtl79\nZY0FtyrT9ub63zT76O0bRrVuWb/hlvHLBgBshraUsIvIbHDlE5e9uMf1d5921VGfn5P/j8/a\nrXlw3ebNK3b66ps33hhz/U41Vrvu4p+nzixVfsXTgkWL/no7QIECBZYW3pZlrW6eqLJtVkTO\n7JlLVpjILZSZlT9i0ZLciGI1ipWOn6ZOnhexwmV2UybNjShUvdZvT47JX7jhsXUbHrtswPeP\nzl4UhbarWzgib+bYGWMnLSm7c9kqxZbbRKGSNbeN+GzGt5Oj4TZ//3gB4J9pywm7iIKNrnri\n3y81vuXuUzs1OfnPbp9c0+CMvc8+f9cHLvv43s6vnfnUwateabdk5B0HNuw89eTnPnmk7arf\nQLalWpubJ6Lm3luVenT8qK9mLjm49LKymzhnYm5kbFusckTU23rXEuPe+d/PU6L0co9EmfPF\nRwuieJXdGkTEkqn/mzz0i6h9wja1l03P5QwZMCtKVWvcKCJidNehl72wpOk9ra5ovvzU3Pxf\npkVkFCi2fO0BAOtmyzrtVXD3ax+/oHaMuP2iR/78AcJrGlzr4qdu3afED/89+bBbhs5YcZXF\nY587+aCrhhZscEbHVulU3drKt2uNw3fMmNZr1NtTf7+KLnfYMxPGRuZeh1YsHhEFtz78uGL5\nRo5/fuiySct5Q77r/XVG9eNr7FYoIvIt+uK7++/68r+vL3vS8KyBX7/web4dTq/ZKDMiMuof\nWKls5H745LhJy018LvjfhAGTosCuFXdf9at+AYC1tSXN2EVEFN7j+scveGW/O0ePjyj6Fwdn\n1r2gz4clOh5/8RV713nluDNOaLVrzbL5pk8YMeilLk+8PW3b4x8d8vDpO/4Tv7Q0o9hRtzf4\n9pRhdx330dRzqu9QYuHYAWO7v7SgyuG7nH/w0oc1Z9TqsMuZwwY/dMHgJWdt27hWgbmjp7zw\n4MTcfepd2aHE0t8Rtjmu/vF9hzx9y5DbZtbad/v8M4ZN7PH4jwUP3vmqk369dK/g7nWvPm/2\ntfeNuODEuce0K1e1dN7PI354qev3P1eudOF1VcttsoMHgARstmFXqEqjJk1K7FB2lReK7H3D\n451GXdNvVhSpU+avDI6IyKp3atfPD7+kz0u93v7w/ecfeSUns0zFKlX3vbL3g0dlb1tsnR5x\ntznIrLRT2QYzi5VYdeEvxf/46zpWkr9q1WtfKf3hixMGDRg3cl5G0Upbn95l1xZ7Zy27pK5g\nicMeaVrvrfFvD5z88vtLCpUttl+n/Q5sWWLZZ5ZV5qSu2Q1eGf/ux5NfHZxXpELxQ++q22q/\nostdlJe5Y4fGTzaZ0qf3tJFvj//fwozCZbP2/feezQ/aqpLH2AHA35Kx9CEXrKWcnJysrKyI\n6NmzZ7t27TbcG7XtO2LDbZw/9Wrzupt6FwBgnW1Z19gBALBGwg4AIBHCDgAgEcIOACARwg4A\nIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEcIO\nACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHC\nDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACAR\nwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAg\nEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4A\nIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEcIOACARwg4AIBHCDgAgEZmb\negdYvVeb193UuwAAbGHM2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAk\nQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEA\nJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgB\nACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELY\nAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC\n2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAkQtgBACRC2AEAJELYAQAk\nQtgBACRC2AEAJCJzU+8AqzfloN1Xu7xi74828p4AAFsKM3YAAIkQdgAAiRB2AACJEHYAAIkQ\ndgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2AACJ\nEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAA\niRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYA\nAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2\nAACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQ\ndgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAiRB2AACJEHYAAIkQdgAAicjc1Dvw\nV0zue2+XD36J4rufelGbqhkrvzrhrTufGJJvn/YXNK/0+4+zarW97IRGhVcauXjGmP+99+EX\nY3+YnpOvZPlK29TcJXuf7Uv8Y1s396dh/fp/8s24KXMzy1ar17h50/pbFVjD0CVjet/534/n\nlNmn/XlLP+XlLZrw5sNdh/68dfPzztynzAbeZwBgOVtkxUzue2/nzp07//vo07uMX/XVCW/d\n2bnz3X0nL/9j56c/nb/CoJmfdjmrSbXytdte8fQH3/4wa/r4j17vcvFBdSpUa9z+0WFzN/wh\nbG6m9b/hoNrVdz3lrj4jfpwxedhLnY5osE3NNv83eNbqBi/49Pojj/xP586d7/39U/79pTEv\nn793wzbnXdu588Mf/LIRdhwAWGaLnLGLiIgiRYrMeeeSM5448O1Tt1nHVX9668x9DntkfM0z\nun5517F1iv62OPfHwbefeOhl7fcaOvqNQf/XpOgfbCExYx88su3VH2xzTv8R9zUpGRERiy6/\nrdnOl1x22H8ajXvkgBVnOmf2O++o64YVKFZkwZwVt7JgVI+Lj27fdXG72/+TddZtAzfWzgMA\nv9kiZ+wiImLb0y49qtysdy7u8MTKs0Z/4qdnzjj2kW8Kt33k7S7LVV1E5N96r0tf6nX+dvOG\n3Xb6dUMXrde93UB+7H9/p063vzEhYt6kIS8/cd9tN910x8M9P5qyeJ228kmXu96bm//AK278\nteoiokDtDqc2jZj6+usfrzh2SvdTj+8ytvq5nU/ZeuXNvHVN+/473ffR0MdO2H7lk94AwMaw\n5c7YzSt20H33fNHv+JcvPqtb61f+VXFt1xv+4I29ZkSdq248cZWLwyIia68ru/bcJ2OP1ruv\n6fKyzcqsQY917jzmhFIlH3/4tmmNDti9csbk9++95t+X1vv3C+/edkDpiIWfPn3Tq6PXuH6F\nZuecuV+5Su3u6Flv4bYtSyz/Uv58+SKiePHiyy3L/fquYzu8NH3nq3vfst8LL6+8sZ2vHvy/\nunWLRsxZ+RUAYKPYcmfsFi9evPVxD95zUOkZr15w5rNT13a1sW+8PjKiRtu29dYwYKvG7Y7c\nq2rRVe7J2CxlZmZGzHnuyid26P7Je90fuP22+7sPGnh9w9lf3H7pQ2PWeisVG7Vtd/yRu5Zb\nbtH09666850o0vjME+v/vmzekKuOvHRgvqa397x210Kr2UzVunX/QaevAWAztOXO2OXl5UVU\nOP7hO5/b8ZRXzzv72eYvHLvVWqw2cuTIiKhbt+46vdldd901derUiMjNza1Ro8bEiRP/yi5v\nIHlF2115VaPfZtby1W7TquYVn3/xySeLo0ZmwUYndGq01lv66b2H7+87bsakr/r3HV5k13P+\n+0ynf233W/r/8kbHo/9veJkjenY/t1b+9X4MAMB6sOXO2P2q8skP3dayxE8vnnfuC9PWZvzs\n2XMiomjRrHV6l8GDB/ft27dv377vvvtu6dKl8+ffrNKmTsOGy1/UVrJkyYgls2b9xZt7MzIy\nCxctuGTe9Mljxvy0cOmyvImPn3jiU5O2Pbv74+0q/O39BQA2jC13xu5325zxyG3P1evQ45xz\nj2n63KFl/2R0yZIlImbOnDkrouSfDF1OsWLFSpQoERF5eXkzZszIy8v7O3u8vpUqVWrVhX9l\nH8vtd2an/SIiInfKa2dmH3pIo8+6De91Qrlhtxx17utzdunU545m6/ChAQAbWQJhF1HtjEdv\nea7+OT06XnBsdrdDMjL+6Pq4ejvuGDFo2LBhEfuu/Ts88cQTS/+Qk5OTlbVus32b1FrdPLGa\nF/JXPPj2yw948pRXbnpoxLE7XX/1kJySezRc8trNnV77dcDkD2ZE5PugS6dOvUvseepFrapu\nkN0HANbFFn8qNiIiMmp0fOzGfbOmPH3Wha/PyMz8o1rd5sCDd8qIyS/1/HBNjwT5ZWivN76Z\nvVlNyW1YU9598Por//PYxyt9ICUrVy4aMWbMmIzaR1x17bUXtNpmy7ihBAD+uZKYsYuIjBrn\nPXZDz50u6nrmRfXbFYuYvcaR23e45pi7jnj20X/ffub7l+24yvH/3Ou8ww97Zv6xL47qfngC\n34e1NjdPzB73+nU3vVEh57BTdm283LWDP44aNSuidvXq+erV67TKPcQfz+/66IeZ+7Tv1GnX\n9bzLAMBflcaMXUREvu3Pf7zznoUmPdn58dV8z9hySh/+4HMXNsgccmWLw+4YNG3Jcq/MHdXj\nrKbHPjOj3jlP35dC1a2l4kecc2LFmPTwf278/Pcn0OX98uZVd7wfBfdqf9KangsDAGxuUpmx\ni4jIt8PFj1/9/M5XfTw74g+zrFTTOz8Y1OCK86+6Yp8a9+/d6oBda5bNN33CiMF9+o4tsc+p\njz9966kN/lEPZCvR+v63H1149EWd9qz3zqEtd6+aNfPbga+8Nixjn4tf6nHhdmu5kXkf//fW\n3kufnbfw09ER8UPfezvNLxMRka/eUdccuW7PlwEA/oItMuwqNT/v2sKLG6/6FbH561763+4Z\nPb5cGKX2+e17Jaq2uujaUrNqNVrxW66KNzj5vnf/dcOYoQM/HDZ2yi85mdVb7HnI2ffvu1fN\n4lvSlWSl9ml/7bWTazUsuOrCGo3W4Xu9itQ7vduXR17zXt+BX4yZOie3YosLD7t6//13qbTm\nTVRqfsG1hfPts7pv74iCjU64du0fngcArC8Zm9mTOzZ3v98V27Nnz3bt2m24N5py0O6rXV6x\n90cb7k0BgC1aOtfYAQD8wwk7AIBECDsAgEQIOwCARAg7AIBECDsAgEQIOwCARAg7AIBECDsA\ngEQIOwCARAg7AIBECDsAgEQIOwCARAg7AIBECDsAgEQIOwCARAg7AIBECDsAgEQIOwCARAg7\nAIBECDsAgEQIOwCARAg7AIBECDsAgEQIOwCARAg7AIBECDsAgEQIOwCARAg7AIBECDsAgEQI\nOwCARAg7AIBECDsAgEQIOwCARAg7AIBECDsAgEQIOwCARAg7AIBECDsAgEQIOwCARAg7AIBE\nCDsAgEQIOwCARAg7AIBECDsAgEQIOwCARAg7AIBECDsAgEQIOwCARAg7AIBECDsAgEQIOwCA\nRAg7AIBECDsAgEQIOwCARAg7AIBECDsAgEQIOwCARAg7AIBECDsAgEQIOwCARAg7AIBECDsA\ngEQIOwCARAg7AIBECDsAgEQIOwD+v717j26yvuM4/k3S5laaUmorpLQVKhS5t4IiSBXxUpwo\n3qaIbqJO8YLMOkXPcIgoXjnq1KEy3FCmE/WgTBQcUxzYnkZE5DI6oMVa2kqLRVIDKSV59kcp\nNF1LHjUlD7+8X3+1T3755ndOvuebT9PkeQAogmAHAACgCIIdAACAIuKivQG0r8f7nmhvAQAA\nHGd4xw4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATB\nDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABA\nEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsA\nAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ\n7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAA\nFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUERft\nDRyvNm7cmJycHO1dAACA2DJmzBiLxdLRrSZN047lbo53fr/f7XZ7vd5AINB5j9L8hGmaFgwG\nO+9RgGZms9lkMolIp3Y1cNjh1yRaDseAyWQym80iEgwG1cg8Pp/P6XR2dCvv2P04dru9vr4+\nJydn69atnfQQ8fHxgwcPFpHq6uqamppOehTgsNTU1J49e4rIpk2bGhsbo70dqK9v375Op9Pn\n85WWlkZ7L1CfzWYbMGCAiFRWVtbW1kZ7O52OYPdTrFy5sqmpqZOK19fXT5kyRUQKCwsvu+yy\nTnoU4LClS5cuWrRIRFatWpWWlhbt7UB9995779dff52bm7ts2bJo7wXq27lzZ2FhoYjMnDmz\noKAg2tuJAIfDcZRbCXY/RUZGRucVT0xMbP4hOTm5d+/enfdAQLNu3bo1/5CZmel2u6O7GcQC\nm80mIna7nRGHY6D5oyYikpKSEgstx7diAQAAFME7doZjNptdLpe0/FELdDabzdbccof/rgU6\nVUJCgsvlOsqnv4EIslgszSPOarVGey/HAt+KBQAAUAT/igUAAFAEwQ4AAEARBDsAAABFEOwM\nxbf5zQeuHt2vu8tudXTNGHz+zXM/rubE7OhQ04537x7RzWwyDX1sezs31699pfDS07NTu9is\nzpRewy6e9qLnu5DP1AZq/v3MbRfmZnVzWm1d0vqMvPK+NzY1hNYI35M6ikAFgdriP91WMCgj\n2WG1Ortl5Rbc9OSKipDzeTbuWDZn8tiBPbs6rHZXj/5nXzd76fY2J7wO25M6ijAnY8PBOs/C\n+y4f2T8rxWm1u3rkjLzirheL60KuxhSRCRaJnjQYDUZxYN2jZyRK6rmzlv6ndn/jDztL/vqb\nAXZzxqS3d0V7ZzAi//bFdwxL6pKTky4iQx7d1vZ27yd397OZT7r06Y+31/v9e8tXPX95L4ut\n37RV3pYVu5ZMyrTY+l+/oKjC6/fvLv1w9nndpcvwR9c1tqzQ0ZPhi0AFgR2vjEsRS9b4p5Zv\nqWvwVm9Ycv+ZKSKpFy+sOLQiWPbyuFRxDS/8+7pqn3/ftxsW/+4Ma86HmgAACFZJREFUl6QU\nvLwt2FIkfE+GL8KcjA2B8gXju5vMGeOf+HDLrgbfdzuKFt482CmWzF+/v7tlTSQmWCR60nAI\ndoZR/vTp8ZJ81Vt7jxwKbJ6da5LUm5bvj962YEzr5+S5EgZc/9qW2vkXtBfsguum9zFJ5u2r\nW+Ur/+ppmSJ97vs8qGma5v/wxhPFNHj25sCRFXuXXJUs8SOeKm/+NXxP6igCFdT9ucAu0uuu\nz1rNouqXxthE3FM/C2qapu1+7ZIEsY6dt7PVvWrmjbVLwkWLml+Jw/ekjiLMydhQ/7dLnSK9\npn7a6kkNfPn7U0Rk+ONlmqZFZoJFpCeNh2BnFJtnDRJJvOGDgyFHK+YOF0m48s19UdoVjGrF\nrBsWbPZpmtbQbrALfjo1TaT3vZ7Qe31xTy+RtFs/DWravneudIqc9kRFyILAssldRU6ZtVnT\n9PSkjiJQwYGVhUN69Rg4vSTk6H9nDRKRM56t0jRt1wtjLGK58FVvyArvwvEWMY95oVrT05M6\nijAnY8SaB0f1P3nI/atDDgYW/9Ik0mPqp5oWmQkWkZ40ID5jZxA/FBdvEhmcm2sJOZyZl5ci\nvqKir6K0LRjV+X9YcEP/jk/vWl5cXCum3NwhoYcH5+VZpLaoqExkQ3HxPknJzc0MWWDOyxsi\nsqWoaI+ungxfBEqIHzt3fXn1xsdOCznafEJru90mEvy82BOQk3NzE0NWJOblnSzBkqKSoJ6e\nDF+EORkrRs1cs3nb+jlntj7W9MXaDZo4Tj21v0hkJlgketKICHYGsaO8XBNLevqJbY673W6R\nqrIyQ39QE4ZTXl4ukpqe3uY063Fud5pIWVmZNJWX7xRJT09vc8cT3W6LSFlZuZ6e1FEEyjrw\n1aLFG8WWf+m4FJFd5eW+9lrB7XaL7Csrq9HRkzqKMCdjUtD//Tdr33lowoQnK065bv6z150g\nEpkJFomeNCKCnUE0NDSI2B2Otpd0ar7ojtfrjcamcNxqaGgQcTgcbY87nU4Rn9cbbGhokPZW\nmJxO+6GGC9+TOopAUd8tv+2yRzbFD5w+95YMkQ5bodUAC9uTOoowJ2NO6cNDTRZHctbwK5+p\nKXh2xSevTOptFum44378BPuZPWlEBDsjaedSnZqmtX8DcFSmo/ZTy6IOVpjCrtCxhK5V1YHt\nb/xq5IQFFe6Jr37w4LBW73b8/1Me0iy6ejJcEeZkbOk3Y70W8H9ftfGDR/O3zj67/6i7lrd6\nl+xnT7AI9aTBEOwMIikpSWS/z9f2yr0+n09Emq9fDOiVlORqaZ4QPp9PpIvLZXYlJUl7K4I+\nn/9Qw4XvSR1FoBitbtXMsSOuea0ye8o7ny26KuPQC1tSB63QaoCF7UkdRZiTMclsS3IPLLjl\nmY/eui3B88ykwne8omf4hO+WSPSkERHsDKJXdrZZAlVV37Y5XllZKXJSTo613XsB7cvOzhbZ\nXVXV5jNHjZWVtSI5OTkSl52dJVJVVdXmjlWVlUEx5eT00dOTOopAJQe3vTpx+PkPFTsu+uOa\n4nmXZBx5/UjLzk5srxUqKytFXDk5PXT0pI4izMnYoR34Yd+B0EOOs8afmyj1K1aslchMsEj0\npBER7AzCeeboPJH1Hk9oJ5eWlHila37+oChtC8epk0aPzhDN41kbevjzEo8mmfn5WSKDRo9O\nknqPpyxkwYGSkvUiufmjE3X1ZPgiUEZgx+uTzpn8Zv2gacs8703NC32rwnT66FFW2e7xhH4T\nuq6kZIfEn5k/wqSnJ8MXYU7GiK1z8qy2xAvm7wo9fNDr3d/yb9BITLBI9KQhReMcK2hP9fyx\nDukyfmGrUx42ldzTVyTzrjVN0dsWDK7989hp2pbZQ+Ok++QVvlZLl01OFUveI83nl2taPS1L\npO/9X7TqrtqF45ziPO+lmuZfw/ekjiJQgu+ze/rFiW3gnZ/Ut79g75JJyWIZObfsyKFg2eOn\nWSRl4nuHzgEWtid1FGFOxoaSe3qLxJ/+eGnrqzs0fHhjdxHnhEV7NS1CEywSPWk8BDvjCJTO\nOz9Fup01Y+mWuv2N3oo18yZmx9kG/Hbl3vD3RczqKNhp+z+fPayLKf2iJ/+1fY/fv2fbR48X\n9DC5Rjy8zt+yYs/K23Ns8X2umV/0TUPj/rrN704flWxKvWD+1sNnctfRk+GL4PgX2DBjgFni\nhj688WDHi6rfvDrD7Bx6x+vranyNvup1r00Z4rBkXvv2kZO4hu/J8EWYk7GhfuUdp9jF3nfi\ncyu3VHt931d+ueSh89LNknz281tamjASEywSPWk4BDtD8W//x+xr8/ud2MUa7+jWa/iEu1/5\nYk+09wQDqps3tqP34C1XvXV4mXfT69OvGJF9gjPe6kzrM+rqGYtLfwipE9ztefHO8XlZyY54\nW2KP/udcP2f5122u8Rq+J3UUwXHuqweO8nnJX/yl4dCyg1Ufz735vEHpLlu8vWvPoeNufW51\nTZuAH7YndRRhTsYG76Y3Zkw6a0B6V0dcnDUhtfewi26d+8+KkNkSkQkWiZ40FpOmtf3GCAAA\nAI5HfHkCAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAE\nwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAA\nQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7\nAAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABF\nEOwAAAAUQbADAABQBMEOAABAEf8DS9AwcDLutK4AAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#isoform stat\n",
    "trannovel = gtfinfo[gtfinfo$novelGene == F,]\n",
    "#trannovel = trannovel[trannovel$TXNAME %in% fsrawdp$transcript_name,]\n",
    "table(trannovel$txClassDescription)\n",
    "trannovel$type = \"NNC\"\n",
    "trannovel[trannovel$txClassDescription %in% c(\"newWithin\",\"newLastExon\",\"newFirstExon\",\"newFirstExon:newLastExon\"),\"type\"] = \"NIC\"\n",
    "trannovel[trannovel$txClassDescription == \"annotation\",\"type\"] = \"Known\"\n",
    "#trannovel[trannovel$txClassDescription == \"allNew\",\"type\"] = \"NNC_all\"\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "trannovelst = trannovel %>% group_by(type) %>% summarise(number = n())\n",
    "trannovelst = trannovelst[order(trannovelst$number,decreasing = F),]\n",
    "trannovelst$type = factor(trannovelst$type,levels = trannovelst$type)\n",
    "p1 = ggplot(trannovelst,aes(y = type,x = number,fill = type)) + \n",
    "    geom_bar(stat = \"identity\",width = 0.7) + \n",
    "    geom_text(aes(y = type,x = 0,label = paste0(\"n=\",number)),\n",
    "              hjust = -0.2) +\n",
    "    xlab(\"\") + ylab(\"\") +\n",
    "    scale_fill_npg() +\n",
    "    theme_pubr() \n",
    "p1\n",
    "ggexport(p1,filename = \"result/figure1/bambu_novel_isoform_stat_bar_25_10_13.pdf\",\n",
    "           width = 4,height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dd647ff-0ccb-48a5-89c5-b824fbb733d6",
   "metadata": {},
   "outputs": [],
   "source": [
    " #isoform number stat\n",
    "genest = fsraw.cor %>% group_by(gene_id) %>% summarise(readnum = length(unique(readid)))\n",
    "genest = genest[order(genest$readnum,decreasing = T),]\n",
    "  #total basic number stat\n",
    "isoformst = fsraw.cor[fsraw.cor$gene_id %in% genest[genest$readnum > 100,\n",
    "                                                     ]$gene_id,] %>% \n",
    "    group_by(gene_id) %>% \n",
    "    summarise(trannum = length(unique(transcript_id)))\n",
    "  \n",
    "  mtransnum = mean(isoformst$trannum)\n",
    "isoformst1 = as.data.frame(table(isoformst$trannum))\n",
    "\n",
    "isoformst1$prop = isoformst1$Freq/sum(isoformst1$Freq)\n",
    "isoformst1f = isoformst1[as.numeric(isoformst1$Var1) <= 15,]\n",
    "isoformst1f2 = isoformst1[as.numeric(isoformst1$Var1) > 15,]\n",
    "isoformst1 = rbind(isoformst1f,data.frame(\"Var1\" = \">15\",\n",
    "                                            \"Freq\" = sum(isoformst1f2$Freq),\n",
    "                                            \"prop\" = sum(isoformst1f2$prop)))\n",
    "isoformst1$prop = round(isoformst1$prop,3)*100\n",
    "  \n",
    "  pis1 = ggplot(isoformst1,aes(x = Var1,y = Freq,fill = Var1)) + \n",
    "    geom_bar(stat = \"identity\") +\n",
    "    scale_y_continuous(limits = c(0,6500)) +\n",
    "    geom_text(aes(label = prop),hjust = 0,angle = 90) +\n",
    "    ggtitle(paste0(\"Mean isoform number = \", round(mtransnum,3))) +\n",
    "    scale_fill_viridis(discrete = T,option = \"A\")+ xlab(\"\") + \n",
    "    ylab(\"Gene number\") +\n",
    "    theme_pubr() +theme(legend.position = \"none\")\n",
    "  pis1\n",
    "  ggexport(pis1,filename = \"result/figure1/bam_iso_basic_stat1_isoform_number_gene_100read_25_10_10.pdf\",\n",
    "           width = 5,height = 4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
